---
title: "DE PCAs"
author: "Jill Ashey"
date: "2024-05-20"
output: html_document
---

Script for plotting PCAs for differential expression of mRNAs, miRNAs, and lncRNAs

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(DESeq2)
```

### mRNA
Read in counts data for differentially expressed genes 
```{r}
mrna_de <- read.csv("../output/Molecular/mRNA/Unique_DEGs.csv")
```

Read in metadata 
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv")
meta <- dplyr::arrange(meta, ID) #%>% # rearrange metadata so IDs are sorted in descending order 
  #mutate(Treatment = ifelse(Treatment == "Acclimation", "Ambient", Treatment))
  
# Set variables as factors 
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP0", "TP5", "TP7"))
meta$Treatment <- factor(meta$Treatment, levels = c("Acclimation","Ambient", "Heat"))

# Remove sample that did not sequence well 
meta <- meta %>%
  dplyr::filter(ID != "AST-1105")
```

Prep data and merge dfs together
```{r}
# Ensure the first column of mirna_de is set as row names
row.names(mrna_de) <- mrna_de[, 1]
mrna_de <- mrna_de[, -1]

# Modify meta IDs to match the column names in mirna_de if necessary
meta$ID <- gsub("-", ".", meta$ID)

# Prepare colData for DESeq2
colData <- meta
row.names(colData) <- colData$ID
colData <- colData[ , -1]  # Remove the ID column as it's now row names
```

Stabilize variance 
```{r}
# Ensure that the columns of mirna_de match the rows of colData
mrna_de <- mrna_de[, row.names(colData)]

# Create DESeq2 dataset
dds <- DESeqDataSetFromMatrix(countData = mrna_de, colData = colData, design = ~ 1)

# Perform variance stabilizing transformation
vsd <- varianceStabilizingTransformation(dds, blind = FALSE)
```

Plot PCA
```{r}
# Define your custom colors for treatments
custom_colors <- c("Heat" = "red", "Acclimation" = "grey", "Ambient" = "blue")

# Plot PCA
pca_data <- plotPCA(vsd, intgroup = c("Timepoint", "Treatment"), returnData = TRUE)
percentVar <- round(100 * attr(pca_data, "percentVar"))

# Plot with ggplot2
mrna_plot <- ggplot(pca_data, aes(x = PC1, y = PC2, color = Treatment, shape = Timepoint)) +
  geom_point(size = 6) +
  labs(
    x = paste0("PC1: ", percentVar[1], "% variance"),
    y = paste0("PC2: ", percentVar[2], "% variance")
  ) +
    scale_color_manual(values = custom_colors) +  # Set custom colors
  theme_bw() +
  theme(
    axis.title = element_text(size = 14, face = "bold"),
    axis.text = element_text(size = 12),
        legend.text = element_text(size = 12),       # Change legend text font size
    legend.title = element_text(size = 14, face = "bold")  # Change legend title font size
  ); mrna_plot

# Save plots 
ggsave("../output/Molecular/mRNA/pca_plot_DEGs.pdf", mrna_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_DEGs.png", mrna_plot, width = 10, height = 8)
```




### miRNA
Read in counts data for differentially expressed genes 
```{r}
mirna_de <- read.csv("../output/Molecular/smRNA/Unique_DEMs.csv")
```

There are some predicted miRNAs that look like duplicates. For example, chromosome_4_131850, chromosome_4_131836, chromosome_4_131810, and chromosome_4_131796 all have the same number of counts, which makes me think that it is just one miRNA in this region of the genome. Look into the mirdeep2 plots

Read in metadata 
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv")
meta <- dplyr::arrange(meta, ID) #%>% # rearrange metadata so IDs are sorted in descending order 
  #mutate(Treatment = ifelse(Treatment == "Acclimation", "Ambient", Treatment))
  
# Set variables as factors 
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP0", "TP5", "TP7"))
meta$Treatment <- factor(meta$Treatment, levels = c("Acclimation","Ambient", "Heat"))

# Remove sample that did not sequence well 
meta <- meta %>%
  dplyr::filter(ID != "AST-1105")
```

Prep data and merge dfs together
```{r}
# Ensure the first column of mirna_de is set as row names
row.names(mirna_de) <- mirna_de[, 1]
mirna_de <- mirna_de[, -1]

# Modify meta IDs to match the column names in mirna_de if necessary
meta$ID <- gsub("-", ".", meta$ID)

# Prepare colData for DESeq2
colData <- meta
row.names(colData) <- colData$ID
colData <- colData[ , -1]  # Remove the ID column as it's now row names
```

Stabilize variance 
```{r}
# Ensure that the columns of mirna_de match the rows of colData
mirna_de <- mirna_de[, row.names(colData)]

# Create DESeq2 dataset
dds <- DESeqDataSetFromMatrix(countData = mirna_de, colData = colData, design = ~ 1)

# Perform variance stabilizing transformation
vsd <- varianceStabilizingTransformation(dds, blind = FALSE)
```

Plot PCA
```{r}
# Define your custom colors for treatments
custom_colors <- c("Heat" = "red", "Acclimation" = "grey", "Ambient" = "blue")

# Plot PCA
pca_data <- plotPCA(vsd, intgroup = c("Timepoint", "Treatment"), returnData = TRUE)
percentVar <- round(100 * attr(pca_data, "percentVar"))

# Plot with ggplot2
mirna_plot <- ggplot(pca_data, aes(x = PC1, y = PC2, color = Treatment, shape = Timepoint)) +
  geom_point(size = 6) +
  labs(
    x = paste0("PC1: ", percentVar[1], "% variance"),
    y = paste0("PC2: ", percentVar[2], "% variance")
  ) +
    scale_color_manual(values = custom_colors) +  # Set custom colors
  theme_bw() +
  theme(
    axis.title = element_text(size = 14, face = "bold"),
    axis.text = element_text(size = 12),
        legend.text = element_text(size = 12),       # Change legend text font size
    legend.title = element_text(size = 14, face = "bold")  # Change legend title font size
  ); mirna_plot

# Save plots 
ggsave("../output/Molecular/smRNA/pca_plot_DEMs.pdf", mirna_plot, width = 10, height = 8)
ggsave("../output/Molecular/smRNA/pca_plot_DEMs.png", mirna_plot, width = 10, height = 8)
```




### lncRNA