---
title: "DESeq2 - Astrangia mRNA"
output: html_document
date: "2023-10-02"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(apeglm)
```

## Load and clean data 

Read in raw count data for mRNA 
```{r}
raw_counts <- read.csv("../data/Molecular/mRNA/Apoc_gene_count_matrix.csv")
rownames(raw_counts) <- raw_counts[,1] #set first column that contains gene names as rownames
raw_counts <- raw_counts[,-1] #remove the column with gene names
```

Remove extraneous info from sample name (eg trimmed.AST.1065_R1_001.fastq.gz.bam.gtf should be AST-1065)
```{r}
colnames(raw_counts)<-gsub("_[^_]+$", "",colnames(raw_counts)) #get rid of "_001.fastq.gz.bam.gtf"
colnames(raw_counts)<-gsub("_[^_]+$", "",colnames(raw_counts)) #get rid of "_R1"
colnames(raw_counts)<-gsub("trimmed.", "",colnames(raw_counts)) #get rid of "trimmed."
colnames(raw_counts)<-gsub("AST.", "AST-",colnames(raw_counts)) # replace "AST." with "AST-"
colnames(raw_counts)<-gsub(" ", "", colnames(raw_counts))

# rename columns all together? if they are not matching with the meta$ID 
colnames(raw_counts) <- c("AST-1065", "AST-1105", "AST-1147", "AST-1412", "AST-1560", "AST-1567", "AST-1617", "AST-1722", "AST-2000", "AST-2007", "AST-2302", "AST-2360", "AST-2398", "AST-2404", "AST-2412", "AST-2512", "AST-2523", "AST-2563", "AST-2729", "AST-2755")
```

Read in metadata 
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv")
meta <- dplyr::arrange(meta, ID) # rearrange metadata so IDs are sorted in descending order 
#meta$ID <- gsub("AST-", "AST.", meta$ID)

# Set variables as factors 
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP0", "TP5", "TP7"))
meta$Treatment <- factor(meta$Treatment, levels = c("Ambient", "Heat"))
```

Data sanity checks!
```{r}
meta$ID %in% colnames(raw_counts) #are all of the sample names (rows) in the metadata df in the gene count matrix? Should be TRUE. 
all(rownames(meta$ID) == colnames(raw_counts)) #are they the same in the same order? Should be TRUE
```

## Filter reads by proportion of samples containing cutoff value 
```{r}
ffun<-filterfun(pOverA(0.85,5))  #set up filtering parameters--LOOK INTO FILTERING PARAMETERS
filt_outrm_poa <- genefilter((raw_counts), ffun) #apply filter
sum(filt_outrm_poa) #count number of genes left

filt_outrm_poa <- raw_counts[filt_outrm_poa,] #keep only rows that passed filter

all(rownames(meta$ID) %in% colnames(filt_outrm_poa)) # must come out TRUE
```

## Calculte dispersions 

"Using negative binomial models requires gene dispersion estimates to be made. This can be achieved in a number of ways. A common way to calculate this for gene i is to use the equation:

Dispersioni = (variancei - meani)/meani2

Starting with raw counts, so using [DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html), but could also import a normalized gene count matrix (i.e. transcripts per million) and calculate disperson manually as in the [vignette](https://cran.r-project.org/web/packages/glmmSeq/vignettes/glmmSeq.html).

```{r}
dds_filt_outrm <- DESeqDataSetFromMatrix(countData = filt_outrm_poa,
                              colData = meta,
                              design = ~Treatment*Timepoint)

dds_filt_outrm <- DESeq(dds_filt_outrm) #apply DEseq dispersion calculation
dispersions_filt_outrm <- setNames(dispersions(dds_filt_outrm), rownames(filt_outrm_poa)) #save dispersions as vector for each gene

dds_norm_counts <- as.data.frame(counts(dds_filt_outrm, normalized=TRUE)) #save normalized counts from DEseq object for downstream use
```

## Do variance stabilizing transformation 

First, estimate size factors to make sure we can use vst on our data
```{r}
size_factors <- estimateSizeFactors(dds_filt_outrm)
print(sizeFactors(size_factors))

# All less than 4, can use vst 
```

Apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
```{r}
vst <- vst(dds_filt_outrm)
```

Make PCA plot using vst object 
```{r}
vst_PCAdata <- plotPCA(vst, intgroup = c("Treatment", "Timepoint"), returnData=TRUE) # create PCA loadings 
percentVar <- round(100*attr(vst_PCAdata, "percentVar")) #plot PCA of samples with all data
PCAplot <- ggplot(vst_PCAdata, aes(PC1, PC2, color=Treatment, shape =Timepoint)) + 
   geom_point(size=3) +
   geom_text(aes(label=name),hjust=0, vjust=0) +
   xlab(paste0("PC1: ",percentVar[1],"% variance")) +
   ylab(paste0("PC2: ",percentVar[2],"% variance")) +
   #scale_color_manual(values = c(control="cadetblue3", mid="palevioletred", high="darkgreen", unknown="black")) +
   coord_fixed() +
   #ggtitle("P. lobata - all genes") +
   theme_bw() + #Set background color
   theme(panel.border = element_blank(), # Set border
         #panel.grid.major = element_blank(), #Set major gridlines
         #panel.grid.minor = element_blank(), #Set minor gridlines
         axis.line = element_line(colour = "black"), #Set axes color
         plot.background=element_blank()) #Set the plot background
PCAplot # there is no clear grouping of Days, so I will not be moving forward with Treatment only as an independent variable in my analysis 
```

Save vst object for downstream analysis 
```{r}
vst <- assay(vst) # call only the transformed counts
vst <- as.data.frame(vst)
vst <- tibble::rownames_to_column(vst, var = "Gene")

write.csv(vst, "../output/vst_mRNA.csv") #save this as csv for downstream analysis
```




20231003
```{r}
dds_filt_outrm <- DESeqDataSetFromMatrix(countData = filt_outrm_poa,
                              colData = meta,
                              design = ~Treatment+Timepoint+Treatment:Timepoint)

dds_filt_outrm <- DESeq(dds_filt_outrm) #apply DEseq dispersion calculation
res <- results(dds_filt_outrm)
resultsNames(dds_filt_outrm)

# [1] "Intercept"                  "Treatment_Heat_vs_Ambient"  "Timepoint_TP5_vs_TP0"       "Timepoint_TP7_vs_TP0"       "TreatmentHeat.TimepointTP5" "TreatmentHeat.TimepointTP7"
## This doesn't really make sense to me. 
```

Trying something that the DESeq2 manual recommended based on the interactions
```{r}
dds_filt_outrm$group <- factor(paste0(dds_filt_outrm$Timepoint, dds_filt_outrm$Treatment))
design(dds_filt_outrm) <- ~group
dds <- DESeq(dds_filt_outrm)
resultsNames(dds)
results(dds)
colData(dds)

# multifactor design
design(dds) <- formula(~ Treatment + Timepoint)
ddsMF <- DESeq(dds)
resMF <- results(ddsMF)
head(resMF)
resMFtype <- results(ddsMF, contrast = c("Treatment", "Ambient", "Heat"))
head(resMFtype)
```

Trying to the quick start 
```{r}
dds <- DESeqDataSetFromMatrix(countData = filt_outrm_poa,
                              colData = meta,
                              design = ~Timepoint + Treatment + Timepoint:Treatment)
dds <- DESeq(dds)
resultsNames(dds)

#result <- results(dds, contrast = list("TreatmentAmbient.TimepointTP5", "TreatmentHeat.TimepointTP7"))
result <- results(dds)
sum(result$padj<0.05, na.rm = T)
sum(result$log2FoldChange>2, na.rm = T)

plotDispEsts(dds)
```

```{r}
meta$trt_time <- factor(paste0(meta$Treatment, "-", meta$Timepoint))

dds <- DESeqDataSetFromMatrix(countData = filt_outrm_poa,
                              colData = meta,
                              design = ~trt_time)
dds <- DESeq(dds)
resultsNames(dds)

result <- results(dds, contrast = list("trt_time_Ambient.TP5_vs_Ambient.TP0", "trt_time_Ambient.TP7_vs_Ambient.TP0"))
sum(result$padj<0.1, na.rm = T)

result <- results(dds, contrast = list("trt_time_Ambient.TP5_vs_Ambient.TP0", "trt_time_Heat.TP0_vs_Ambient.TP0"))
sum(result$padj<0.1, na.rm = T)
sum(result$log2FoldChange>5, na.rm = T)

```

I am confused. When I run the DESeq code with the design set to treatment + timepoint + treatment:timepoint, the results names are this: "Intercept", "Timepoint_TP5_vs_TP0", "Timepoint_TP7_vs_TP0", "Treatment_Heat_vs_Ambient",  "TimepointTP5.TreatmentHeat", "TimepointTP7.TreatmentHeat". This is not all of the comparisons between the two factors. The comparisons should be: Ambient v Heat, TP0 v TP5, TP0 v TP7, TP5 v TP7, Amb TP0 v Heat TP0, Amb TP0 v Heat TP5, 






