---
title: "DESeq2 compare"
author: "Jill Ashey"
date: "`r Sys.Date()`"
output: html_document
---

I used bowtie2 and hisat2 as aligners for my data. Now I want to compare the gene count matrices that each aligner produced. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(car)

library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(car)
library(gplots)
library(wesanderson)
library(cowplot)
library(gridExtra)
```

Talked with Zoe about the various issues I was running into We agreed that there is an issue with TP0 only have one 'treatment'. We discussed a couple of different options moving forward:

- 1) Merge the Treatment and Timepoint columns together into one column called 'Condition'. Then run DESeq2 so the design is ~Condition
- 2) Remove TP0 and only analyze TP5 Amb v Heat and TP7 Amb v Heat. Then run DESeq2 so the design is ~Timepoint + Treatment + Timepoint:Treatment OR Timepoint * Treatment 
- 3) Remove TP5 and TP7 Amb samples and only keep TP0, TP5 Heat and TP7 Heat. Then run DESeq2 so the design is ~Timepoint + Treatment + Timepoint:Treatment OR Timepoint * Treatment 
- 4) Remove Heat samples from TP5 and TP7 and only analyze the Amb samples for TP0, TP5, and TP7. The DESeq2 design would be ~Timepoint 

I'm going to do all of these iterations and see where that leaves me. Also need to discuss w/ Hollie what the best way forward might be. 

## Prep data for all iterations 
Let's try to run some DESeq2 with the bowtie data. Read in counts. 
```{r}
bowtie_matrix <- read_csv("../data/Molecular/mRNA/Apoc_gene_count_matrix_bowtie.csv")
bowtie_matrix <- as.data.frame(bowtie_matrix)
rownames(bowtie_matrix) <- bowtie_matrix[,1] #set first column that contains gene names as rownames
bowtie_matrix <- bowtie_matrix[,-1] # remove column w/ gene names 
dim(bowtie_matrix)
```

There are 47156 genes total. 

Remove extraneous info from sample names. 
```{r}
colnames(bowtie_matrix) <- sub("align.trimmed.", "", colnames(bowtie_matrix))
colnames(bowtie_matrix) <- sub("_R1_001.fastq.gz.bam.gtf", "", colnames(bowtie_matrix))
```

Remove any genes that have 0 counts across all samples (ie these genes were not expressed)
```{r}
bowtie_matrix<-bowtie_matrix %>%
     mutate(Total = rowSums(.[, 1:20]))%>%
    filter(!Total==0)%>%
    dplyr::select(!Total)

dim(bowtie_matrix)
```

After removing genes with 0s across all samples, 43967 genes are left.

Read in metadata 
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv")
meta <- dplyr::arrange(meta, ID) %>% # rearrange metadata so IDs are sorted in descending order 
  mutate(Treatment = ifelse(Treatment == "Acclimation", "Ambient", Treatment))
  
# Set variables as factors 
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP0", "TP5", "TP7"))
meta$Treatment <- factor(meta$Treatment, levels = c("Ambient", "Heat"))
```

Data sanity checks!
```{r}
meta$ID %in% colnames(bowtie_matrix) #are all of the sample names (rows) in the metadata df in the gene count matrix? Should be TRUE. 
all(rownames(meta$ID) == colnames(bowtie_matrix)) #are they the same in the same order? Should be TRUE
```

Filter data with pOverA. There are 20 samples with an n=4 samples per timepoint and treatment. Genes will pass filtering if they are present in 4/20=0.2 of the samples because we expect differential expression by timepoint and treatment. Also filtering for a minimum gene count of 10, such that ~20% of the samples must have a gene count of >10 in order for the gene to remain in the data set.  
```{r}
ffun<-filterfun(pOverA(0.2,10))  #set up filtering parameters--LOOK INTO FILTERING PARAMETERS
filt_outrm_poa <- genefilter((bowtie_matrix), ffun) #apply filter
sum(filt_outrm_poa) #count number of genes left

filt_outrm_poa <- bowtie_matrix[filt_outrm_poa,] #keep only rows that passed filter

all(rownames(meta$ID) %in% colnames(filt_outrm_poa)) # must come out TRUE

#write.csv(filt_outrm_poa, "../output/Molecular/mRNA/filtered_gene_counts.csv")
```

28720 genes remain after filtering. 

Remove AST-1105 from the data, as it did not have good mapping and has basically no counts. 
```{r}
# Remove column from gene count matrix 
filt_outrm_poa <- filt_outrm_poa %>%
  dplyr::select(-"AST-1105")
  
# Remove row from metadata
meta <- meta %>% 
  filter(ID != "AST-1105")

all(rownames(meta$ID) %in% colnames(filt_outrm_poa)) # must come out TRUE
```

## 1) Merge the Treatment and Timepoint columns together into one column called 'Condition'. Then run DESeq2 so the design is ~Condition

Merge Treatment and Timepoint into a single column 
```{r}
meta$Condition <- paste(meta$Timepoint, meta$Treatment, sep = "_")
meta$Condition <- factor(meta$Condition, levels = c("TP0_Ambient", "TP5_Ambient", "TP5_Heat", "TP7_Ambient", "TP7_Heat"))
```

Run DESeq2
```{r}
data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = ~Condition)
dds <- DESeq(data)
resultsNames(dds)
```

Look at contrasts

TP0 v TP5 Ambient
```{r}
# Look at results 
TP0_v_5_amb <- results(dds, contrast=c("Condition", "TP0_Ambient" , "TP5_Ambient"))
head(TP0_v_5_amb)

# Save results for all genes 
TP0_v_5_amb_df <- as.data.frame(TP0_v_5_amb)
write.csv(TP0_v_5_amb_df, file = "../output/Molecular/mRNA/TP0_v_5_amb.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP0_v_5_amb); ma_plot

# Subset genes with significant padj values
sum(TP0_v_5_amb$padj <0.05, na.rm=T) # genes with significant p-values
TP0_v_5_amb.sig <- subset(TP0_v_5_amb, padj<0.05,) #subset signficant pvalues with 5%FDR
sum(TP0_v_5_amb.sig$log2FoldChange <0, na.rm=T) # genes downregulated in TP5 ambient relative to TP0 ambient OR upregulated in TP0
sum(TP0_v_5_amb.sig$log2FoldChange >0, na.rm=T) # genes upregulated in TP5 ambient relative to TP0 ambient
TP0_v_5_amb.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_5_amb.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP0_v_5_amb.rsig <- rlog(TP0_v_5_amb.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
pca_plot <- plotPCA(TP0_v_5_amb.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP0_v_5_amb.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP0_v_5_amb.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP0_amb_v_TP5_amb_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP0_amb_v_TP5_amb_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP0_amb_v_TP5_amb_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP0_v_5_amb.sig <- as.data.frame(TP0_v_5_amb.sig)
TP0_v_5_amb.sig.list <- as.data.frame(counts(TP0_v_5_amb.sig.list))
TP0_v_5_amb.sig.list.full <- cbind(TP0_v_5_amb.sig, TP0_v_5_amb.sig.list)
write.csv(TP0_v_5_amb.sig.list.full, file = "../output/Molecular/mRNA/TP0_amb_v_TP5_amb_DEG.csv")
```

TP0 v TP7 Ambient
```{r}
# Look at results 
TP0_v_7_amb <- results(dds, contrast=c("Condition", "TP0_Ambient" , "TP7_Ambient"))
head(TP0_v_7_amb)

# Save results for all genes 
TP0_v_7_amb_df <- as.data.frame(TP0_v_7_amb)
write.csv(TP0_v_7_amb_df, file = "../output/Molecular/mRNA/TP0_v_7_amb.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP0_v_7_amb); ma_plot

# Subset genes with significant padj values
sum(TP0_v_7_amb$padj <0.05, na.rm=T) # genes with significant p-values
TP0_v_7_amb.sig <- subset(TP0_v_7_amb, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_7_amb.sig$log2FoldChange <0, na.rm=T) # genes are downregulated in TP5 ambient relative to TP0 ambient OR upregulated in TP0
sum(TP0_v_7_amb.sig$log2FoldChange >0, na.rm=T) # genes are upregulated in TP5 ambient relative to TP0 ambient
TP0_v_7_amb.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_7_amb.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP0_v_7_amb.rsig <- rlog(TP0_v_7_amb.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
pca_plot <- plotPCA(TP0_v_7_amb.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP0_v_7_amb.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP0_v_7_amb.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP0_amb_v_TP7_amb_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP0_amb_v_TP7_amb_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP0_amb_v_TP7_amb_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP0_v_7_amb.sig <- as.data.frame(TP0_v_7_amb.sig)
TP0_v_7_amb.sig.list <- as.data.frame(counts(TP0_v_7_amb.sig.list))
TP0_v_7_amb.sig.list.full <- cbind(TP0_v_7_amb.sig, TP0_v_7_amb.sig.list)
write.csv(TP0_v_7_amb.sig.list.full, file = "../output/Molecular/mRNA/TP0_amb_v_TP7_amb_DEG.csv")
```

TP5 v TP7 Ambient 
```{r}
# Look at results 
TP5_v_7_amb <- results(dds, contrast=c("Condition", "TP5_Ambient" , "TP7_Ambient"))
head(TP5_v_7_amb)

# Save results for all genes 
TP5_v_7_amb_df <- as.data.frame(TP5_v_7_amb)
write.csv(TP5_v_7_amb_df, file = "../output/Molecular/mRNA/TP5_v_7_amb.csv")

# Volcano plot for LFC 
ma_plot<- plotMA(TP5_v_7_amb); ma_plot

# Subset genes with significant padj values
sum(TP5_v_7_amb$padj <0.05, na.rm=T) # genes with significant p-values
TP5_v_7_amb.sig <- subset(TP5_v_7_amb, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP5_v_7_amb.sig$log2FoldChange <0, na.rm=T) # genes downregulated in TP7 ambient relative to TP5 ambientn OR upregulated in TP5
sum(TP5_v_7_amb.sig$log2FoldChange >0, na.rm=T) # genes upregulated in TP7 ambient relative to TP5 ambient
TP5_v_7_amb.sig.list <- data[which(rownames(dds) %in% rownames(TP5_v_7_amb.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP5_v_7_amb.rsig <- rlog(TP5_v_7_amb.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
pca_plot <- plotPCA(TP5_v_7_amb.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP5_v_7_amb.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP5_v_7_amb.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP5_amb_v_TP7_amb_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP5_amb_v_TP7_amb_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP5_amb_v_TP7_amb_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP5_v_7_amb.sig <- as.data.frame(TP5_v_7_amb.sig)
TP5_v_7_amb.sig.list <- as.data.frame(counts(TP5_v_7_amb.sig.list))
TP5_v_7_amb.sig.list.full <- cbind(TP5_v_7_amb.sig, TP5_v_7_amb.sig.list)
write.csv(TP5_v_7_amb.sig.list.full, file = "../output/Molecular/mRNA/TP5_amb_v_TP7_amb_DEG.csv")
```

TP0 v TP5 Heat
```{r}
# Look at results 
TP0_v_5_heat <- results(dds, contrast=c("Condition", "TP0_Ambient" , "TP5_Heat"))
head(TP0_v_5_heat)

# Save results for all genes 
TP0_v_5_heat_df <- as.data.frame(TP0_v_5_heat)
#write.csv(TP0_v_5_heat_df, file = "../output/Molecular/mRNA/TP0_v_5_heat.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP0_v_5_heat); ma_plot

# Subset genes with significant padj values
sum(TP0_v_5_heat$padj <0.05, na.rm=T) # genes with significant p-values
TP0_v_5_heat.sig <- subset(TP0_v_5_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_5_heat.sig$log2FoldChange <0, na.rm=T) # genes downregulated in TP5 heat relative to TP0 ambient OR upregulated in TP0
sum(TP0_v_5_heat.sig$log2FoldChange >0, na.rm=T) # genes  upregulated in TP5 heat relative to TP0 ambient
TP0_v_5_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_5_heat.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP0_v_5_heat.rsig <- rlog(TP0_v_5_heat.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
pca_plot <- plotPCA(TP0_v_5_heat.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP0_v_5_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP0_v_5_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP0_amb_v_TP5_heat_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP0_amb_v_TP5_heat_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP0_amb_v_TP5_heat_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP0_v_5_heat.sig <- as.data.frame(TP0_v_5_heat.sig)
TP0_v_5_heat.sig.list <- as.data.frame(counts(TP0_v_5_heat.sig.list))
TP0_v_5_heat.sig.list.full <- cbind(TP0_v_5_heat.sig, TP0_v_5_heat.sig.list)
#write.csv(TP0_v_5_heat.sig.list.full, file = "../output/Molecular/mRNA/TP0_amb_v_TP5_heat_DEG.csv")
```

TP0 v TP7 Heat
```{r}
# Look at results 
TP0_v_7_heat <- results(dds, contrast=c("Condition", "TP0_Ambient" , "TP7_Heat"))
head(TP0_v_7_heat)

# Save results for all genes 
TP0_v_7_heat_df <- as.data.frame(TP0_v_7_heat)
#write.csv(TP0_v_7_heat_df, file = "../output/Molecular/mRNA/TP0_v_7_heat.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP0_v_7_heat); ma_plot

# Subset genes with significant padj values
sum(TP0_v_7_heat$padj <0.05, na.rm=T) # genes with significant p-values
TP0_v_7_heat.sig <- subset(TP0_v_7_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_7_heat.sig$log2FoldChange <0, na.rm=T) # genes downregulated in TP7 heat relative to TP0 ambient OR upregulated in TP0
sum(TP0_v_7_heat.sig$log2FoldChange >0, na.rm=T) # genes upregulated in TP7 heat relative to TP0 ambient
TP0_v_7_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_7_heat.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP0_v_7_heat.rsig <- rlog(TP0_v_7_heat.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
pca_plot <- plotPCA(TP0_v_7_heat.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP0_v_7_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP0_v_7_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP0_amb_v_TP7_heat_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP0_amb_v_TP7_heat_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP0_amb_v_TP7_heat_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP0_v_7_heat.sig <- as.data.frame(TP0_v_7_heat.sig)
TP0_v_7_heat.sig.list <- as.data.frame(counts(TP0_v_7_heat.sig.list))
TP0_v_7_heat.sig.list.full <- cbind(TP0_v_7_heat.sig, TP0_v_7_heat.sig.list)
#write.csv(TP0_v_7_heat.sig.list.full, file = "../output/Molecular/mRNA/TP0_amb_v_TP7_heat_DEG.csv")
```

TP5 Heat v TP7 Heat
```{r}
# Look at results 
TP5_heat_v_7_heat <- results(dds, contrast=c("Condition", "TP5_Heat" , "TP7_Heat"))
head(TP5_heat_v_7_heat)

# Save results for all genes 
TP5_heat_v_7_heat_df <- as.data.frame(TP5_heat_v_7_heat)
#write.csv(TP5_heat_v_7_heat_df, file = "../output/Molecular/mRNA/TP5_heat_v_7_heat.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP5_heat_v_7_heat); ma_plot

# Subset genes with significant padj values
sum(TP5_heat_v_7_heat$padj <0.05, na.rm=T) # genes with significant p-values
TP5_heat_v_7_heat.sig <- subset(TP5_heat_v_7_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP5_heat_v_7_heat.sig$log2FoldChange <0, na.rm=T) # genes downregulated in TP7 heat relative to TP5 heat OR upregulated in TP5 heat
sum(TP5_heat_v_7_heat.sig$log2FoldChange >0, na.rm=T) # genes upregulated in TP7 heat relative to TP5 heat
TP5_heat_v_7_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP5_heat_v_7_heat.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP5_heat_v_7_heat.rsig <- rlog(TP5_heat_v_7_heat.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
pca_plot <- plotPCA(TP5_heat_v_7_heat.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP5_heat_v_7_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP5_heat_v_7_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP5_heat_v_TP7_heat_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP5_heat_v_TP7_heat_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP5_heat_v_TP7_heat_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP5_heat_v_7_heat.sig <- as.data.frame(TP5_heat_v_7_heat.sig)
TP5_heat_v_7_heat.sig.list <- as.data.frame(counts(TP5_heat_v_7_heat.sig.list))
TP5_heat_v_7_heat.sig.list.full <- cbind(TP5_heat_v_7_heat.sig, TP5_heat_v_7_heat.sig.list)
#write.csv(TP5_heat_v_7_heat.sig.list.full, file = "../output/Molecular/mRNA/TP5_heat_v_TP7_heat_DEG.csv")
```

TP5 Amb v TP5 Heat
```{r}
# Look at results 
TP5_amb_v_TP5_heat <- results(dds, contrast=c("Condition", "TP5_Ambient" , "TP5_Heat"))
head(TP5_amb_v_TP5_heat)

# Save results for all genes 
TP5_amb_v_TP5_heat_df <- as.data.frame(TP5_amb_v_TP5_heat)
write.csv(TP5_amb_v_TP5_heat_df, file = "../output/Molecular/mRNA/TP5_amb_v_TP5_heat.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP5_amb_v_TP5_heat)

# Subset genes with significant padj values
sum(TP5_amb_v_TP5_heat$padj <0.05, na.rm=T) # genes with significant p-values
TP5_amb_v_TP5_heat.sig <- subset(TP5_amb_v_TP5_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP5_amb_v_TP5_heat.sig$log2FoldChange <0, na.rm=T) # genes downregulated in TP5 heat relative to TP5 amb OR upregulated in TP5 amb
sum(TP5_amb_v_TP5_heat.sig$log2FoldChange >0, na.rm=T) # genes upregulated in TP5 heat relative to TP5 amb
TP5_amb_v_TP5_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP5_amb_v_TP5_heat.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP5_amb_v_TP5_heat.rsig <- rlog(TP5_amb_v_TP5_heat.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
pca_plot <- plotPCA(TP5_amb_v_TP5_heat.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP5_amb_v_TP5_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP5_amb_v_TP5_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP5_amb_v_TP5_heat_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP5_amb_v_TP5_heat_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP5_amb_v_TP5_heat_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP5_amb_v_TP5_heat.sig <- as.data.frame(TP5_amb_v_TP5_heat.sig)
TP5_amb_v_TP5_heat.sig.list <- as.data.frame(counts(TP5_amb_v_TP5_heat.sig.list))
TP5_amb_v_TP5_heat.sig.list.full <- cbind(TP5_amb_v_TP5_heat.sig, TP5_amb_v_TP5_heat.sig.list)
write.csv(TP5_amb_v_TP5_heat.sig.list.full, file = "../output/Molecular/mRNA/TP5_amb_v_TP5_heat_DEG.csv")
```

TP7 Amb v TP7 Heat
```{r}
# Look at results 
TP7_amb_v_TP7_heat <- results(dds, contrast=c("Condition", "TP7_Ambient" , "TP7_Heat"))
head(TP7_amb_v_TP7_heat)

# Save results for all genes 
TP7_amb_v_TP7_heat_df <- as.data.frame(TP7_amb_v_TP7_heat)
write.csv(TP7_amb_v_TP7_heat_df, file = "../output/Molecular/mRNA/TP7_amb_v_TP7_heat.csv")

# Volcano plot for LFC
plotMA(TP7_amb_v_TP7_heat)

# Subset genes with significant padj values
sum(TP7_amb_v_TP7_heat$padj <0.05, na.rm=T) # genes with significant p-values
TP7_amb_v_TP7_heat.sig <- subset(TP7_amb_v_TP7_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP7_amb_v_TP7_heat.sig$log2FoldChange <0, na.rm=T) # genes downregulated in TP7 heat relative to TP7 amb OR upregulated in TP7 amb 
sum(TP7_amb_v_TP7_heat.sig$log2FoldChange >0, na.rm=T) # genes upregulated in TP7 heat relative to TP7 amb
TP7_amb_v_TP7_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP7_amb_v_TP7_heat.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP7_amb_v_TP7_heat.rsig <- rlog(TP7_amb_v_TP7_heat.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
plotPCA(TP7_amb_v_TP7_heat.rsig, intgroup = c("Condition")) #Plot PCA of all samples for DEG only
mat <- assay(TP7_amb_v_TP7_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP7_amb_v_TP7_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row",
                         clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, 
                         show_colnames =TRUE) #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP7_amb_v_TP7_heat_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP7_amb_v_TP7_heat_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP7_amb_v_TP7_heat_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP7_amb_v_TP7_heat.sig <- as.data.frame(TP7_amb_v_TP7_heat.sig)
TP7_amb_v_TP7_heat.sig.list <- as.data.frame(counts(TP7_amb_v_TP7_heat.sig.list))
TP7_amb_v_TP7_heat.sig.list.full <- cbind(TP7_amb_v_TP7_heat.sig, TP7_amb_v_TP7_heat.sig.list)
write.csv(TP7_amb_v_TP7_heat.sig.list.full, file = "../output/Molecular/mRNA/TP7_amb_v_TP7_heat_DEG.csv")
```

TP5 Heat v TP7 Amb
```{r}
# Look at results 
TP5_heat_v_TP7_amb <- results(dds, contrast=c("Condition", "TP5_Heat" , "TP7_Ambient"))
head(TP5_heat_v_TP7_amb)

# Save results for all genes 
TP5_heat_v_TP7_amb_df <- as.data.frame(TP5_heat_v_TP7_amb)
#write.csv(TP5_heat_v_TP7_amb_df, file = "../output/Molecular/mRNA/TP5_heat_v_TP7_amb.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP5_heat_v_TP7_amb)

# Subset genes with significant padj values
sum(TP5_heat_v_TP7_amb$padj <0.05, na.rm=T) 
TP5_heat_v_TP7_amb.sig <- subset(TP5_heat_v_TP7_amb, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP5_heat_v_TP7_amb.sig$log2FoldChange <0, na.rm=T) # upregulated in TP7 amb
sum(TP5_heat_v_TP7_amb.sig$log2FoldChange >0, na.rm=T) # upregulated in TP5 heat
TP5_heat_v_TP7_amb.sig.list <- data[which(rownames(dds) %in% rownames(TP5_heat_v_TP7_amb.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP5_heat_v_TP7_amb.rsig <- rlog(TP5_heat_v_TP7_amb.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
plotPCA(TP5_heat_v_TP7_amb.rsig, intgroup = c("Condition")) #Plot PCA of all samples for DEG only
mat <- assay(TP5_heat_v_TP7_amb.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP5_heat_v_TP7_amb.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row",
                         clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, 
                         show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP5_heat_v_TP7_amb_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP5_heat_v_TP7_amb_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP5_heat_v_TP7_amb_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP5_heat_v_TP7_amb.sig <- as.data.frame(TP5_heat_v_TP7_amb.sig)
TP5_heat_v_TP7_amb.sig.list <- as.data.frame(counts(TP5_heat_v_TP7_amb.sig.list))
TP5_heat_v_TP7_amb.sig.list.full <- cbind(TP5_heat_v_TP7_amb.sig, TP5_heat_v_TP7_amb.sig.list)
#write.csv(TP5_heat_v_TP7_amb.sig.list.full, file = "../output/Molecular/mRNA/TP5_heat_v_TP7_amb_DEG.csv")
```

Look at unique genes 
```{r}
TP0_v_5_amb.DEGs <- as.data.frame(row.names(TP0_v_5_amb.sig.list))
colnames(TP0_v_5_amb.DEGs) <- "DEGs"
length(t(unique(TP0_v_5_amb.DEGs)))

TP0_v_7_amb.DEGs <- as.data.frame(row.names(TP0_v_7_amb.sig.list))
colnames(TP0_v_7_amb.DEGs) <- "DEGs"
length(t(unique(TP0_v_7_amb.DEGs)))

TP5_v_7_amb.DEGs <- as.data.frame(row.names(TP5_v_7_amb.sig.list))
colnames(TP5_v_7_amb.DEGs) <- "DEGs"
length(t(unique(TP5_v_7_amb.DEGs)))

TP0_v_5_heat.DEGs <- as.data.frame(row.names(TP0_v_5_heat.sig.list))
colnames(TP0_v_5_heat.DEGs) <- "DEGs"
length(t(unique(TP0_v_5_heat.DEGs)))

TP0_v_7_heat.DEGs <- as.data.frame(row.names(TP0_v_7_heat.sig.list))
colnames(TP0_v_7_heat.DEGs) <- "DEGs"
length(t(unique(TP0_v_7_heat.DEGs)))

TP5_heat_v_7_heat.DEGs <- as.data.frame(row.names(TP5_heat_v_7_heat.sig.list))
colnames(TP5_heat_v_7_heat.DEGs) <- "DEGs"
length(t(unique(TP5_heat_v_7_heat.DEGs)))

TP5_amb_v_TP5_heat.DEGs <- as.data.frame(row.names(TP5_amb_v_TP5_heat.sig.list))
colnames(TP5_amb_v_TP5_heat.DEGs) <- "DEGs"
length(t(unique(TP5_amb_v_TP5_heat.DEGs)))

TP7_amb_v_7_heat.DEGs <- as.data.frame(row.names(TP7_amb_v_TP7_heat.sig.list))
colnames(TP7_amb_v_7_heat.DEGs) <- "DEGs"
length(t(unique(TP7_amb_v_7_heat.DEGs)))

TP5_heat_v_7_amb.DEGs <- as.data.frame(row.names(TP5_heat_v_TP7_amb.sig.list))
colnames(TP5_heat_v_7_amb.DEGs) <- "DEGs"
length(t(unique(TP5_heat_v_7_amb.DEGs)))
```

Bind genes from all contrasts
```{r}
DEGs.all <- rbind(TP0_v_5_amb.DEGs, TP0_v_7_amb.DEGs, TP5_v_7_amb.DEGs, TP0_v_5_heat.DEGs, 
                TP0_v_7_heat.DEGs, TP5_heat_v_7_heat.DEGs, TP5_amb_v_TP5_heat.DEGs, 
                TP7_amb_v_7_heat.DEGs, TP5_heat_v_7_amb.DEGs)
DEGs.all <- unique(DEGs.all)
length(unique(DEGs.all$DEGs))

Unique.sig.list <- data[which(rownames(data) %in% DEGs.all$DEGs),] #subset list of sig transcripts from original count data
Unique.rsig <- rlog(Unique.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
#write.csv(counts(Unique.sig.list), file="../output/Molecular/mRNA/Unique_DEGs.csv")
```

Visualize DEGs. 

PCA with DEGs
```{r}
pca_plot <- plotPCA(Unique.rsig, intgroup = c("Condition")); pca_plot
# percentVar_pca <- round(100*attr(pca_plot, "percentVar")) #plot PCA of samples with all data
# 
# pca_plot_degs <- ggplot(pca_plot, aes(PC1, PC2, color = Condition)) + 
#   geom_point(size=10) +
#   xlab(paste0("PC1: ",percentVar_pca[1],"% variance")) +
#   ylab(paste0("PC2: ",percentVar_pca[2],"% variance")) +
#   #scale_color_manual(values = wes_palette("GrandBudapest1", n=3)) +
#   #ggtitle(label = "M. capitata") +
#   theme_bw() + #Set background color
#   theme(legend.text = element_text(size=13), 
#         #legend.position="none",
#         plot.background = element_blank(),
#         legend.title = element_text(size=18, face="bold"), 
#         #legend.title=element_blank(),
#         axis.text = element_text(size=13), 
#         axis.title = element_text(size=15,  face="bold"), 
#         axis.title.y = element_text(vjust=-1.5),
#         plot.title = element_text(size = 20, face = "italic", hjust = 0.5)) +
#   coord_equal(ratio = 2); pca_plot_degs
##^^ code above wasn't working 

# Customize the PCA plot
pca_plot2 <- pca_plot + 
  #labs(x = paste0("PC1: ", percentVar_pca[1], "% variance"),
  #     y = paste0("PC2: ", percentVar_pca[2], "% variance")) +
  theme_bw() +
  theme(legend.text = element_text(size = 13),
        plot.background = element_blank(),
        legend.title = element_text(size = 18, face = "bold"),
        axis.text = element_text(size = 13),
        axis.title = element_text(size = 15, face = "bold"),
        axis.title.y = element_text(vjust = -1.5),
        plot.title = element_text(size = 20, face = "italic", hjust = 0.5)); pca_plot2

```

Need to make the plot prettier and figure out how to incorporate Treatment and timepoint 

Heatmap with DEGs
```{r}
mat <- assay(Unique.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(Unique.rsig)[,c("Treatment", "Timepoint")]) #make dataframe
df <- df[order(df$Treatment),]
#df$Timepoint <- factor(df$Timepoint, levels = c("TP0", "TP5", "TP7"), labels = c("February", "June", "August"))
treatment_colors <- c("Ambient" = "#1f78b4", "Heat" = "#e31a1c")  # blue vs red
timepoint_colors <- c("TP0" = "#8a36b6", "TP5" = "#f19e21", "TP7" = "#5eca4a")
ann_colors <- list(Treatment = treatment_colors, Timepoint = timepoint_colors)
heatmap <- pheatmap(
  mat,
  annotation_col = df,
  annotation_colors = ann_colors,
  clustering_method = "average",
  scale = "row",
  clustering_distance_rows = "euclidean",
  show_rownames = F,
  fontsize_row = 2,
  fontsize = 15,
  cluster_cols = FALSE,
  cluster_rows = T,
  show_colnames = F
); heatmap
```

Also need to make this prettier 

Save DEG plots 
```{r}
ggsave("../output/Molecular/mRNA/pca_plot_DEGs.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_DEGs.png", pca_plot, width = 10, height = 8)

pdf("../output/Molecular/mRNA/heatmap_DEGs.pdf", width = 10, height = 11)
print(heatmap)
dev.off()
```

Save unique DEGs in a list for subsetting fasta files
```{r}
DEGs.list <- read.csv("../output/Molecular/mRNA/Unique_DEGs.csv")
length(unique(DEGs.list$X))

list <- as.character(DEGs.list$X)

writeLines(list, "../output/Molecular/mRNA/DEG_list.txt")
```
