---
title: "DESeq2 June and August treatments"
author: "Jill Ashey"
date: "2025-05-01"
output: html_document
---

DESeq2 analysis for June and August ambient and heat samples (ie February is removed).

Load libraries 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(car)

library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(car)
library(gplots)
library(wesanderson)
library(cowplot)
library(gridExtra)
```

Read in filtered data 
```{r}
filt_counts <- read.csv("../output/Molecular/mRNA/filtered_gene_counts.csv")
rownames(filt_counts) <- filt_counts[,1] #set first column that contains gene names as rownames
filt_counts <- filt_counts[,-1] # remove column w/ gene names 
colnames(filt_counts) <- sub("AST\\.", "AST-", colnames(filt_counts))

# Remove AST-1105, as it did not sequence well 
filt_counts <- filt_counts[, !colnames(filt_counts) %in% "AST-1105"]
```

Read in metadata
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv")
meta <- dplyr::arrange(meta, ID) %>% # rearrange metadata so IDs are sorted in descending order 
  mutate(Treatment = ifelse(Treatment == "Acclimation", "Ambient", Treatment))
  
# Set variables as factors 
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP0", "TP5", "TP7"))
meta$Treatment <- factor(meta$Treatment, levels = c("Ambient", "Heat"))
```

Remove TP0 samples from metadata and AST-1105
```{r}
meta <- meta %>%
  filter(!Timepoint == "TP0") %>%
  filter(!ID == "AST-1105")
```

Remove TP0 samples from counts df based on metadata
```{r}
filt_counts <- filt_counts[, colnames(filt_counts) %in% meta$ID]
```

Set variables as factors 
```{r}
meta$Treatment <- factor(meta$Treatment, levels = c("Ambient", "Heat"))
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP5", "TP7"))
```

Data sanity check!
```{r}
meta$ID %in% colnames(filt_counts) #are all of the sample names (rows) in the metadata df in the gene count matrix? Should be TRUE. 
all(rownames(meta$ID) == colnames(filt_counts)) #are they the same in the same order? Should be TRUE
```

No data filtering needed, as this has already been done. 

Merge Treatment and Timepoint into a single column 
```{r}
meta$Condition <- paste(meta$Timepoint, meta$Treatment, sep = "_")
meta$Condition <- factor(meta$Condition, levels = c("TP5_Ambient", "TP5_Heat", "TP7_Ambient", "TP7_Heat"))
```

Put data and metadata in DESeq2 matrix 
```{r}
data <- DESeqDataSetFromMatrix(countData = filt_counts, colData = meta, design = ~Condition)
# data <- DESeqDataSetFromMatrix(countData = filt_counts, colData = meta, design = ~Treatment*Timepoint)
```

Visualize expression
```{r}
rld <- rlog(data, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
head(assay(rld), 3) #view data
sampleDists <- dist(t(assay(rld))) #calculate distance matix
sampleDistMatrix <- as.matrix(sampleDists) #distance matrix
rownames(sampleDistMatrix) <- colnames(rld) #assign row names
colnames(sampleDistMatrix) <- NULL #assign col names
pheatmap(sampleDistMatrix, #plot matrix of expression similarity
         clustering_distance_rows=sampleDists, #cluster rows
         clustering_distance_cols=sampleDists) #cluster columns

plotPCA(rld, intgroup = c("Treatment", "Timepoint")) #plot PCA of samples with all data
```

Run differential gene expression analysis 
```{r}
DEG.int <- DESeq(data) #run differential expression test by group using the wald test 
resultsNames(DEG.int)
```
Look at contrats 

TP5 Amb v TP5 Heat
```{r}
# Look at results 
TP5_amb_v_TP5_heat <- results(DEG.int, contrast=c("Condition", "TP5_Ambient" , "TP5_Heat"))
head(TP5_amb_v_TP5_heat)

# Save results for all genes 
TP5_amb_v_TP5_heat_df <- as.data.frame(TP5_amb_v_TP5_heat)
write.csv(TP5_amb_v_TP5_heat_df, file = "../output/Molecular/mRNA/TP5_amb_v_TP5_heat.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP5_amb_v_TP5_heat)

# Subset genes with significant padj values
sum(TP5_amb_v_TP5_heat$padj <0.05, na.rm=T) # genes with significant p-values
TP5_amb_v_TP5_heat.sig <- subset(TP5_amb_v_TP5_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP5_amb_v_TP5_heat.sig$log2FoldChange <0, na.rm=T) # genes downregulated in TP5 heat relative to TP5 amb OR upregulated in TP5 amb
sum(TP5_amb_v_TP5_heat.sig$log2FoldChange >0, na.rm=T) # genes upregulated in TP5 heat relative to TP5 amb
TP5_amb_v_TP5_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP5_amb_v_TP5_heat.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP5_amb_v_TP5_heat.rsig <- rlog(TP5_amb_v_TP5_heat.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
pca_plot <- plotPCA(TP5_amb_v_TP5_heat.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP5_amb_v_TP5_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP5_amb_v_TP5_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP5_amb_v_TP5_heat_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP5_amb_v_TP5_heat_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP5_amb_v_TP5_heat_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP5_amb_v_TP5_heat.sig <- as.data.frame(TP5_amb_v_TP5_heat.sig)
TP5_amb_v_TP5_heat.sig.list <- as.data.frame(counts(TP5_amb_v_TP5_heat.sig.list))
TP5_amb_v_TP5_heat.sig.list.full <- cbind(TP5_amb_v_TP5_heat.sig, TP5_amb_v_TP5_heat.sig.list)
write.csv(TP5_amb_v_TP5_heat.sig.list.full, file = "../output/Molecular/mRNA/TP5_amb_v_TP5_heat_DEG.csv")
```

TP7 Amb v TP7 Heat
```{r}
# Look at results 
TP7_amb_v_TP7_heat <- results(DEG.int, contrast=c("Condition", "TP7_Ambient" , "TP7_Heat"))
head(TP7_amb_v_TP7_heat)

# Save results for all genes 
TP7_amb_v_TP7_heat_df <- as.data.frame(TP7_amb_v_TP7_heat)
write.csv(TP7_amb_v_TP7_heat_df, file = "../output/Molecular/mRNA/TP7_amb_v_TP7_heat.csv")

# Volcano plot for LFC
plotMA(TP7_amb_v_TP7_heat)

# Subset genes with significant padj values
sum(TP7_amb_v_TP7_heat$padj <0.05, na.rm=T) # genes with significant p-values
TP7_amb_v_TP7_heat.sig <- subset(TP7_amb_v_TP7_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP7_amb_v_TP7_heat.sig$log2FoldChange <0, na.rm=T) # genes downregulated in TP7 heat relative to TP7 amb OR upregulated in TP7 amb 
sum(TP7_amb_v_TP7_heat.sig$log2FoldChange >0, na.rm=T) # genes upregulated in TP7 heat relative to TP7 amb
TP7_amb_v_TP7_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP7_amb_v_TP7_heat.sig)),] #subset list of sig transcripts from original count data

# Apply a regularized log transformation to minimize effects of small counts
TP7_amb_v_TP7_heat.rsig <- rlog(TP7_amb_v_TP7_heat.sig.list, blind=FALSE) 

# Plot PCA and heatmap of DEGs for this comparison
plotPCA(TP7_amb_v_TP7_heat.rsig, intgroup = c("Condition")) #Plot PCA of all samples for DEG only
mat <- assay(TP7_amb_v_TP7_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP7_amb_v_TP7_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row",
                         clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, 
                         show_colnames =TRUE) #plot heatmap of all DEG by group

## Save plots 
# ggsave("../output/Molecular/mRNA/ma_plot_TP7_amb_v_TP7_heat_DEG.pdf", ma_plot, width = 10, height = 8)
# ggsave("../output/Molecular/mRNA/pca_plot_TP7_amb_v_TP7_heat_DEG.pdf", pca_plot, width = 10, height = 8)
# pdf("../output/Molecular/mRNA/heatmap_TP7_amb_v_TP7_heat_DEG.pdf", width = 8, height = 10)
# print(heatmap)
# dev.off()

# Save info with pvalues and LFC
TP7_amb_v_TP7_heat.sig <- as.data.frame(TP7_amb_v_TP7_heat.sig)
TP7_amb_v_TP7_heat.sig.list <- as.data.frame(counts(TP7_amb_v_TP7_heat.sig.list))
TP7_amb_v_TP7_heat.sig.list.full <- cbind(TP7_amb_v_TP7_heat.sig, TP7_amb_v_TP7_heat.sig.list)
write.csv(TP7_amb_v_TP7_heat.sig.list.full, file = "../output/Molecular/mRNA/TP7_amb_v_TP7_heat_DEG.csv")
```

Look at unique genes 
```{r}
TP5_amb_v_TP5_heat.DEGs <- as.data.frame(row.names(TP5_amb_v_TP5_heat.sig.list))
colnames(TP5_amb_v_TP5_heat.DEGs) <- "DEGs"
length(t(unique(TP5_amb_v_TP5_heat.DEGs)))

TP7_amb_v_7_heat.DEGs <- as.data.frame(row.names(TP7_amb_v_TP7_heat.sig.list))
colnames(TP7_amb_v_7_heat.DEGs) <- "DEGs"
length(t(unique(TP7_amb_v_7_heat.DEGs)))
```

Bind genes from all contrasts
```{r}
DEGs.all <- rbind(TP5_amb_v_TP5_heat.DEGs, TP7_amb_v_7_heat.DEGs)
DEGs.all <- unique(DEGs.all)
length(unique(DEGs.all$DEGs))

Unique.sig.list <- data[which(rownames(data) %in% DEGs.all$DEGs),] #subset list of sig transcripts from original count data
Unique.rsig <- rlog(Unique.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
write.csv(counts(Unique.sig.list), file="../output/Molecular/mRNA/Unique_DEGs_Treatment.csv")
```

Visualize DEGs. 

PCA with DEGs
```{r}
pca_plot <- plotPCA(Unique.rsig, intgroup = c("Condition")); pca_plot
# percentVar_pca <- round(100*attr(pca_plot, "percentVar")) #plot PCA of samples with all data
# 
# pca_plot_degs <- ggplot(pca_plot, aes(PC1, PC2, color = Condition)) + 
#   geom_point(size=10) +
#   xlab(paste0("PC1: ",percentVar_pca[1],"% variance")) +
#   ylab(paste0("PC2: ",percentVar_pca[2],"% variance")) +
#   #scale_color_manual(values = wes_palette("GrandBudapest1", n=3)) +
#   #ggtitle(label = "M. capitata") +
#   theme_bw() + #Set background color
#   theme(legend.text = element_text(size=13), 
#         #legend.position="none",
#         plot.background = element_blank(),
#         legend.title = element_text(size=18, face="bold"), 
#         #legend.title=element_blank(),
#         axis.text = element_text(size=13), 
#         axis.title = element_text(size=15,  face="bold"), 
#         axis.title.y = element_text(vjust=-1.5),
#         plot.title = element_text(size = 20, face = "italic", hjust = 0.5)) +
#   coord_equal(ratio = 2); pca_plot_degs
##^^ code above wasn't working 

# Customize the PCA plot
pca_plot2 <- pca_plot + 
  #labs(x = paste0("PC1: ", percentVar_pca[1], "% variance"),
  #     y = paste0("PC2: ", percentVar_pca[2], "% variance")) +
  theme_bw() +
  theme(legend.text = element_text(size = 13),
        plot.background = element_blank(),
        legend.title = element_text(size = 18, face = "bold"),
        axis.text = element_text(size = 13),
        axis.title = element_text(size = 15, face = "bold"),
        axis.title.y = element_text(vjust = -1.5),
        plot.title = element_text(size = 20, face = "italic", hjust = 0.5)); pca_plot2

```



