---
title: "DESeq2 compare"
author: "Jill Ashey"
date: "`r Sys.Date()`"
output: html_document
---

I used bowtie2 and hisat2 as aligners for my data. Now I want to compare the gene count matrices that each aligner produced. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(car)

library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(car)
library(gplots)
library(wesanderson)
library(cowplot)
library(gridExtra)
```

Read in data
```{r}
bowtie_matrix <- read_csv("data/Molecular/mRNA/Apoc_gene_count_matrix_bowtie.csv")
hisat2_matrix <- read_csv("data/Molecular/mRNA/Apoc_gene_count_matrix_hisat2.csv")

head(bowtie_matrix)
head(hisat2_matrix)
```

Invesigate data. First, compare row sums for each alignment method.
```{r}
row_sums <- rowSums(select(bowtie_matrix, -gene_id))
bowtie_matrix$sum_across_rows_bowtie <- row_sums
last_column_bowtie <- bowtie_matrix[, ncol(bowtie_matrix)]

row_sums <- rowSums(select(hisat2_matrix, -gene_id))
hisat2_matrix$sum_across_rows_hisat2 <- row_sums
last_column_hisat2 <- hisat2_matrix[, ncol(hisat2_matrix)]

merge <- merge(bowtie_matrix, hisat2_matrix, by = "gene_id")

summary <- merge %>%
  select(gene_id, sum_across_rows_bowtie, sum_across_rows_hisat2) %>%
  mutate(difference = sum_across_rows_bowtie - sum_across_rows_hisat2) %>%
  summarize(
    positive_count = sum(difference > 0),
    negative_count = sum(difference < 0),
    zero_count = sum(difference == 0))

print(summary)
```

For ~37000 genes, bowtie2 generated more counts than hisat2. 

Compare column sums for each alignment method.
```{r}
col_sums_bowtie <- colSums(select(bowtie_matrix, -gene_id))

col_sums_hisat2 <- colSums(select(hisat2_matrix, -gene_id))

col_sums <- rbind(col_sums_bowtie, col_sums_hisat2)
col_sums <- as.data.frame(col_sums)
```

For all samples, there are more reads aligned with bowtie than with hisat2. 




Let's try to run some DESeq2 with the bowtie data. Read in counts. 
```{r}
bowtie_matrix <- read_csv("../data/Molecular/mRNA/Apoc_gene_count_matrix_bowtie.csv")
bowtie_matrix <- as.data.frame(bowtie_matrix)
rownames(bowtie_matrix) <- bowtie_matrix[,1] #set first column that contains gene names as rownames
bowtie_matrix <- bowtie_matrix[,-1] # remove column w/ gene names 
```

Remove extraneous info from sample names. 
```{r}
colnames(bowtie_matrix) <- sub("align.trimmed.", "", colnames(bowtie_matrix))
colnames(bowtie_matrix) <- sub("_R1_001.fastq.gz.bam.gtf", "", colnames(bowtie_matrix))
```

Read in metadata 
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv")
meta <- dplyr::arrange(meta, ID) %>% # rearrange metadata so IDs are sorted in descending order 
  mutate(Treatment = ifelse(Treatment == "Acclimation", "Ambient", Treatment))
  
# Set variables as factors 
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP0", "TP5", "TP7"))
meta$Treatment <- factor(meta$Treatment, levels = c("Ambient", "Heat"))
```

Data sanity checks!
```{r}
meta$ID %in% colnames(bowtie_matrix) #are all of the sample names (rows) in the metadata df in the gene count matrix? Should be TRUE. 
all(rownames(meta$ID) == colnames(bowtie_matrix)) #are they the same in the same order? Should be TRUE
```

Filter reads by proportion of samples containing cutoff value 
```{r}
ffun<-filterfun(pOverA(0.75,5))  #set up filtering parameters--LOOK INTO FILTERING PARAMETERS
filt_outrm_poa <- genefilter((bowtie_matrix), ffun) #apply filter
sum(filt_outrm_poa) #count number of genes left

filt_outrm_poa <- bowtie_matrix[filt_outrm_poa,] #keep only rows that passed filter

all(rownames(meta$ID) %in% colnames(filt_outrm_poa)) # must come out TRUE
```

Remove AST-1105 from the data, as it did not have good mapping and has basically no counts. 
```{r}
# Remove column from gene count matrix 
filt_outrm_poa <- filt_outrm_poa %>%
  select(-"AST-1105")
  
# Remove row from metadata
meta <- meta %>% 
  filter(ID != "AST-1105")

all(rownames(meta$ID) %in% colnames(filt_outrm_poa)) # must come out TRUE
```

```{r}
data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = ~Timepoint+Treatment)

dds <- DESeq(data) #apply DEseq dispersion calculation
resultsNames(dds)
# [1] "Intercept"            "Timepoint_TP5_vs_TP0" "Timepoint_TP7_vs_TP0"
blah <- results(dds, contrast=c("Timepoint", "TP5", "TP7"))
results(dds, contrast=c("Timepoint", "TP0", "TP5"))
results(dds, contrast=c("Timepoint", "TP0", "TP7"))




## merge TP and trt columns into a 'Condition' column 
meta$Condition <- paste(meta$Timepoint, meta$Treatment, sep = "_")
meta$Condition <- factor(meta$Condition, levels = c("TP0_Ambient", "TP5_Ambient", "TP5_Heat", "TP7_Ambient", "TP7_Heat"))
data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = ~Condition)
dds <- DESeq(data)
resultsNames(dds)
blah <- results(dds, contrast=c("Condition", "TP7_Ambient" , "TP0_Ambient"))

```

I can run it without the interaction term. but if I include the interaction term, I need to remove TP0, as this makes the model not full rank because TP0 only has one trt (ambient). 

```{r}
data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = ~Timepoint)
dds <- DESeq(data) #apply DEseq dispersion calculation
resultsNames(dds)
# [1] "Intercept"            "Timepoint_TP5_vs_TP0" "Timepoint_TP7_vs_TP0"
blah <- results(dds, contrast=c("Timepoint", "TP5", "TP7"))
results(dds, contrast=c("Timepoint", "TP0", "TP5"))
results(dds, contrast=c("Timepoint", "TP0", "TP7"))

data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = ~Treatment)
dds <- DESeq(data) #apply DEseq dispersion calculation
resultsNames(dds)
# [1] "Intercept"                        "Treatment_Ambient_vs_Acclimation" "Treatment_Heat_vs_Acclimation"   
```

Why is it not giving me all of the comparisons?

Let's look at timepoint first 
```{r}
data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = ~Timepoint)
dds <- DESeq(data) #apply DEseq dispersion calculation
resultsNames(dds)
# [1] "Intercept"            "Timepoint_TP5_vs_TP0" "Timepoint_TP7_vs_TP0" -- not giving me all the comparisons

## TP7 vs 5
TP7_v_5 <- results(dds, contrast=c("Timepoint", "TP7", "TP5"))
head(TP7_v_5)
plotMA(TP7_v_5, ylim=c(-7,6))
sum(TP7_v_5$padj <0.05, na.rm=T)
TP7_v_5.sig <- subset(TP7_v_5, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP7_v_5.sig$log2FoldChange <0, na.rm=T)
sum(TP7_v_5.sig$log2FoldChange >0, na.rm=T)
TP7v5.sig.list <- data[which(rownames(dds) %in% rownames(TP7_v_5.sig)),] #subset list of sig transcripts from original count data
#write.csv(counts(SpawnCon.sig.list), file="Output/Spawn_compare_DEG.csv")
TP7v5.rsig <- rlog(TP7v5.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
plotPCA(TP7v5.rsig, intgroup = c("Treatment", "Timepoint")) #Plot PCA of all samples for DEG only
mat <- assay(TP7v5.rsig) #make an expression object
col.order <- c("AST-1147", "AST-1567", "AST-1722", "AST-2398", # Acclimation
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP7v5.rsig)[,c("Treatment", "Timepoint")]) #make dataframe
df <- df[order(df$Timepoint),]

pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row",
                         clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, 
                         show_colnames =TRUE) #plot heatmap of all DEG by group


## TP5 vs 0
TP5_v_0 <- results(dds, contrast=c("Timepoint", "TP5", "TP0"))
head(TP5_v_0)
plotMA(TP5_v_0, ylim=c(-7,6))
sum(TP5_v_0$padj <0.05, na.rm=T)
TP5_v_0.sig <- subset(TP5_v_0, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP5_v_0.sig$log2FoldChange <0, na.rm=T)
sum(TP5_v_0.sig$log2FoldChange >0, na.rm=T)
TP5v0.sig.list <- data[which(rownames(dds) %in% rownames(TP5_v_0.sig)),] #subset list of sig transcripts from original count data
#write.csv(counts(SpawnCon.sig.list), file="Output/Spawn_compare_DEG.csv")
TP5v0.rsig <- rlog(TP5v0.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
plotPCA(TP5v0.rsig, intgroup = c("Treatment", "Timepoint")) #Plot PCA of all samples for DEG only
mat <- assay(TP5v0.rsig) #make an expression object
col.order <- c("AST-1147", "AST-1567", "AST-1722", "AST-2398", # Acclimation
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP5v0.rsig)[,c("Treatment", "Timepoint")]) #make dataframe
df <- df[order(df$Timepoint),]

pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row",
                         clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, 
                         show_colnames =TRUE) #plot heatmap of all DEG by group


## TP7 vs 0
TP7_v_0 <- results(dds, contrast=c("Timepoint", "TP7", "TP0"))
head(TP7_v_0)
plotMA(TP7_v_0, ylim=c(-7,6))
sum(TP7_v_0$padj <0.05, na.rm=T)
TP7_v_0.sig <- subset(TP7_v_0, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP7_v_0.sig$log2FoldChange <0, na.rm=T)
sum(TP7_v_0.sig$log2FoldChange >0, na.rm=T)
```

Let's look at timepoint first 
```{r}
data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = ~Treatment)
dds <- DESeq(data) #apply DEseq dispersion calculation
resultsNames(dds)
# [1] "Intercept"                        "Treatment_Ambient_vs_Acclimation" "Treatment_Heat_vs_Acclimation"   

## Acclimation vs Ambient
Acc_v_Amb <- results(dds, contrast=c("Treatment", "Ambient", "Acclimation"))
head(Acc_v_Amb)
plotMA(Acc_v_Amb, ylim=c(-7,6))
sum(Acc_v_Amb$padj <0.05, na.rm=T)
Acc_v_Amb.sig <- subset(Acc_v_Amb, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(Acc_v_Amb.sig$log2FoldChange <0, na.rm=T)
sum(Acc_v_Amb.sig$log2FoldChange >0, na.rm=T)

## Acclimation vs Heat
Acc_v_Heat <- results(dds, contrast=c("Treatment", "Heat", "Acclimation"))
head(Acc_v_Heat)
plotMA(Acc_v_Heat, ylim=c(-7,6))
sum(Acc_v_Heat$padj <0.05, na.rm=T)
Acc_v_Heat.sig <- subset(Acc_v_Heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(Acc_v_Heat.sig$log2FoldChange <0, na.rm=T)
sum(Acc_v_Heat.sig$log2FoldChange >0, na.rm=T)

## Ambient vs Heat
Amb_v_Heat <- results(dds, contrast=c("Treatment", "Heat", "Ambient"))
head(Amb_v_Heat)
plotMA(Amb_v_Heat, ylim=c(-7,6))
sum(Amb_v_Heat$padj <0.05, na.rm=T)
Amb_v_Heat.sig <- subset(Amb_v_Heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(Amb_v_Heat.sig$log2FoldChange <0, na.rm=T)
sum(Amb_v_Heat.sig$log2FoldChange >0, na.rm=T)
```

Now let's try combining Timepoint and Treatment in the design formula 
```{r}
# Assuming you want to include the main effects and interaction term for Timepoint and Treatment
design_formula <- ~ Timepoint+Treatment+Timepoint:Treatment

data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = design_formula)

data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = ~Timepoint*Treatment)
dds <- DESeq(data) #apply DEseq dispersion calculation
resultsNames(dds)
```

Still not working. Saying that my model matrix is not full rank. Let's troubleshoot. 

Check for perfect collinearity. 
```{r}
# Assuming 'meta' is your metadata dataframe
design_matrix <- model.matrix(~ Timepoint + Treatment + Timepoint:Treatment, data = meta)

# Check for collinearity using VIF
vif_results <- vif(design_matrix)
print(vif_results)

## this is giving me errors. 
```

Check factor levels
```{r}
## Check unique levels 
unique_levels_timepoint <- unique(meta$Timepoint)
unique_levels_treatment <- unique(meta$Treatment)

print(unique_levels_timepoint)
print(unique_levels_treatment)

## Check freq of levels
table_timepoint <- table(meta$Timepoint)
table_treatment <- table(meta$Treatment)

print(table_timepoint)
print(table_treatment)

barplot(table_timepoint, main = "Distribution of Timepoint Levels")
barplot(table_treatment, main = "Distribution of Treatment Levels")
```

I bet I know what's wrong...acclimation and TP0 are the same. What if I removed the TP0 samples? The TP0 samples are AST-1147, AST-1567, AST-1722, and AST-2398. 
```{r}
# Remove rows from metadata
meta_subset <- meta %>% 
  filter(Treatment != "Acclimation")

# Remove column from gene count matrix 
filt_outrm_poa_subset <- filt_outrm_poa %>%
  select(-c("AST-1147", "AST-1567", "AST-1722", "AST-2398"))

all(rownames(meta_subset$ID) %in% colnames(filt_outrm_poa_subset)) # must come out TRUE
```

Now run DESeq2
```{r}
data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa_subset, colData = meta_subset, design = ~Timepoint + Treatment + Timepoint:Treatment)
dds <- DESeq(data) #apply DEseq dispersion calculation
resultsNames(dds)
# [1] "Intercept"                  "Timepoint_TP7_vs_TP5"       "Treatment_Heat_vs_Ambient"  "TimepointTP7.TreatmentHeat" -- still not giving all of the comparisons...

## TP7 vs 5
TP7v5 <- results(dds, contrast=c("Timepoint", "TP7", "TP5"))
head(TP7v5)
plotMA(TP7v5)
sum(TP7v5$padj <0.05, na.rm=T) 
TP7v5.sig <- subset(TP7v5, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP7v5.sig$log2FoldChange <0, na.rm=T)
sum(TP7v5.sig$log2FoldChange >0, na.rm=T)
TP7v5.sig.list <- data[which(rownames(dds) %in% rownames(TP7v5.sig)),] #subset list of sig transcripts from original count data
#write.csv
TP7v5.rsig <- rlog(TP7v5.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
plotPCA(TP7v5.rsig, intgroup = c("Treatment", "Timepoint")) #Plot PCA of all samples for DEG only
mat <- assay(TP7v5.rsig) #make an expression object
col.order <- c( "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP7v5.rsig)[,c("Treatment", "Timepoint")]) #make dataframe
df <- df[order(df$Timepoint),]

pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row",
                         clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, 
                         show_colnames =TRUE) #plot heatmap of all DEG by group




## Heat vs Ambient
```




What if I mutated the metadata file so that Acclimation was Ambient? Or would that still run into the same problem?

I feel like I made progress today! I am still very confused about how to interpret my data and why the comparisons are not all showing up. I need to sit down with Hollie, Zoe or Danielle and talk through this with them. My main questions

- Why are the results different when I set the design to ~Treatment or ~Timepoint vs ~Timepoint + Treatment + Timepoint:Treatment?







20240117

Talked with Zoe about the various issues I was running into above. We agreed that there is an issue with TP0 only have one 'treatment'. We discussed a couple of different options moving forward:

- 1) Merge the Treatment and Timepoint columns together into one column called 'Condition'. Then run DESeq2 so the design is ~Condition
- 2) Remove TP0 and only analyze TP5 Amb v Heat and TP7 Amb v Heat. Then run DESeq2 so the design is ~Timepoint + Treatment + Timepoint:Treatment OR Timepoint * Treatment 
- 3) Remove TP5 and TP7 Amb samples and only keep TP0, TP5 Heat and TP7 Heat. Then run DESeq2 so the design is ~Timepoint + Treatment + Timepoint:Treatment OR Timepoint * Treatment 
- 4) Remove Heat samples from TP5 and TP7 and only analyze the Amb samples for TP0, TP5, and TP7. The DESeq2 design would be ~Timepoint 

I'm going to do all of these iterations and see where that leaves me. Also need to discuss w/ Hollie what the best way forward might be. 

## Prep data for all iterations 
Let's try to run some DESeq2 with the bowtie data. Read in counts. 
```{r}
bowtie_matrix <- read_csv("../data/Molecular/mRNA/Apoc_gene_count_matrix_bowtie.csv")
bowtie_matrix <- as.data.frame(bowtie_matrix)
rownames(bowtie_matrix) <- bowtie_matrix[,1] #set first column that contains gene names as rownames
bowtie_matrix <- bowtie_matrix[,-1] # remove column w/ gene names 
```

Remove extraneous info from sample names. 
```{r}
colnames(bowtie_matrix) <- sub("align.trimmed.", "", colnames(bowtie_matrix))
colnames(bowtie_matrix) <- sub("_R1_001.fastq.gz.bam.gtf", "", colnames(bowtie_matrix))
```

Read in metadata 
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv")
meta <- dplyr::arrange(meta, ID) %>% # rearrange metadata so IDs are sorted in descending order 
  mutate(Treatment = ifelse(Treatment == "Acclimation", "Ambient", Treatment))
  
# Set variables as factors 
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP0", "TP5", "TP7"))
meta$Treatment <- factor(meta$Treatment, levels = c("Ambient", "Heat"))
```

Data sanity checks!
```{r}
meta$ID %in% colnames(bowtie_matrix) #are all of the sample names (rows) in the metadata df in the gene count matrix? Should be TRUE. 
all(rownames(meta$ID) == colnames(bowtie_matrix)) #are they the same in the same order? Should be TRUE
```

Filter reads by proportion of samples containing cutoff value 
```{r}
ffun<-filterfun(pOverA(0.75,5))  #set up filtering parameters--LOOK INTO FILTERING PARAMETERS
filt_outrm_poa <- genefilter((bowtie_matrix), ffun) #apply filter
sum(filt_outrm_poa) #count number of genes left

filt_outrm_poa <- bowtie_matrix[filt_outrm_poa,] #keep only rows that passed filter

all(rownames(meta$ID) %in% colnames(filt_outrm_poa)) # must come out TRUE
```

Remove AST-1105 from the data, as it did not have good mapping and has basically no counts. 
```{r}
# Remove column from gene count matrix 
filt_outrm_poa <- filt_outrm_poa %>%
  select(-"AST-1105")
  
# Remove row from metadata
meta <- meta %>% 
  filter(ID != "AST-1105")

all(rownames(meta$ID) %in% colnames(filt_outrm_poa)) # must come out TRUE
```

## 1) Merge the Treatment and Timepoint columns together into one column called 'Condition'. Then run DESeq2 so the design is ~Condition

Merge Treatment and Timepoint into a single column 
```{r}
meta$Condition <- paste(meta$Timepoint, meta$Treatment, sep = "_")
meta$Condition <- factor(meta$Condition, levels = c("TP0_Ambient", "TP5_Ambient", "TP5_Heat", "TP7_Ambient", "TP7_Heat"))
```

Run DESeq2
```{r}
data <- DESeqDataSetFromMatrix(countData = filt_outrm_poa, colData = meta, design = ~Condition)
dds <- DESeq(data)
resultsNames(dds)
```

Look at contrasts

TP0 v TP5 Ambient
```{r}
TP0_v_5_amb <- results(dds, contrast=c("Condition", "TP0_Ambient" , "TP5_Ambient"))
head(TP0_v_5_amb)
ma_plot <- plotMA(TP0_v_5_amb); ma_plot
sum(TP0_v_5_amb$padj <0.05, na.rm=T) # 541 genes with significant p-values
TP0_v_5_amb.sig <- subset(TP0_v_5_amb, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_5_amb.sig$log2FoldChange <0, na.rm=T) # 432 genes are downregulated in TP5 ambient relative to TP0 ambient
sum(TP0_v_5_amb.sig$log2FoldChange >0, na.rm=T) # 109 genes are upregulated in TP5 ambient relative to TP0 ambient
TP0_v_5_amb.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_5_amb.sig)),] #subset list of sig transcripts from original count data
write.csv(counts(TP0_v_5_amb.sig.list), file = "../output/Molecular/mRNA/TP0_amb_v_TP5_amb_DEG.csv")
TP0_v_5_amb.rsig <- rlog(TP0_v_5_amb.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
pca_plot <- plotPCA(TP0_v_5_amb.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP0_v_5_amb.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP0_v_5_amb.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/Molecular/mRNA/ma_plot_TP0_amb_v_TP5_amb_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_TP0_amb_v_TP5_amb_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/heatmap_TP0_amb_v_TP5_amb_DEG.pdf", heatmap, width = 8, height = 10)
```

TP0 v TP7 Ambient
```{r}
TP0_v_7_amb <- results(dds, contrast=c("Condition", "TP0_Ambient" , "TP7_Ambient"))
head(TP0_v_7_amb)
ma_plot <- plotMA(TP0_v_7_amb); ma_plot
sum(TP0_v_7_amb$padj <0.05, na.rm=T) # 599 genes with significant p-values
TP0_v_7_amb.sig <- subset(TP0_v_7_amb, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_7_amb.sig$log2FoldChange <0, na.rm=T) # 424 genes are downregulated in TP5 ambient relative to TP0 ambient
sum(TP0_v_7_amb.sig$log2FoldChange >0, na.rm=T) # 175 genes are upregulated in TP5 ambient relative to TP0 ambient
TP0_v_7_amb.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_7_amb.sig)),] #subset list of sig transcripts from original count data
write.csv(counts(TP0_v_7_amb.sig.list), file = "../output/Molecular/mRNA/TP0_amb_v_TP7_amb_DEG.csv")
TP0_v_7_amb.rsig <- rlog(TP0_v_7_amb.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
pca_plot <- plotPCA(TP0_v_7_amb.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP0_v_7_amb.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP0_v_7_amb.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/Molecular/mRNA/ma_plot_TP0_amb_v_TP7_amb_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_TP0_amb_v_TP7_amb_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/heatmap_TP0_amb_v_TP7_amb_DEG.pdf", heatmap, width = 8, height = 10)
```

TP5 v TP7 Ambient 
```{r}
TP5_v_7_amb <- results(dds, contrast=c("Condition", "TP5_Ambient" , "TP7_Ambient"))
head(TP5_v_7_amb)
ma_plot<- plotMA(TP5_v_7_amb); ma_plot
sum(TP5_v_7_amb$padj <0.05, na.rm=T) # 61 genes with significant p-values
TP5_v_7_amb.sig <- subset(TP5_v_7_amb, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP5_v_7_amb.sig$log2FoldChange <0, na.rm=T) # 29 genes are downregulated in TP7 ambient relative to TP5 ambient
sum(TP5_v_7_amb.sig$log2FoldChange >0, na.rm=T) # 32 genes are upregulated in TP7 ambient relative to TP5 ambient
TP5_v_7_amb.sig.list <- data[which(rownames(dds) %in% rownames(TP5_v_7_amb.sig)),] #subset list of sig transcripts from original count data
write.csv(counts(TP5_v_7_amb.sig.list), file = "../output/Molecular/mRNA/TP5_amb_v_TP7_amb_DEG.csv")
TP5_v_7_amb.rsig <- rlog(TP5_v_7_amb.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
pca_plot <- plotPCA(TP5_v_7_amb.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP5_v_7_amb.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP5_v_7_amb.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/Molecular/mRNA/ma_plot_TP5_amb_v_TP7_amb_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_TP5_amb_v_TP7_amb_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/heatmap_TP5_amb_v_TP7_amb_DEG.pdf", heatmap, width = 8, height = 10)
```

TP0 v TP5 Heat
```{r}
TP0_v_5_heat <- results(dds, contrast=c("Condition", "TP0_Ambient" , "TP5_Heat"))
head(TP0_v_5_heat)
ma_plot <- plotMA(TP0_v_5_heat); ma_plot
sum(TP0_v_5_heat$padj <0.05, na.rm=T) # 540 genes with significant p-values
TP0_v_5_heat.sig <- subset(TP0_v_5_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_5_heat.sig$log2FoldChange <0, na.rm=T) # 376 genes are downregulated in TP5 heat relative to TP0 ambient
sum(TP0_v_5_heat.sig$log2FoldChange >0, na.rm=T) # 164 genes are upregulated in TP5 heat relative to TP0 ambient
TP0_v_5_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_5_heat.sig)),] #subset list of sig transcripts from original count data
write.csv(counts(TP0_v_5_heat.sig.list), file = "../output/Molecular/mRNA/TP0_amb_v_TP5_heat_DEG.csv")
TP0_v_5_heat.rsig <- rlog(TP0_v_5_heat.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
pca_plot <- plotPCA(TP0_v_5_heat.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP0_v_5_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP0_v_5_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/Molecular/mRNA/ma_plot_TP0_amb_v_TP5_heat_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_TP0_amb_v_TP5_heat_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/heatmap_TP0_amb_v_TP5_heat_DEG.pdf", heatmap, width = 8, height = 10)
```

TP0 v TP7 Heat
```{r}
TP0_v_7_heat <- results(dds, contrast=c("Condition", "TP0_Ambient" , "TP7_Heat"))
head(TP0_v_7_heat)
ma_plot <- plotMA(TP0_v_7_heat); ma_plot
sum(TP0_v_7_heat$padj <0.05, na.rm=T) # 863 genes with significant p-values
TP0_v_7_heat.sig <- subset(TP0_v_7_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_7_heat.sig$log2FoldChange <0, na.rm=T) # 549 genes are downregulated in TP7 heat relative to TP0 ambient
sum(TP0_v_7_heat.sig$log2FoldChange >0, na.rm=T) # 314 genes are upregulated in TP7 heat relative to TP0 ambient
TP0_v_7_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_7_heat.sig)),] #subset list of sig transcripts from original count data
write.csv(counts(TP0_v_7_heat.sig.list), file = "../output/Molecular/mRNA/TP0_amb_v_TP7_heat_DEG.csv")
TP0_v_7_heat.rsig <- rlog(TP0_v_7_heat.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
pca_plot <- plotPCA(TP0_v_7_heat.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP0_v_7_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP0_v_7_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/Molecular/mRNA/ma_plot_TP0_amb_v_TP7_heat_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_TP0_amb_v_TP7_heat_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/heatmap_TP0_amb_v_TP7_heat_DEG.pdf", heatmap, width = 8, height = 10)
```

TP5 Heat v TP7 Heat
```{r}
TP5_heat_v_7_heat <- results(dds, contrast=c("Condition", "TP5_Heat" , "TP7_Heat"))
head(TP5_heat_v_7_heat)
ma_plot <- plotMA(TP5_heat_v_7_heat); ma_plot
sum(TP5_heat_v_7_heat$padj <0.05, na.rm=T) # 9 genes with significant p-values
TP5_heat_v_7_heat.sig <- subset(TP5_heat_v_7_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP5_heat_v_7_heat.sig$log2FoldChange <0, na.rm=T) # 5 genes are downregulated in TP7 heat relative to TP5 heat
sum(TP5_heat_v_7_heat.sig$log2FoldChange >0, na.rm=T) # 4 genes are upregulated in TP7 heat relative to TP5 heat
TP5_heat_v_7_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP5_heat_v_7_heat.sig)),] #subset list of sig transcripts from original count data
write.csv(counts(TP5_heat_v_7_heat.sig.list), file = "../output/Molecular/mRNA/TP5_heat_v_TP7_heat_DEG.csv")
TP5_heat_v_7_heat.rsig <- rlog(TP5_heat_v_7_heat.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
pca_plot <- plotPCA(TP5_heat_v_7_heat.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP5_heat_v_7_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP5_heat_v_7_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/Molecular/mRNA/ma_plot_TP5_heat_v_TP7_heat_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_TP5_heat_v_TP7_heat_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/heatmap_TP5_heat_v_TP7_heat_DEG.pdf", heatmap, width = 8, height = 10)
```

TP5 Amb v TP5 Heat
```{r}
TP5_amb_v_TP5_heat <- results(dds, contrast=c("Condition", "TP5_Ambient" , "TP5_Heat"))
head(TP5_amb_v_TP5_heat)
ma_plot <- plotMA(TP5_amb_v_TP5_heat)
sum(TP5_amb_v_TP5_heat$padj <0.05, na.rm=T) # 9 genes with significant p-values
TP5_amb_v_TP5_heat.sig <- subset(TP5_amb_v_TP5_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP5_amb_v_TP5_heat.sig$log2FoldChange <0, na.rm=T) # 16 genes are downregulated in TP5 heat relative to TP5 amb
sum(TP5_amb_v_TP5_heat.sig$log2FoldChange >0, na.rm=T) # 29 genes are upregulated in TP5 heat relative to TP5 amb
TP5_amb_v_TP5_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP5_amb_v_TP5_heat.sig)),] #subset list of sig transcripts from original count data
write.csv(counts(TP5_amb_v_TP5_heat.sig.list), file = "../output/Molecular/mRNA/TP5_amb_v_TP5_heat_DEG.csv")
TP5_amb_v_TP5_heat.rsig <- rlog(TP5_amb_v_TP5_heat.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
pca_plot <- plotPCA(TP5_amb_v_TP5_heat.rsig, intgroup = c("Condition")); pca_plot #Plot PCA of all samples for DEG only
mat <- assay(TP5_amb_v_TP5_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP5_amb_v_TP5_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row", clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/Molecular/mRNA/ma_plot_TP5_amb_v_TP5_heat_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_TP5_amb_v_TP5_heat_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/heatmap_TP5_amb_v_TP5_heat_DEG.pdf", heatmap, width = 8, height = 10)
```

TP7 Amb v TP7 Heat
```{r}
TP7_amb_v_TP7_heat <- results(dds, contrast=c("Condition", "TP7_Ambient" , "TP7_Heat"))
head(TP7_amb_v_TP7_heat)
plotMA(TP7_amb_v_TP7_heat)
sum(TP7_amb_v_TP7_heat$padj <0.05, na.rm=T) # 2 genes with significant p-values
TP7_amb_v_TP7_heat.sig <- subset(TP7_amb_v_TP7_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP7_amb_v_TP7_heat.sig$log2FoldChange <0, na.rm=T) # 2 genes are downregulated in TP7 heat relative to TP7 amb
sum(TP7_amb_v_TP7_heat.sig$log2FoldChange >0, na.rm=T) # 0 genes are upregulated in TP7 heat relative to TP7 amb
TP7_amb_v_TP7_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP7_amb_v_TP7_heat.sig)),] #subset list of sig transcripts from original count data
write.csv(counts(TP7_amb_v_TP7_heat.sig.list), file = "../output/Molecular/mRNA/TP7_amb_v_TP7_heat_DEG.csv")
TP7_amb_v_TP7_heat.rsig <- rlog(TP7_amb_v_TP7_heat.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size

##### the rlog() step isn't working. Giving me this error: Error in estimateSizeFactorsForMatrix(counts(object), locfunc = locfunc, : every gene contains at least one zero, cannot compute log geometric means. I tried to apply a psuedo count to the filt_outrm_poa (ie adding 1 to each gene count) but it resulted in differenti numbers of DEGs for other comparisons. May need to come back to this. 





plotPCA(TP7_amb_v_TP7_heat.rsig, intgroup = c("Condition")) #Plot PCA of all samples for DEG only
mat <- assay(TP7_amb_v_TP7_heat.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP7_amb_v_TP7_heat.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row",
                         clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, 
                         show_colnames =TRUE) #plot heatmap of all DEG by group
```

TP5 Heat v TP7 Amb
```{r}
TP5_heat_v_TP7_amb <- results(dds, contrast=c("Condition", "TP5_Heat" , "TP7_Ambient"))
head(TP5_heat_v_TP7_amb)
ma_plot <- plotMA(TP5_heat_v_TP7_amb)
sum(TP5_heat_v_TP7_amb$padj <0.05, na.rm=T) 
TP5_heat_v_TP7_amb.sig <- subset(TP5_heat_v_TP7_amb, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP5_heat_v_TP7_amb.sig$log2FoldChange <0, na.rm=T)
sum(TP5_heat_v_TP7_amb.sig$log2FoldChange >0, na.rm=T)
TP5_heat_v_TP7_amb.sig.list <- data[which(rownames(dds) %in% rownames(TP5_heat_v_TP7_amb.sig)),] #subset list of sig transcripts from original count data
write.csv(counts(TP5_heat_v_TP7_amb.sig.list), file = "../output/Molecular/mRNA/TP5_heat_v_TP7_amb_DEG.csv")
TP5_heat_v_TP7_amb.rsig <- rlog(TP5_heat_v_TP7_amb.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
plotPCA(TP5_heat_v_TP7_amb.rsig, intgroup = c("Condition")) #Plot PCA of all samples for DEG only
mat <- assay(TP5_heat_v_TP7_amb.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(TP5_heat_v_TP7_amb.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row",
                         clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, 
                         show_colnames =TRUE); heatmap #plot heatmap of all DEG by group

## Save plots 
ggsave("../output/Molecular/mRNA/ma_plot_TP5_heat_v_TP7_amb_DEG.pdf", ma_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_TP5_heat_v_TP7_amb_DEG.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/heatmap_TP5_heat_v_TP7_amb_DEG.pdf", heatmap, width = 8, height = 10)
```

Look at unique genes 
```{r}
TP0_v_5_amb.DEGs <- as.data.frame(row.names(TP0_v_5_amb.sig.list))
colnames(TP0_v_5_amb.DEGs) <- "DEGs"
length(t(unique(TP0_v_5_amb.DEGs)))

TP0_v_7_amb.DEGs <- as.data.frame(row.names(TP0_v_7_amb.sig.list))
colnames(TP0_v_7_amb.DEGs) <- "DEGs"
length(t(unique(TP0_v_7_amb.DEGs)))

TP5_v_7_amb.DEGs <- as.data.frame(row.names(TP5_v_7_amb.sig.list))
colnames(TP5_v_7_amb.DEGs) <- "DEGs"
length(t(unique(TP5_v_7_amb.DEGs)))

TP0_v_5_heat.DEGs <- as.data.frame(row.names(TP0_v_5_heat.sig.list))
colnames(TP0_v_5_heat.DEGs) <- "DEGs"
length(t(unique(TP0_v_5_heat.DEGs)))

TP0_v_7_heat.DEGs <- as.data.frame(row.names(TP0_v_7_heat.sig.list))
colnames(TP0_v_7_heat.DEGs) <- "DEGs"
length(t(unique(TP0_v_7_heat.DEGs)))

TP5_heat_v_7_heat.DEGs <- as.data.frame(row.names(TP5_heat_v_7_heat.sig.list))
colnames(TP5_heat_v_7_heat.DEGs) <- "DEGs"
length(t(unique(TP5_heat_v_7_heat.DEGs)))

TP5_amb_v_TP5_heat.DEGs <- as.data.frame(row.names(TP5_amb_v_TP5_heat.sig.list))
colnames(TP5_amb_v_TP5_heat.DEGs) <- "DEGs"
length(t(unique(TP5_amb_v_TP5_heat.DEGs)))

TP7_amb_v_7_heat.DEGs <- as.data.frame(row.names(TP7_amb_v_TP7_heat.sig.list))
colnames(TP7_amb_v_7_heat.DEGs) <- "DEGs"
length(t(unique(TP7_amb_v_7_heat.DEGs)))

TP5_heat_v_7_amb.DEGs <- as.data.frame(row.names(TP5_heat_v_TP7_amb.sig.list))
colnames(TP5_heat_v_7_amb.DEGs) <- "DEGs"
length(t(unique(TP5_heat_v_7_amb.DEGs)))
```

Bind genes from all contrasts
```{r}
DEGs.all <- rbind(TP0_v_5_amb.DEGs, TP0_v_7_amb.DEGs, TP5_v_7_amb.DEGs, TP0_v_5_heat.DEGs, 
                TP0_v_7_heat.DEGs, TP5_heat_v_7_heat.DEGs, TP5_amb_v_TP5_heat.DEGs, 
                TP7_amb_v_7_heat.DEGs, TP5_heat_v_7_amb.DEGs)
DEGs.all <- unique(DEGs.all)
length(unique(DEGs.all$DEGs))

Unique.sig.list <- data[which(rownames(data) %in% DEGs.all$DEGs),] #subset list of sig transcripts from original count data
Unique.rsig <- rlog(Unique.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
write.csv(counts(Unique.sig.list), file="../output/Molecular/mRNA/Unique_DEGs.csv")
```

Visualize DEGs. 

PCA with DEGs
```{r}
pca_plot <- plotPCA(Unique.rsig, intgroup = c("Condition")); pca_plot
# percentVar_pca <- round(100*attr(pca_plot, "percentVar")) #plot PCA of samples with all data
# 
# pca_plot_degs <- ggplot(pca_plot, aes(PC1, PC2, color = Condition)) + 
#   geom_point(size=10) +
#   xlab(paste0("PC1: ",percentVar_pca[1],"% variance")) +
#   ylab(paste0("PC2: ",percentVar_pca[2],"% variance")) +
#   #scale_color_manual(values = wes_palette("GrandBudapest1", n=3)) +
#   #ggtitle(label = "M. capitata") +
#   theme_bw() + #Set background color
#   theme(legend.text = element_text(size=13), 
#         #legend.position="none",
#         plot.background = element_blank(),
#         legend.title = element_text(size=18, face="bold"), 
#         #legend.title=element_blank(),
#         axis.text = element_text(size=13), 
#         axis.title = element_text(size=15,  face="bold"), 
#         axis.title.y = element_text(vjust=-1.5),
#         plot.title = element_text(size = 20, face = "italic", hjust = 0.5)) +
#   coord_equal(ratio = 2); pca_plot_degs
##^^ code above wasn't working 

# Customize the PCA plot
pca_plot2 <- pca_plot + 
  #labs(x = paste0("PC1: ", percentVar_pca[1], "% variance"),
  #     y = paste0("PC2: ", percentVar_pca[2], "% variance")) +
  theme_bw() +
  theme(legend.text = element_text(size = 13),
        plot.background = element_blank(),
        legend.title = element_text(size = 18, face = "bold"),
        axis.text = element_text(size = 13),
        axis.title = element_text(size = 15, face = "bold"),
        axis.title.y = element_text(vjust = -1.5),
        plot.title = element_text(size = 20, face = "italic", hjust = 0.5)); pca_plot2

```

Need to make the plot prettier and figure out how to incorporate Treatment and timepoint 

Heatmap with DEGs
```{r}
mat <- assay(Unique.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(Unique.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row",
                         clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, 
                         show_colnames =TRUE); heatmap #plot heatmap of all DEG by group
```

Also need to make this prettier 

Save DEG plots 
```{r}
ggsave("../output/Molecular/mRNA/pca_plot_DEGs.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/mRNA/pca_plot_DEGs.png", pca_plot, width = 10, height = 8)

ggsave("../output/Molecular/mRNA/heatmap_DEGs.pdf", heatmap, width = 8, height = 10)
ggsave("../output/Molecular/mRNA/heatmap_DEGs.png", heatmap, width = 8, height = 10)
```

