---
title: "DESeq2 miRNAs"
author: "Jill Ashey"
date: "2024-03-19"
output: html_document
---

Use DESeq2 to evaluate differential expression of miRNAs; using miRNAs generated through ShortStack.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(car)
library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(car)
library(gplots)
library(wesanderson)
library(cowplot)
library(gridExtra)
```

Read in data and select only counts data 
```{r}
counts <- read.table("../output/Molecular/smRNA/shortstack/Counts.csv", header = T) 
```

Select only miRNAs with a Y in MIRNA column 
```{r}
counts <- counts %>%
  dplyr::filter(MIRNA == "Y")
```

Should be 51 miRNAs.

Combine Name and Coords columns for miRNA name. Remove MIRNA column 
```{r}
counts <- counts %>%
  mutate(miRNA_name = paste(Name, Coords, sep = "_")) %>%
  dplyr::select(-c(Name, Coords, MIRNA))
```

Make miRNA_name the row names and remove it as a column 
```{r}
rownames(counts) <- counts$miRNA_name
counts <- counts %>%
  dplyr::select(-miRNA_name)
```

Remove extra info from column names
```{r}
colnames(counts) <- gsub("_R1_001.fastq.gz_1", "", colnames(counts))
colnames(counts) <- gsub("AST.", "AST-", colnames(counts))
```

Read in metadata 
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv")
meta <- dplyr::arrange(meta, ID) #%>% # rearrange metadata so IDs are sorted in descending order 
  #mutate(Treatment = ifelse(Treatment == "Acclimation", "Ambient", Treatment))
  
# Set variables as factors 
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP0", "TP5", "TP7"))
meta$Treatment <- factor(meta$Treatment, levels = c("Acclimation","Ambient", "Heat"))

# Remove sample that did not sequence well 
meta <- meta %>%
  dplyr::filter(ID != "AST-1105")
```

Data sanity checks!
```{r}
meta$ID %in% colnames(counts) #are all of the sample names (rows) in the metadata df in the gene count matrix? Should be TRUE. 
all(rownames(meta$ID) == colnames(counts)) #are they the same in the same order? Should be TRUE
```

Typically, I filter reads by a proportion of samples containing cutoff value. Because I only have 51 miRNAs, I am not going to filter initially. 

Merge the Treatment and Timepoint columns together into one column called 'Condition'. Then run DESeq2 so the design is ~Condition

Merge Treatment and Timepoint into a single column 
```{r}
meta$Condition <- paste(meta$Timepoint, meta$Treatment, sep = "_")
meta$Condition <- factor(meta$Condition, levels = c("TP0_Acclimation", "TP5_Ambient", "TP5_Heat", "TP7_Ambient", "TP7_Heat"))
```

Run DESeq2
```{r}
data <- DESeqDataSetFromMatrix(countData = counts, colData = meta, design = ~Condition)
dds <- DESeq(data)
resultsNames(dds)
```

TP0 v TP5 Ambient
```{r}
# Look at results 
TP0_v_5_amb <- results(dds, contrast=c("Condition", "TP0_Acclimation" , "TP5_Ambient"))
head(TP0_v_5_amb)

# Save results for all miRNAs 
TP0_v_5_amb_df <- as.data.frame(TP0_v_5_amb)
write.csv(TP0_v_5_amb_df, file = "../output/Molecular/smRNA/shortstack/TP0_v_5_amb_miRNA.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP0_v_5_amb); ma_plot

# Subset miRNAs with significant padj values
sum(TP0_v_5_amb$padj <0.05, na.rm=T) # miRNA with significant p-values
TP0_v_5_amb.sig <- subset(TP0_v_5_amb, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_5_amb.sig$log2FoldChange <0, na.rm=T) # miRNAs downregulated in TP5 ambient relative to TP0 ambient
sum(TP0_v_5_amb.sig$log2FoldChange >0, na.rm=T) # miRNAs upregulated in TP5 ambient relative to TP0 ambient
TP0_v_5_amb.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_5_amb.sig)),] #subset list of sig transcripts from original count data

# Save info with pvalues and LFC
TP0_v_5_amb.sig <- as.data.frame(TP0_v_5_amb.sig)
TP0_v_5_amb.sig.list <- as.data.frame(counts(TP0_v_5_amb.sig.list))
TP0_v_5_amb.sig.list.full <- cbind(TP0_v_5_amb.sig, TP0_v_5_amb.sig.list)
write.csv(TP0_v_5_amb.sig.list.full, file = "../output/Molecular/smRNA/shortstack/TP0_amb_v_TP5_amb_DEM.csv")
```

TP0 v TP7 Ambient
```{r}
# Look at results 
TP0_v_7_amb <- results(dds, contrast=c("Condition", "TP0_Acclimation" , "TP7_Ambient"))
head(TP0_v_7_amb)

# Save results for all miRNAs 
TP0_v_7_amb_df <- as.data.frame(TP0_v_7_amb)
write.csv(TP0_v_7_amb_df, file = "../output/Molecular/smRNA/shortstack/TP0_v_7_amb_miRNA.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP0_v_7_amb); ma_plot

# Subset miRNAs with significant padj values
sum(TP0_v_7_amb$padj <0.05, na.rm=T) # miRNAs with significant p-values
TP0_v_7_amb.sig <- subset(TP0_v_7_amb, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_7_amb.sig$log2FoldChange <0, na.rm=T) # miRNAs downregulated in TP5 ambient relative to TP0 ambient
sum(TP0_v_7_amb.sig$log2FoldChange >0, na.rm=T) # miRNAs upregulated in TP5 ambient relative to TP0 ambient
TP0_v_7_amb.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_7_amb.sig)),] #subset list of sig transcripts from original count data

# Save info with pvalues and LFC
TP0_v_7_amb.sig <- as.data.frame(TP0_v_7_amb.sig)
TP0_v_7_amb.sig.list <- as.data.frame(counts(TP0_v_7_amb.sig.list))
TP0_v_7_amb.sig.list.full <- cbind(TP0_v_7_amb.sig, TP0_v_7_amb.sig.list)
write.csv(TP0_v_7_amb.sig.list.full, file = "../output/Molecular/smRNA/shortstack/TP0_amb_v_TP7_amb_DEM.csv")
```

TP5 v TP7 Ambient 
```{r}
# Look at results 
TP5_v_7_amb <- results(dds, contrast=c("Condition", "TP5_Ambient" , "TP7_Ambient"))
head(TP5_v_7_amb)

# Save results for all miRNAs 
TP5_v_7_amb_df <- as.data.frame(TP5_v_7_amb)
#write.csv(TP5_v_7_amb_df, file = "../output/Molecular/smRNA/TP5_v_7_amb_miRNA.csv")

# Volcano plot for LFC 
ma_plot<- plotMA(TP5_v_7_amb); ma_plot

# Subset miRNAs with significant padj values
sum(TP5_v_7_amb$padj <0.05, na.rm=T) # no miRNAs with significant p-values
```

TP0 v TP5 Heat
```{r}
# Look at results 
TP0_v_5_heat <- results(dds, contrast=c("Condition", "TP0_Acclimation" , "TP5_Heat"))
head(TP0_v_5_heat)

# Save results for all miRNAs 
TP0_v_5_heat_df <- as.data.frame(TP0_v_5_heat)
write.csv(TP0_v_5_heat_df, file = "../output/Molecular/smRNA/shortstack/TP0_v_5_heat_miRNA.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP0_v_5_heat); ma_plot

# Subset miRNAs with significant padj values
sum(TP0_v_5_heat$padj <0.05, na.rm=T) # miRNAs with significant p-values
TP0_v_5_heat.sig <- subset(TP0_v_5_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_5_heat.sig$log2FoldChange <0, na.rm=T) # miRNAs downregulated in TP5 heat relative to TP0 ambient
sum(TP0_v_5_heat.sig$log2FoldChange >0, na.rm=T) # miRNAs upregulated in TP5 heat relative to TP0 ambient
TP0_v_5_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_5_heat.sig)),] #subset list of sig transcripts from original count data

# Save info with pvalues and LFC
TP0_v_5_heat.sig <- as.data.frame(TP0_v_5_heat.sig)
TP0_v_5_heat.sig.list <- as.data.frame(counts(TP0_v_5_heat.sig.list))
TP0_v_5_heat.sig.list.full <- cbind(TP0_v_5_heat.sig, TP0_v_5_heat.sig.list)
write.csv(TP0_v_5_heat.sig.list.full, file = "../output/Molecular/smRNA/shortstack/TP0_amb_v_TP5_heat_DEM.csv")
```

TP0 v TP7 Heat
```{r}
# Look at results 
TP0_v_7_heat <- results(dds, contrast=c("Condition", "TP0_Acclimation" , "TP7_Heat"))
head(TP0_v_7_heat)

# Save results for all miRNAs 
TP0_v_7_heat_df <- as.data.frame(TP0_v_7_heat)
write.csv(TP0_v_7_heat_df, file = "../output/Molecular/smRNA/shortstack/TP0_v_7_heat_miRNA.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP0_v_7_heat); ma_plot

# Subset miRNAs with significant padj values
sum(TP0_v_7_heat$padj <0.05, na.rm=T) # miRNAs with significant p-values
TP0_v_7_heat.sig <- subset(TP0_v_7_heat, padj<0.05,) #identify signficant pvalues with 5%FDR
sum(TP0_v_7_heat.sig$log2FoldChange <0, na.rm=T) # miRNAs downregulated in TP7 heat relative to TP0 ambient
sum(TP0_v_7_heat.sig$log2FoldChange >0, na.rm=T) # miRNAs upregulated in TP7 heat relative to TP0 ambient
TP0_v_7_heat.sig.list <- data[which(rownames(dds) %in% rownames(TP0_v_7_heat.sig)),] #subset list of sig transcripts from original count data

# Save info with pvalues and LFC
TP0_v_7_heat.sig <- as.data.frame(TP0_v_7_heat.sig)
TP0_v_7_heat.sig.list <- as.data.frame(counts(TP0_v_7_heat.sig.list))
TP0_v_7_heat.sig.list.full <- cbind(TP0_v_7_heat.sig, TP0_v_7_heat.sig.list)
write.csv(TP0_v_7_heat.sig.list.full, file = "../output/Molecular/smRNA/shortstack/TP0_amb_v_TP7_heat_DEM.csv")
```

TP5 Heat v TP7 Heat
```{r}
# Look at results 
TP5_heat_v_7_heat <- results(dds, contrast=c("Condition", "TP5_Heat" , "TP7_Heat"))
head(TP5_heat_v_7_heat)

# Save results for all miRNAs 
TP5_heat_v_7_heat_df <- as.data.frame(TP5_heat_v_7_heat)
write.csv(TP5_heat_v_7_heat_df, file = "../output/Molecular/smRNA/shortstack/TP5_heat_v_7_heat_miRNA.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP5_heat_v_7_heat); ma_plot

# Subset miRNAs with significant padj values
sum(TP5_heat_v_7_heat$padj <0.05, na.rm=T) # no miRNAs with significant p-values
```

TP5 Amb v TP5 Heat
```{r}
# Look at results 
TP5_amb_v_TP5_heat <- results(dds, contrast=c("Condition", "TP5_Ambient" , "TP5_Heat"))
head(TP5_amb_v_TP5_heat)

# Save results for all miRNAs 
TP5_amb_v_TP5_heat_df <- as.data.frame(TP5_amb_v_TP5_heat)
write.csv(TP5_amb_v_TP5_heat_df, file = "../output/Molecular/smRNA/shortstack/TP5_amb_v_TP5_heat_miRNA.csv")

# Volcano plot for LFC 
ma_plot <- plotMA(TP5_amb_v_TP5_heat)

# Subset miRNAs with significant padj values
sum(TP5_amb_v_TP5_heat$padj <0.05, na.rm=T) # no miRNAs with significant p-values
```

TP7 Amb v TP7 Heat
```{r}
# Look at results
TP7_amb_v_TP7_heat <- results(dds, contrast=c("Condition", "TP7_Ambient" , "TP7_Heat"))
head(TP7_amb_v_TP7_heat)

# Save results for all miRNAs 
TP7_amb_v_TP7_heat_df <- as.data.frame(TP7_amb_v_TP7_heat)
write.csv(TP7_amb_v_TP7_heat_df, file = "../output/Molecular/smRNA/shortstack/TP7_amb_v_TP7_heat_miRNA.csv")

# Volcano plot for LFC
plotMA(TP7_amb_v_TP7_heat)

# Subset miRNAs with significant padj values
sum(TP7_amb_v_TP7_heat$padj <0.05, na.rm=T) # no miRNAs with significant p-values
```

TP5 Heat v TP7 Amb
```{r}
# Look at results
TP5_heat_v_TP7_amb <- results(dds, contrast=c("Condition", "TP5_Heat" , "TP7_Ambient"))
head(TP5_heat_v_TP7_amb)

# Save results for all miRNAs 
TP5_heat_v_TP7_amb_df <- as.data.frame(TP5_heat_v_TP7_amb)
write.csv(TP5_heat_v_TP7_amb_df, file = "../output/Molecular/smRNA/shortstack/TP5_heat_v_TP7_amb_miRNA.csv")

# Volcano plot for LFC
ma_plot <- plotMA(TP5_heat_v_TP7_amb)

# Subset miRNAs with significant padj values
sum(TP5_heat_v_TP7_amb$padj <0.05, na.rm=T) # no miRNAs with significant p-values
```

Look at unique DEMs
```{r}
TP0_v_5_amb.DEMs <- as.data.frame(row.names(TP0_v_5_amb.sig.list))
colnames(TP0_v_5_amb.DEMs) <- "DEMs"
length(t(unique(TP0_v_5_amb.DEMs)))

TP0_v_7_amb.DEMs <- as.data.frame(row.names(TP0_v_7_amb.sig.list))
colnames(TP0_v_7_amb.DEMs) <- "DEMs"
length(t(unique(TP0_v_7_amb.DEMs)))

TP0_v_5_heat.DEMs <- as.data.frame(row.names(TP0_v_5_heat.sig.list))
colnames(TP0_v_5_heat.DEMs) <- "DEMs"
length(t(unique(TP0_v_5_heat.DEMs)))

TP0_v_7_heat.DEMs <- as.data.frame(row.names(TP0_v_7_heat.sig.list))
colnames(TP0_v_7_heat.DEMs) <- "DEMs"
length(t(unique(TP0_v_7_heat.DEMs)))
```

Bind genes from all contrasts
```{r}
DEMs.all <- rbind(TP0_v_5_amb.DEMs, 
                  TP0_v_7_amb.DEMs, 
                  TP0_v_5_heat.DEMs, 
                  TP0_v_7_heat.DEMs) 
DEMs.all <- unique(DEMs.all)
length(unique(DEMs.all$DEMs))

Unique.sig.list <- data[which(rownames(data) %in% DEMs.all$DEMs),] #subset list of sig transcripts from original count data
Unique.rsig <- rlog(Unique.sig.list, blind=FALSE) #apply a regularized log transformation to minimize effects of small counts and normalize wrt library size
write.csv(counts(Unique.sig.list), file="../output/Molecular/smRNA/shortstack/Unique_DEMs.csv")
```

There are 12 DEMs. 

PCA with DEMs
```{r}
pca_plot <- plotPCA(Unique.rsig, intgroup = c("Condition")); pca_plot
# percentVar_pca <- round(100*attr(pca_plot, "percentVar")) #plot PCA of samples with all data
# 
# pca_plot_degs <- ggplot(pca_plot, aes(PC1, PC2, color = Condition)) + 
#   geom_point(size=10) +
#   xlab(paste0("PC1: ",percentVar_pca[1],"% variance")) +
#   ylab(paste0("PC2: ",percentVar_pca[2],"% variance")) +
#   #scale_color_manual(values = wes_palette("GrandBudapest1", n=3)) +
#   #ggtitle(label = "M. capitata") +
#   theme_bw() + #Set background color
#   theme(legend.text = element_text(size=13), 
#         #legend.position="none",
#         plot.background = element_blank(),
#         legend.title = element_text(size=18, face="bold"), 
#         #legend.title=element_blank(),
#         axis.text = element_text(size=13), 
#         axis.title = element_text(size=15,  face="bold"), 
#         axis.title.y = element_text(vjust=-1.5),
#         plot.title = element_text(size = 20, face = "italic", hjust = 0.5)) +
#   coord_equal(ratio = 2); pca_plot_degs
##^^ code above wasn't working 

# Customize the PCA plot
pca_plot2 <- pca_plot + 
  #labs(x = paste0("PC1: ", percentVar_pca[1], "% variance"),
  #     y = paste0("PC2: ", percentVar_pca[2], "% variance")) +
  theme_bw() +
  theme(legend.text = element_text(size = 13),
        plot.background = element_blank(),
        legend.title = element_text(size = 18, face = "bold"),
        axis.text = element_text(size = 13),
        axis.title = element_text(size = 15, face = "bold"),
        axis.title.y = element_text(vjust = -1.5),
        plot.title = element_text(size = 20, face = "italic", hjust = 0.5)); pca_plot2

```

Need to make the plot prettier and figure out how to incorporate Treatment and timepoint 

Heatmap with DEMs
```{r}
mat <- assay(Unique.rsig) #make an expression object
col.order <- c( "AST-1147", "AST-1567", "AST-1722", "AST-2398", # TP0, Ambient
              "AST-1412", "AST-1617", "AST-2000", "AST-2360", # TP5, Ambient
              "AST-2404", "AST-2412", "AST-2512", "AST-2563",  # TP5, Heat
              "AST-1560", "AST-2302", "AST-2523",  # TP7, Ambient
              "AST-1065", "AST-2007", "AST-2729", "AST-2755")  # TP7, Heat 
mat <- mat[,col.order]
df <- as.data.frame(colData(Unique.rsig)[,c("Treatment", "Timepoint", "Condition")]) #make dataframe
df <- df[order(df$Condition),]
heatmap <- pheatmap(mat, annotation_col=df, clustering_method = "average", scale="row",
                         clustering_distance_rows="euclidean", show_rownames =TRUE,fontsize_row = 4, cluster_cols=FALSE, cluster_rows=TRUE, 
                         show_colnames =TRUE); heatmap #plot heatmap of all DEG by group
```

Also need to make this prettier 

Save DEM plots 
```{r}
ggsave("../output/Molecular/smRNA/shortstack/pca_plot_DEMs.pdf", pca_plot, width = 10, height = 8)
ggsave("../output/Molecular/smRNA/shortstack/pca_plot_DEMs.png", pca_plot, width = 10, height = 8)

pdf("../output/Molecular/smRNA/shortstack/heatmap_DEMs.pdf", width = 8, height = 10)
print(heatmap)
dev.off()
```
