---
title: "DESeq2 - Astrangia mRNA"
output: html_document
date: "2023-10-02"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
#library(apeglm)
```

## Load and clean data 

Read in raw count data for mRNA 
```{r}
raw_counts <- read.csv("../data/Molecular/mRNA/Apoc_gene_count_matrix.csv")
rownames(raw_counts) <- raw_counts[,1] #set first column that contains gene names as rownames
raw_counts <- raw_counts[,-1] #remove the column with gene names
```

Remove extraneous info from sample name (eg trimmed.AST.1065_R1_001.fastq.gz.bam.gtf should be AST-1065)
```{r}
colnames(raw_counts)<-gsub("_[^_]+$", "",colnames(raw_counts)) #get rid of "_001.fastq.gz.bam.gtf"
colnames(raw_counts)<-gsub("_[^_]+$", "",colnames(raw_counts)) #get rid of "_R1"
colnames(raw_counts)<-gsub("trimmed.", "",colnames(raw_counts)) #get rid of "trimmed."
colnames(raw_counts)<-gsub("AST.", "AST-",colnames(raw_counts)) # replace "AST." with "AST-"
colnames(raw_counts)<-gsub(" ", "", colnames(raw_counts))

# rename columns all together? if they are not matching with the meta$ID 
colnames(raw_counts) <- c("AST-1065", "AST-1105", "AST-1147", "AST-1412", "AST-1560", "AST-1567", "AST-1617", "AST-1722", "AST-2000", "AST-2007", "AST-2302", "AST-2360", "AST-2398", "AST-2404", "AST-2412", "AST-2512", "AST-2523", "AST-2563", "AST-2729", "AST-2755")
```

Read in metadata 
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv")
meta <- dplyr::arrange(meta, ID) # rearrange metadata so IDs are sorted in descending order 
#meta$ID <- gsub("AST-", "AST.", meta$ID)

# Set variables as factors 
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP0", "TP5", "TP7"))
meta$Treatment <- factor(meta$Treatment, levels = c("Acclimation", "Ambient", "Heat"))
```

Data sanity checks!
```{r}
meta$ID %in% colnames(raw_counts) #are all of the sample names (rows) in the metadata df in the gene count matrix? Should be TRUE. 
all(rownames(meta$ID) == colnames(raw_counts)) #are they the same in the same order? Should be TRUE
```

## Filter reads by proportion of samples containing cutoff value 
```{r}
ffun<-filterfun(pOverA(0.75,5))  #set up filtering parameters--LOOK INTO FILTERING PARAMETERS
filt_outrm_poa <- genefilter((raw_counts), ffun) #apply filter
sum(filt_outrm_poa) #count number of genes left

filt_outrm_poa <- raw_counts[filt_outrm_poa,] #keep only rows that passed filter

all(rownames(meta$ID) %in% colnames(filt_outrm_poa)) # must come out TRUE
```

## Calculte dispersions 

"Using negative binomial models requires gene dispersion estimates to be made. This can be achieved in a number of ways. A common way to calculate this for gene i is to use the equation:

Dispersioni = (variancei - meani)/meani2

Starting with raw counts, so using [DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html), but could also import a normalized gene count matrix (i.e. transcripts per million) and calculate disperson manually as in the [vignette](https://cran.r-project.org/web/packages/glmmSeq/vignettes/glmmSeq.html).

```{r}
dds_filt_outrm <- DESeqDataSetFromMatrix(countData = filt_outrm_poa,
                              colData = meta,
                              design = ~Timepoint*Treatment)

dds_filt_outrm <- DESeq(dds_filt_outrm) #apply DEseq dispersion calculation
dispersions_filt_outrm <- setNames(dispersions(dds_filt_outrm), rownames(filt_outrm_poa)) #save dispersions as vector for each gene

dds_norm_counts <- as.data.frame(counts(dds_filt_outrm, normalized=TRUE)) #save normalized counts from DEseq object for downstream use


```


20231221
Error in checkFullRank(modelMatrix) :
the model matrix is not full rank, so the model cannot be fit as specified.
One or more variables or interaction terms in the design formula are linear
combinations of the others and must be removed.

Please read the vignette section 'Model matrix not full rank':

vignette('DESeq2')

Went to the DESeq2 manual and read about model matrix not full rank: "There are two main reasons for this problem: either one or more columns in the model matrix are linear combinations of other columns, or there are levels of factors or combinations of levels of multiple factors which are missing samples." I think in my case, it is an issue of linear combos. Let's see what deseq2 recommends us to do... 

An example: 

```{r}
coldata <- DataFrame(grp=factor(rep(c("X","Y"),each=6)),
                     ind=factor(rep(1:6,each=2)),
                     cnd=factor(rep(c("A","B"),6)))
coldata
as.data.frame(coldata)

```





















## Do variance stabilizing transformation 

First, estimate size factors to make sure we can use vst on our data
```{r}
size_factors <- estimateSizeFactors(dds_filt_outrm)
print(sizeFactors(size_factors))

# All less than 4, can use vst 
```

Apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
```{r}
vst <- vst(dds_filt_outrm)
```

Make PCA plot using vst object 
```{r}
vst_PCAdata <- plotPCA(vst, intgroup = c("Treatment", "Timepoint"), returnData=TRUE) # create PCA loadings 
percentVar <- round(100*attr(vst_PCAdata, "percentVar")) #plot PCA of samples with all data
PCAplot <- ggplot(vst_PCAdata, aes(PC1, PC2, color=Treatment, shape =Timepoint)) + 
   geom_point(size=3) +
   geom_text(aes(label=name),hjust=0, vjust=0) +
   xlab(paste0("PC1: ",percentVar[1],"% variance")) +
   ylab(paste0("PC2: ",percentVar[2],"% variance")) +
   #scale_color_manual(values = c(control="cadetblue3", mid="palevioletred", high="darkgreen", unknown="black")) +
   coord_fixed() +
   #ggtitle("P. lobata - all genes") +
   theme_bw() + #Set background color
   theme(panel.border = element_blank(), # Set border
         #panel.grid.major = element_blank(), #Set major gridlines
         #panel.grid.minor = element_blank(), #Set minor gridlines
         axis.line = element_line(colour = "black"), #Set axes color
         plot.background=element_blank()) #Set the plot background
PCAplot 
```

Save vst object for downstream analysis 
```{r}
vst <- assay(vst) # call only the transformed counts
vst <- as.data.frame(vst)
vst <- tibble::rownames_to_column(vst, var = "Gene")

write.csv(vst, "../output/vst_mRNA.csv") #save this as csv for downstream analysis
```




20231003
```{r}
dds_filt_outrm <- DESeqDataSetFromMatrix(countData = filt_outrm_poa,
                              colData = meta,
                              design = ~Treatment + Timepoint + Treatment*Timepoint)

dds_filt_outrm <- DESeq(dds_filt_outrm) #apply DEseq dispersion calculation
res <- results(dds_filt_outrm)
resultsNames(dds_filt_outrm)

# [1] "Intercept"                  "Treatment_Heat_vs_Ambient"  "Timepoint_TP5_vs_TP0"       "Timepoint_TP7_vs_TP0"       "TreatmentHeat.TimepointTP5" "TreatmentHeat.TimepointTP7"
## This doesn't really make sense to me. Why are not all of the pairwise comparisons present?
```

Trying something that the DESeq2 manual recommended based on the interactions
```{r}
dds_filt_outrm$group <- factor(paste0(dds_filt_outrm$Timepoint, dds_filt_outrm$Treatment))
design(dds_filt_outrm) <- ~group
dds <- DESeq(dds_filt_outrm)
resultsNames(dds)
results(dds)
colData(dds)
# Does not have all the interactions that I want...

# multifactor design
design(dds) <- formula(~ Treatment + Timepoint)
ddsMF <- DESeq(dds)
resMF <- results(ddsMF)
head(resMF)
resMFtype <- results(ddsMF, contrast = c("Treatment", "Ambient", "Heat"))
summary(resMFtype)
```

Trying to the quick start 
```{r}
design <- formula(~Timepoint + Treatment + Timepoint:Treatment)

dds <- DESeqDataSetFromMatrix(countData = filt_outrm_poa,
                              colData = meta,
                              design = design)
dds <- DESeq(dds)
resultsNames(dds)

trt <- results(dds, contrast = c("Treatment", "Ambient", "Heat"))
summary(trt)
sum(trt$padj<0.05, na.rm = T)

tp_1 <- results(dds, contrast = c("Timepoint", "TP0", "TP5"))
summary(tp_1)
sum(tp_1$padj<0.05, na.rm = T)

tp_2 <- results(dds, contrast = c("Timepoint", "TP0", "TP7"))
summary(tp_2)
sum(tp_2$padj<0.05, na.rm = T)

tp_3 <- results(dds, contrast = c("Timepoint", "TP5", "TP7"))
summary(tp_3)
sum(tp_3$padj<0.05, na.rm = T)

blah <- results(dds, contrast = c("Ambient", "TP0"))
summary(blah)
sum(tp_3$padj<0.05, na.rm = T)

# But I can't contrast timepoint and treatment at the same time 
```

Trying to suggestion where I combine both variables into one and run it that way
```{r}
meta$trt_time <- factor(paste0(meta$Treatment, "_", meta$Timepoint))

dds <- DESeqDataSetFromMatrix(countData = filt_outrm_poa,
                              colData = meta,
                              design = ~trt_time)
dds <- DESeq(dds)
resultsNames(dds)

test1 <- results(dds, contrast = c("trt_time", "Ambient_TP0", "Heat_TP0"))
summary(test1)
sum(test1$padj<0.05, na.rm = T) 

test2 <- results(dds, contrast = c("trt_time", "Ambient_TP5", "Heat_TP5"))
summary(test2)
sum(test2$padj<0.05, na.rm = T)
plotMA(test2)

test3 <- results(dds, contrast = c("trt_time", "Ambient_TP7", "Heat_TP7"))
summary(test3)
sum(test3$padj<0.05, na.rm = T)
plotMA(test3)

test4 <- results(dds, contrast = c("trt_time", "Ambient_TP0", "Ambient_TP5"))
summary(test4)
sum(test4$padj<0.05, na.rm = T)
plotMA(test4)

test5 <- results(dds, contrast = c("trt_time", "Ambient_TP5", "Ambient_TP7"))
summary(test5)
sum(test5$padj<0.05, na.rm = T)
plotMA(test5)

## Could continue with pairwise comparisons. I still want to figure out if I can just do the design as ~Treatment + Timepoint + Treatment:Timepoint instead of having to combine them into one group. Though the combining of them into a group is one of the recommendations from the manual.
```

Lots of my adjusted p-values are set to NAs. This is the explanation from the DESeq manual that best describes thats going on: "If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p value will be set to NA. Description and customization of independent filtering is described below."

So a lot of my data is not passing the independent filtering threshold that DESeq is setting. Notes on independent filtering of results from the manual: "The results function of the DESeq2 package performs independent filtering by default using the mean of normalized counts as a filter statistic. A threshold on the filter statistic is found which optimizes the number of adjusted p values lower than a significance level alpha (we use the standard variable name for significance level, though it is unrelated to the dispersion parameter α
). The theory behind independent filtering is discussed in greater detail below. The adjusted p values for the genes which do not pass the filter threshold are set to NA."

```{r}
metadata(test2)$alpha

metadata(test2)$filterThreshold # what does the filter threshold mean? That 81% of my genes do not pass the threshold? 

plot(metadata(test2)$filterNumRej, 
     type="b", ylab="number of rejections",
     xlab="quantiles of filter")
lines(metadata(test2)$lo.fit, col="red")
abline(v=metadata(test2)$filterTheta)
```


What if we removed TP0??
















20231221
Step 1 - create deseq dataset 
```{r}
# Ensure column names in count data match the IDs in the metadata
colnames(filt_outrm_poa) <- meta$ID

# Create DESeqDataSet
dds <- DESeqDataSetFromMatrix(
    countData = filt_outrm_poa,
    colData = meta,
    design = ~ Timepoint + Treatment
)
```

Still running into this 'Model matrix not full rank'

Check levels in variables 
```{r}
table(meta$Timepoint)
table(meta$Treatment)
# they have more than 1 level each 
```

Check for perfect correlation - Inspect if there is a perfect correlation between Timepoint and Treatment. If every level of Timepoint always goes with the same level of Treatment (or vice versa), they cannot both be used in the model.


Maybe remove TP0 in this case? 
```{r}
meta_sub <- subset(meta, Timepoint != "TP0")

filt_outrm_poa_sub <- filt_outrm_poa[, meta_sub$ID]

# Create DESeqDataSet
dds_sub <- DESeqDataSetFromMatrix(
    countData = filt_outrm_poa_sub,
    colData = meta_sub,
    design = ~ Treatment
)

# Run deseq
dds_sub <- DESeq(dds_sub)

# look at results 
resultsNames(dds_sub)

# comparing treatments within a specific timepoint 
res <- results(dds_sub, contrast=list("Treatment_Heat_vs_Ambient"))
res$log2FoldChange

res <- results(dds_sub, contrast=list("Timepoint_TP7_vs_TP5", "Treatment_Heat"))
```




















