---
title: "mirtarrnaseq across time point and treatment"
author: "Jill Ashey"
date: "2024-05-15"
output: html_document
---

This script will use the R package [mirTarRnaSeq](https://bioconductor.org/packages/3.13/bioc/html/mirTarRnaSeq.html), which an be used for interactive mRNA miRNA sequencing statistical analysis, to assess coexpression between mRNAs and miRNAs. This package utilizes differential expression mRNA and miRNA sequencing results and performs correlation and various analyses between mRNA and miRNA expriments. These experiments can be time point experiments, and or condition expriments.

I will be using Part 3 of their workflow (see [vignette](https://bioconductor.org/packages/3.13/bioc/vignettes/mirTarRnaSeq/inst/doc/mirTarRnaSeq.pdf)) to assess is there are significant miRNA mRNA interrelationships between time points and/or conditions. The required input is fold change and p-value information on the genes and miRNAs, which I obtained from running DESeq2. In that code, I used pairwise comparisons to compare 'condition' (ie a combo of the timepoint and treatment). I may pivot to using the log ratio method or glmmseq(). 

The comparisons I can analyze are TP0_v_5_amb, TP0_v_7_amb, TP0_v_5_heat, and TP0_v_7_heat because these were the only comparisons that had differentially expressed miRNAs. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#BiocManager::install("mirTarRnaSeq")

library(tidyverse)
library(mirTarRnaSeq)
library(goseq)
library(pheatmap)
library(viridis)
```

I will be initially comparing TP0 amb v TP5 amb, TP0 amb v TP7 amb, TP0 amb v TP5 heat, and TP0 amb v TP7 heat 

Read in mRNA file of interest. In this first iteration, I'm interested in TP0 amb v TP5 amb. 
```{r}
TP0_v_5_amb_df <- read.csv("../output/Molecular/mRNA/TP0_v_5_amb.csv") %>%
  dplyr::select(X, log2FoldChange, padj) %>%
  dplyr::rename(gene_id = X, pvalue = padj)
```

Get mRNAs with particular fold change and pvalues 
```{r}
mrna_files <- list(TP0_v_5_amb_df)
mrna_data <- one2OneRnaMiRNA(mrna_files, gene_colname = "gene_id", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

Read in miRNA file of interest
```{r}
TP0_v_5_amb_miRNA_df <- read.csv("../output/Molecular/smRNA/TP0_v_5_amb_miRNA.csv") %>%
  dplyr::select(X, log2FoldChange, padj) %>%
  dplyr::rename(gene_id = X, pvalue = padj)
```

Get miRNAs with particular fold change and pvalues 
```{r}
mirna_files <- list(TP0_v_5_amb_miRNA_df)
mirna_data <- one2OneRnaMiRNA(mirna_files, gene_colname = "gene_id", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

Produce a mRNA miRNA interrelation df 
```{r}
inter0 <- twoTimePoint(mrna_data, mirna_data)
```

This function is calculating the absolute value of the difference between the mRNA and miRNA fold changes for each potential miRNA mRNA combo. 

Make background distribution
```{r, echo=F}
outs_background <- twoTimePointSamp(mrna_data, mirna_data, Shrounds = 100, Srounds = 1000)
```

Identify relationships below specific threshold
```{r}
sig_interRel <- threshSigInter(inter0, outs_background)
```

This function uses the background distribution to set a pvalue threshold, which is then applied to the value column.

Read in miranda data 
```{r}
miranda_data <- read.csv("~/Desktop/PutnamLab/Astrangia/2020-2021/miranda/miranda_gene_ids_all.csv", header = T)
```

The vignette includes the following information for the example mouse miranda data:  
```{r}
head(miRanda)
```

V1 is the miRNA id, V2 is the gene id, V3 is the score, V4 is energy in kcal/mol, V5 is the subject % identify, and V6 is the query % identity. Select these columns from my data, rearrange columns, and rename columns to V1-6. This will allow the following functions to properly read the miranda file. 
```{r}
# Select specific columns
miranda_data_filt <- miranda_data %>%
  dplyr::select(mirna, Score, Energy_kcal_mol, Subject_Identity, Query_Identity, gene_id)

# Remove any duplicate rows
miranda_data_filt <- unique(miranda_data_filt)

# Rearrange columns
miranda_data_filt <- miranda_data_filt[,c("mirna", "gene_id", "Score", "Energy_kcal_mol", "Subject_Identity", "Query_Identity")]

# Rename columns V1-6
miranda_data_filt <- miranda_data_filt %>%    
  rename(V1 = mirna, 
         V2 = gene_id,
         V3 = Score,
         V4 = Energy_kcal_mol, 
         V5 = Subject_Identity,
         V6 = Query_Identity)
```

Intersect miranda data with significant relationships
```{r}
results_TP0_v_TP5_amb <- mirandaIntersectInter(sig_interRel, outs_background, mrna_data, mirna_data, miranda_data_filt)

tail(results_TP0_v_TP5_amb$corrs)
```

Make plots showing miRNA and mRNA FC relationships 
```{r}
# Create results file for plotting 
results_TP0_v_TP5_amb_df <- finInterResult(results_TP0_v_TP5_amb)

par(mar=c(4,4,2,1))
drawInterPlots(mrna_data, mirna_data, results_TP0_v_TP5_amb_df)
```

Make heatmap for significant relationships
```{r}
mirRnaHeatmapDiff(results_TP0_v_TP5_amb$corrs, fontsize = 10)

# Save plot as PDF
pdf("../output/Molecular/interactions/TP0_v_TP5_amb_comparisons_sigcorr.pdf", height = 20, width = 25)
mirRnaHeatmapDiff(results_TP0_v_TP5_amb$corrs, fontsize = 25)
dev.off() # Close the PDF device

# Save plot as png
#png("../output/Molecular/interactions/TP0_v_TP5_amb_comparisons_sigcorr.png", height = 500, width = 500)
#mirRnaHeatmapDiff(results_TP0_v_TP5_amb$corrs, fontsize = 10)
#dev.off() # Close the png device
```

Combine results data into a single df 
```{r}
combined_df <- cbind(results_TP0_v_TP5_amb$mirna, results_TP0_v_TP5_amb$mrna, results_TP0_v_TP5_amb$corrs)

colnames(combined_df) <- c("mirna_FC", "mrna_FC", "mirna", "mrna", "value", "score", "energy", "subj_identity", "query_identity", "pvalue")

length(unique(combined_df$mrna))
unique(combined_df$mrna)

#write.csv(combined_df, file = "../output/Molecular/interactions/TP0_v_5_amb_miRNA_mRNA_correlations.csv")
```

Some of the mRNA FC are really high (low 20s) relative to the other ones, which is driving up the value calculation for the miRNA mRNA combo. I might filter out the extremely high mRNA FCs...
```{r}
# Have to rename the cols in combined_df so that it matches the column name that the mirRnaHeatmapDiff function is expecting 
colnames(combined_df) <- c("mirna_FC", "mrna_FC", "V1", "V2", "value", "V3", "V4", "V5", "V6", "pvalue")

# Remove any mRNA FC values that are >15 or <-15
combined_df_subset <- combined_df %>%
  dplyr::filter(value >= -15 & value <= 15)

mirRnaHeatmapDiff(combined_df_subset, fontsize = 10)

# Save plot as PDF
pdf("../output/Molecular/interactions/TP0_v_TP5_amb_comparisons_sigcorr_subset.pdf", height = 20, width = 25)
mirRnaHeatmapDiff(combined_df_subset, fontsize = 25)
dev.off() # Close the PDF device
```

Plot FCs
```{r}
ggplot(combined_df, aes(x = mrna_FC, y = mirna_FC)) +
  geom_point(size = 6) +
  geom_smooth(method = "lm", se = T) # Add regression line

# Better plot
plot1 <- ggplot(combined_df_subset, aes(x = mrna_FC, y = mirna_FC)) +
  geom_point(size = 4, shape = 21, color = "black", fill = "black") +  # Customize points
  #geom_vline(xintercept = 0, linetype = "solid", color = "black", size = 0.5) +  # Add vertical line at x = 0
  #geom_hline(yintercept = 0, linetype = "solid", color = "black", size = 0.5) +  # Add horizontal line at y = 0
  geom_smooth(method = "lm", se = TRUE, color = "black", linetype = "dashed", size = 1) +  # Customize regression line
  labs(
    x = "mRNA Fold Change",
    y = "miRNA Fold Change"
  ) +  # Add titles and labels
  theme_minimal() +  # Use a minimal theme for a clean look
  theme(
    plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),  # Center and style title
    axis.title = element_text(size = 14, face = "bold"),  # Style axis titles
    axis.text = element_text(size = 12)  # Style axis text
    #panel.grid.major = element_line(color = "grey80")  # Customize grid lines
    #panel.grid.minor = element_blank()  # Remove minor grid lines
  ); plot1

ggsave("../output/Molecular/interactions/TP0_amb_v_TP5_amb_FC_plot.pdf", plot1, width = 10, height = 8)
ggsave("../output/Molecular/interactions/TP0_amb_v_TP5_amb_FC_plot.png", plot1, width = 10, height = 8)
```

I'm interested in the functions of these genes that are correlated with specific miRNAs. Perform GOSeq to identify any functional enrichment

Read in expressed genes post-filtering
```{r}
mrna_matrix <- read.csv("../output/Molecular/mRNA/filtered_gene_counts.csv") %>%
    #rename(gene_id = X) %>%
    dplyr::select(-"AST.1105") # Remove AST-1105 from mRNA df due to low counts
```

Read in gene lengths
```{r}
length <- read.csv("../output/Molecular/mRNA/gene_length.txt", header = F)

length <- length %>%
  separate(V1, into = c("gene_id", "length"), sep = " ", remove = FALSE) %>%
  dplyr::select(-V1)

str(length)

length$gene_id <- gsub("model", "TU", length$gene_id)
```

Merge length data with filtered count data
```{r}
length_merge <- left_join(length, mrna_matrix, by = c("gene_id"="X")) %>%
  na.omit()

str(length_merge)
```

GOseq requires a vector of all genes, all differentially expressed genes, and gene lengths. Make vectors 
```{r}
# Make DEG vector
cor_gene <- length_merge[length_merge$gene_id %in% combined_df$mrna, ]
cor_gene_names <- as.vector(combined_df$mrna)
gene_vector <- as.integer(length_merge$gene_id%in%cor_gene_names)
names(gene_vector) <- length_merge$gene_id

# Make ID vector
ID_vector <- as.vector(length_merge$gene_id)

# Make length vector
length_vector <- as.numeric(length_merge$length)
```

Calculate probability weighting function 
```{r}
DEG.pwf <- nullp(gene_vector, ID_vector, bias.data = length_vector)

str(DEG.pwf)
```

Code above giving me this warning: Warning: initial point very close to some inequality constraintsWarning: collapsing to unique 'x' values

Read in GO annotation information (annotation from Kate)
```{r}
annot <- read.delim("/Users/jillashey/Desktop/PutnamLab/Astrangia_Genome/Apoculata_v2.0_GeneAnnotation_combined_prelim.txt", header = T)
```

This annot file has annotation information from SwissProt, Trembl, and NCBI but only presents GO information for SwissProt and Trembl. 

Unite the GO terms from SwissProt and Trembl, select GO terms, and split so there is only one GO term per row. 
```{r}
annot_GO <- annot %>%
  unite("GOs", GO_Swiss.Prot, GO_Trembl, sep = ";", na.rm = TRUE) %>%
  dplyr::select(Protein_ID, GOs) %>%
  separate_rows(GOs, sep = ";") %>%
  dplyr::rename("gene_id" = "Protein_ID",
          "GO.ID" = "GOs") %>%
  filter(GO.ID != "")

# Remove leading and trailing whitespaces
annot_GO$GO.ID <- trimws(annot_GO$GO.ID)

# Set gene and go ids as factors 
annot_GO$gene_id <- as.factor(annot_GO$gene_id)
annot_GO$GO.ID <- as.factor(annot_GO$GO.ID)

# Select only unique rows
annot_GO <- unique(annot_GO)

# Look at unique gene and go ids 
length(unique(annot_GO$gene_id)) # 31823 genes have GO annotations
length(unique(annot_GO$GO.ID)) # there are 14581 unique GO annotations 

# Swap model with TU
annot_GO$gene_id <- gsub("model", "TU", annot_GO$gene_id)

# Look at unique gene and go ids that are present in this dataset 
length(unique(annot_GO$gene_id))
length(unique(annot_GO$GO.ID))

str(annot_GO)

# Convert filt_annot_GO from tibble to df 
annot_GO <-as.data.frame(annot_GO)
```

Perform GOSeq
```{r}
GO.wall<-goseq::goseq(DEG.pwf, ID_vector, gene2cat=annot_GO, method="Wallenius", use_genes_without_cat=T)

GO <- GO.wall[order(GO.wall$over_represented_pvalue),]
#write.csv(GO, file = "../output/Molecular/functional_enrichment/GOSeq_GO_all.csv")
```

Filter by significantly enriched GO terms (p < 0.05)
```{r}
GO_05 <- GO %>%
  dplyr::filter(over_represented_pvalue<0.05) %>%
        dplyr::arrange(., ontology, over_represented_pvalue)

#write.csv(GO_05, file = "../output/Molecular/functional_enrichment/GOSeq_GO_05.csv")
```

All GO terms in the original list were significantly over-represented. 

Plot all ontologies (BP, CC, MF) and order by p-value 
```{r}
GO_05_plot <- GO_05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_text(aes(label = over_represented_pvalue), hjust = -1, vjust = 0, size = 2) +
  geom_point(size=1, aes(colour = ontology)) +
  coord_flip() +
  ylim(0,10) +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()); GO_05_plot #Set the plot background #set title attributes

#ggsave("../output/Molecular/functional_enrichment/GOSeq_GO_05.pdf", GO_05_plot, width = 20, height = 10)
#ggsave("../output/Molecular/functional_enrichment/GOSeq_GO_05.png", GO_05_plot, width = 20, height = 10)
```

Merge significant GO terms and annot list together 
```{r}
GO_genes <- annot_GO %>%
  inner_join(GO_05, by = c("GO.ID" = "category"))

length(unique(GO_genes$gene_id))
```

Merge correlated genes with significant GO terms 
```{r}
GO_cor <- GO_genes %>%
  inner_join(combined_df, by = c("gene_id" = "mrna"))

length(unique(GO_cor$gene_id)) 
unique(GO_cor$gene_id)

#write.csv(GO_deg, file = "../output/Molecular/functional_enrichment/DEG_GO.csv")
```

Do we care about significantly enriched genes? Do we need goseq at all? Or do we just need to know the gene function?

Look at the next comparison: TP0 amb v TP7 amb 

Read in mRNA file of interest
```{r}
TP0_v_7_amb_df <- read.csv("../output/Molecular/mRNA/TP0_v_7_amb.csv") %>%
  dplyr::select(X, log2FoldChange, padj) %>%
  rename(gene_id = X, pvalue = padj)
```

Get mRNAs with particular fold change and pvalues 
```{r}
mrna_files <- list(TP0_v_7_amb_df)
mrna_data <- one2OneRnaMiRNA(mrna_files, gene_colname = "gene_id", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

Read in miRNA file of interest
```{r}
TP0_v_7_amb_miRNA_df <- read.csv("../output/Molecular/smRNA/TP0_v_7_amb_miRNA.csv") %>%
  dplyr::select(X, log2FoldChange, padj) %>%
  rename(gene_id = X, pvalue = padj)
```

Get miRNAs with particular fold change and pvalues 
```{r}
mirna_files <- list(TP0_v_7_amb_miRNA_df)
mirna_data <- one2OneRnaMiRNA(mirna_files, gene_colname = "gene_id", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

Produce a mRNA miRNA interrelation df 
```{r}
inter0 <- twoTimePoint(mrna_data, mirna_data)
```

This function is calculating the absolute value of the difference between the mRNA and miRNA fold changes for each potential miRNA mRNA combo. 

Make background distribution
```{r}
outs_background <- twoTimePointSamp(mrna_data, mirna_data, Shrounds = 100, Srounds = 1000)
```

Identify relationships below specific threshold
```{r}
sig_interRel <- threshSigInter(inter0, outs_background)
```

This function uses the background distribution to set a pvalue threshold, which is then applied to the value column.

Read in miranda data (doesn't need to be read in again if it was already read in above)
```{r}
#miranda_data <- read.csv("~/Desktop/PutnamLab/Astrangia/2020-2021/miranda/miranda_gene_ids_all.csv", header = T)
```

The vignette includes the following information for the example mouse miranda data:  
```{r}
#head(miRanda)
```

V1 is the miRNA id, V2 is the gene id, V3 is the score, V4 is energy in kcal/mol, V5 is the subject % identify, and V6 is the query % identity. Select these columns from my data, rearrange columns, and rename columns to V1-6. This will allow the following functions to properly read the miranda file. 
```{r}
# Select specific columns
# miranda_data_filt <- miranda_data %>%
#   dplyr::select(mirna, Score, Energy_kcal_mol, Subject_Identity, Query_Identity, gene_id)
# 
# # Remove any duplicate rows
# miranda_data_filt <- unique(miranda_data_filt)
# 
# # Rearrange columns
# miranda_data_filt <- miranda_data_filt[,c("mirna", "gene_id", "Score", "Energy_kcal_mol", "Subject_Identity", "Query_Identity")]
# 
# # Rename columns V1-6
# miranda_data_filt <- miranda_data_filt %>%    
#   rename(V1 = mirna, 
#          V2 = gene_id,
#          V3 = Score,
#          V4 = Energy_kcal_mol, 
#          V5 = Subject_Identity,
#          V6 = Query_Identity)
```

Intersect miranda data with significant relationships
```{r}
results_TP0_v_TP7_amb <- mirandaIntersectInter(sig_interRel, outs_background, mrna_data, mirna_data, miranda_data_filt)
```

Make plots showing miRNA and mRNA FC relationships 
```{r}
# Create results file for plotting 
results_TP0_v_TP7_amb_df <- finInterResult(results_TP0_v_TP7_amb)

par(mar=c(4,4,2,1))
drawInterPlots(mrna_data, mirna_data, results_TP0_v_TP7_amb_df)
```

Make heatmap for significant relationships
```{r}
mirRnaHeatmapDiff(results_TP0_v_TP7_amb$corrs, fontsize = 10)

pdf("../output/Molecular/interactions/TP0_v_TP7_amb_comparisons_sigcorr.pdf", height = 20, width = 25)
mirRnaHeatmapDiff(results_TP0_v_TP7_amb$corrs, fontsize = 20)
dev.off() # Close the PDF device

# Save plot as png
#png("../output/Molecular/interactions/TP0_v_TP7_amb_comparisons_sigcorr.png")
#mirRnaHeatmapDiff(results_TP0_v_TP7_amb$corrs, fontsize = 10)
#dev.off() # Close the png device
```

Combine results data into a single df 
```{r}
combined_df <- cbind(results_TP0_v_TP7_amb$mirna, results_TP0_v_TP7_amb$mrna, results_TP0_v_TP7_amb$corrs)

colnames(combined_df) <- c("mirna_FC", "mrna_FC", "mirna", "mrna", "value", "score", "energy", "subj_identity", "query_identity", "pvalue")

length(unique(combined_df$mrna))
unique(combined_df$mrna)

#write.csv(combined_df, file = "../output/Molecular/interactions/TP0_v_7_amb_miRNA_mRNA_correlations.csv")
```

Some of the mRNA FC are really high (low 20s) relative to the other ones, which is driving up the value calculation for the miRNA mRNA combo. I might filter out the extremely high mRNA FCs...
```{r}
# Have to rename the cols in combined_df so that it matches the column name that the mirRnaHeatmapDiff function is expecting 
colnames(combined_df) <- c("mirna_FC", "mrna_FC", "V1", "V2", "value", "V3", "V4", "V5", "V6", "pvalue")

# Remove any mRNA FC values that are >15 or <-15. Also remove Ser gene because it is likely an ncRNA
combined_df_subset <- combined_df %>%
  dplyr::filter(value >= -15 & value <= 15) %>%
  dplyr::filter(!V2 == "Ser_4434")

mirRnaHeatmapDiff(combined_df_subset, fontsize = 10)

# Save plot as PDF
pdf("../output/Molecular/interactions/TP0_v_TP7_amb_comparisons_sigcorr_subset.pdf", height = 20, width = 25)
mirRnaHeatmapDiff(combined_df_subset, fontsize = 25)
dev.off() # Close the PDF device
```

Plot FCs
```{r}
ggplot(combined_df, aes(x = mrna_FC, y = mirna_FC)) +
  geom_point(size = 6) +
  geom_smooth(method = "lm", se = T) # Add regression line

# Better plot
plot2 <- ggplot(combined_df_subset, aes(x = mrna_FC, y = mirna_FC)) +
  geom_point(size = 4, shape = 21, color = "black", fill = "black") +  # Customize points
  #geom_vline(xintercept = 0, linetype = "solid", color = "black", size = 0.5) +  # Add vertical line at x = 0
  #geom_hline(yintercept = 0, linetype = "solid", color = "black", size = 0.5) +  # Add horizontal line at y = 0
  geom_smooth(method = "lm", se = TRUE, color = "black", linetype = "dashed", size = 1) +  # Customize regression line
  labs(
    x = "mRNA Fold Change",
    y = "miRNA Fold Change"
  ) +  # Add titles and labels
  theme_minimal() +  # Use a minimal theme for a clean look
  theme(
    plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),  # Center and style title
    axis.title = element_text(size = 14, face = "bold"),  # Style axis titles
    axis.text = element_text(size = 12)  # Style axis text
    #panel.grid.major = element_line(color = "grey80")  # Customize grid lines
    #panel.grid.minor = element_blank()  # Remove minor grid lines
  ); plot2

ggsave("../output/Molecular/interactions/TP0_amb_v_TP7_amb_FC_plot.pdf", plot2, width = 10, height = 8)
ggsave("../output/Molecular/interactions/TP0_amb_v_TP7_amb_FC_plot.png", plot2, width = 10, height = 8)
```



Read in annotation information (annotation from Kate)
```{r}
annot <- read.delim("/Users/jillashey/Desktop/PutnamLab/Astrangia_Genome/Apoculata_v2.0_GeneAnnotation_combined_prelim.txt", header = T)

head(annot)

# Swap model out for TU in gene ids 
annot$Protein_ID <- gsub("model", "TU", annot$Protein_ID)
```

Remove extra info from NCBI gene descriptions
```{r}
annot <- annot %>%
  mutate(NCBI_Description = str_replace(NCBI_Description, "\\s*\\[.*\\]", "")) %>%
  mutate(NCBI_Description = str_replace(NCBI_Description, "LOC\\S+", "")) %>%
  mutate(NCBI_Description = str_replace(NCBI_Description, "PREDICTED:\\s*", "")) %>%
  mutate(NCBI_Description = str_trim(NCBI_Description))  # Trim any leading or trailing whitespace
```

Look at similarities between the two results
```{r}
blah <- read.csv("../output/Molecular/interactions/TP0_v_7_amb_miRNA_mRNA_correlations.csv")

yay <- read.csv("../output/Molecular/interactions/TP0_v_5_amb_miRNA_mRNA_correlations.csv")
```

Join dfs? 
```{r}
all <- yay %>%
  inner_join(blah, by = "mrna") %>%
  dplyr::select(mirna_FC.x, mrna_FC.x, mirna.x, mrna, pvalue.x, mirna_FC.y, mrna_FC.y, mirna.y, pvalue.y)

length(unique(all$mrna))
unique(all$mrna)

poo <- yay %>%
  inner_join(blah, by = "mirna") %>%
  dplyr::select(mirna_FC.x, mrna_FC.x, mirna, mrna.x, pvalue.x, mirna_FC.y, mrna_FC.y, mrna.y, pvalue.y)

length(unique(poo$mirna))
unique(poo$mirna)
```

```{r}
blah$comparison <- "TP0_v_TP7_amb"
yay$comparison <- "TP0_v_TP5_amb"

all <- rbind(blah, yay)
length(unique(all$mirna))
unique(all$mirna)

# Remove high mRNA FC and Ser gene 
all_subset <- all %>%
  dplyr::filter(value >= -15 & value <= 15) %>%
  dplyr::filter(!mrna == "Ser_4434")

length(unique(all_subset$mirna))
unique(all_subset$mirna)
length(unique(all_subset$mrna))
```

Save all_subset as csv
```{r}
#write.csv(all_subset, file = "../output/Molecular/interactions/sigcorr_amb_all.csv")
```

Create a new column called "pattern" and use conditional statements to determine its values
```{r}
all_subset$pattern <- ifelse(all_subset$mirna_FC > 0 & all_subset$mrna_FC < 0, "negative regulation",
                     ifelse(all_subset$mirna_FC < 0 & all_subset$mrna_FC > 0, "inverse regulation",
                            ifelse(all_subset$mirna_FC > 0 & all_subset$mrna_FC > 0, "coregulation increase",
                                   ifelse(all_subset$mirna_FC < 0 & all_subset$mrna_FC < 0, "coregulation decrease", NA))))
```

Subset negative regulation and TP0 amb v TP5 amb
```{r}
neg_reg <- all_subset %>%
  dplyr::filter(pattern == "negative regulation") %>%
  dplyr::filter(comparison == "TP0_v_TP5_amb")

unique(neg_reg$mirna)
```

Join neg reg with annotation info 
```{r}
neg_reg <- neg_reg %>%
  inner_join(annot, by = c("mrna" = "Protein_ID"))
```

Plot
```{r}
plot1 <- ggplot(neg_reg, aes(x=NCBI_Description, y=mirna, color=pvalue, size=value)) +
  geom_point() +
  scale_color_viridis(option="D", name="p-value") +
    scale_size_continuous(name="Value", range=c(15, 25)) +  # Adjust the range to make points larger
  labs(x="Function", y="miRNA", color="p-value", size="Value") +
  theme_classic() +
  theme(
    legend.title = element_text(size=40, face="bold"),
    legend.text = element_text(size=38),
    legend.position = "right",
    legend.key.size = unit(3, "lines"),
    axis.title = element_text(size = 45, face = "bold"),
    axis.text.x = element_text(size=41, color="black", angle=45, hjust=1),
    axis.text.y = element_text(size=41, color="black"),
    axis.line = element_line(linewidth = 2, color = "black"),
    panel.grid.major = element_line(linewidth = 1.25, color = "black")  # Increase major grid line thickness
  ); plot1

ggsave("../output/Molecular/interactions/sigcorr_downreg_TP0vTP5amb.pdf", plot1, width = 30, height = 18)
ggsave("../output/Molecular/interactions/sigcorr_downreg_TP0vTP5amb.png", plot1, width = 30, height = 18)
```

Subset negative regulation and TP0 amb v TP7 amb
```{r}
neg_reg <- all_subset %>%
  dplyr::filter(pattern == "negative regulation") %>%
  dplyr::filter(comparison == "TP0_v_TP7_amb")

unique(neg_reg$mirna)
```

Join neg reg with blast 
```{r}
neg_reg <- neg_reg %>%
  inner_join(annot, by = c("mrna" = "Protein_ID"))
```

Plot
```{r}
plot2 <- ggplot(neg_reg, aes(x=NCBI_Description, y=mirna, color=pvalue, size=value)) +
  geom_point() +
  scale_color_viridis(option="D", name="p-value") +
    scale_size_continuous(name="Value", range=c(15, 25)) +  # Adjust the range to make points larger
  labs(x="Function", y="miRNA", color="p-value", size="Value") +
  theme_classic() +
  theme(
    legend.title = element_text(size=40, face="bold"),
    legend.text = element_text(size=38),
    legend.position = "right",
    legend.key.size = unit(3, "lines"),
    axis.title = element_text(size = 45, face = "bold"),
    axis.text.x = element_text(size=41, color="black", angle=45, hjust=1),
    axis.text.y = element_text(size=41, color="black"),
    axis.line = element_line(linewidth = 2, color = "black"),
    panel.grid.major = element_line(linewidth = 1.25, color = "black")  # Increase major grid line thickness
  ); plot2

ggsave("../output/Molecular/interactions/sigcorr_downreg_TP0vTP7amb.pdf", plot2, width = 30, height = 18)
ggsave("../output/Molecular/interactions/sigcorr_downreg_TP0vTP7amb.png", plot2, width = 30, height = 18)
```

Subset inverse regulation and TP0 amb v TP5 amb
```{r}
inv_reg <- all_subset %>%
  dplyr::filter(pattern == "inverse regulation") %>%
  dplyr::filter(comparison == "TP0_v_TP5_amb")
unique(inv_reg$mirna)
```

Join inverse reg with blast 
```{r}
inv_reg <- inv_reg %>%
  inner_join(blast, by = "mrna")
```

Plot
```{r}
plot3 <- ggplot(inv_reg, aes(x=function., y=mirna, color=pvalue, size=value)) +
  geom_point() +
  scale_color_viridis(option="D", name="p-value") +
    scale_size_continuous(name="Value", range=c(15, 25)) +  # Adjust the range to make points larger
  labs(x="Function", y="miRNA", color="p-value", size="Value") +
  theme_classic() +
  theme(
    legend.title = element_text(size=40, face="bold"),
    legend.text = element_text(size=38),
    legend.position = "right",
    legend.key.size = unit(3, "lines"),
    axis.title = element_text(size = 45, face = "bold"),
    axis.text.x = element_text(size=41, color="black", angle=45, hjust=1),
    axis.text.y = element_text(size=41, color="black"),
    axis.line = element_line(linewidth = 2, color = "black"),
    panel.grid.major = element_line(linewidth = 1.25, color = "black")  # Increase major grid line thickness
  ); plot3

ggsave("../output/Molecular/interactions/sigcorr_upreg_TP0vTP5amb.pdf", plot3, width = 30, height = 25)
ggsave("../output/Molecular/interactions/sigcorr_upreg_TP0vTP5amb.png", plot3, width = 30, height = 25)
```

Subset inverse regulation and TP0 amb v TP7 amb
```{r}
inv_reg <- all_subset %>%
  dplyr::filter(pattern == "inverse regulation") %>%
  dplyr::filter(comparison == "TP0_v_TP7_amb")

unique(inv_reg$mirna)
```

Join inverse reg with blast 
```{r}
inv_reg <- inv_reg %>%
  inner_join(blast, by = "mrna")
```

Plot
```{r}
plot4 <- ggplot(inv_reg, aes(x=function., y=mirna, color=pvalue, size=value)) +
  geom_point() +
  scale_color_viridis(option="D", name="p-value") +
    scale_size_continuous(name="Value", range=c(15, 25)) +  # Adjust the range to make points larger
  labs(x="Function", y="miRNA", color="p-value", size="Value") +
  theme_classic() +
  theme(
    legend.title = element_text(size=40, face="bold"),
    legend.text = element_text(size=38),
    legend.position = "right",
    legend.key.size = unit(3, "lines"),
    axis.title = element_text(size = 45, face = "bold"),
    axis.text.x = element_text(size=41, color="black", angle=45, hjust=1),
    axis.text.y = element_text(size=41, color="black"),
    axis.line = element_line(linewidth = 2, color = "black"),
    panel.grid.major = element_line(linewidth = 1.25, color = "black")  # Increase major grid line thickness
  ); plot4

ggsave("../output/Molecular/interactions/sigcorr_upreg_TP0vTP7amb.pdf", plot4, width = 30, height = 25)
ggsave("../output/Molecular/interactions/sigcorr_upreg_TP0vTP7amb.png", plot4, width = 30, height = 25)
```



Should we be more interested in the gene overlap or the miRNA overlap between comparisons? 

























Next comparison: TP0 amb v TP5 heat

Read in mRNA file of interest
```{r}
TP0_v_5_heat_df <- read.csv("../output/Molecular/mRNA/TP0_v_5_heat.csv") %>%
  dplyr::select(X, log2FoldChange, padj) %>%
  dplyr::rename(gene_id = X, pvalue = padj)
```

Get mRNAs with particular fold change and pvalues 
```{r}
mrna_files <- list(TP0_v_5_heat_df)
mrna_data <- one2OneRnaMiRNA(mrna_files, gene_colname = "gene_id", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

Read in miRNA file of interest
```{r}
TP0_v_5_heat_miRNA_df <- read.csv("../output/Molecular/smRNA/TP0_v_5_heat_miRNA.csv") %>%
  dplyr::select(X, log2FoldChange, padj) %>%
  dplyr::rename(gene_id = X, pvalue = padj)
```

Get miRNAs with particular fold change and pvalues 
```{r}
mirna_files <- list(TP0_v_5_heat_miRNA_df)
mirna_data <- one2OneRnaMiRNA(mirna_files, gene_colname = "gene_id", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

Produce a mRNA miRNA interrelation df 
```{r}
inter0 <- twoTimePoint(mrna_data, mirna_data)
```

This function is calculating the absolute value of the difference between the mRNA and miRNA fold changes for each potential miRNA mRNA combo. 

Make background distribution
```{r}
outs_background <- twoTimePointSamp(mrna_data, mirna_data, Shrounds = 100, Srounds = 1000)
```

Read in miranda data (doesn't need to be read in again if it was already read in above)
```{r}
#miranda_data <- read.csv("~/Desktop/PutnamLab/Astrangia/2020-2021/miranda/miranda_gene_ids_all.csv", header = T)
```

The vignette includes the following information for the example mouse miranda data:  
```{r}
#head(miRanda)
```

V1 is the miRNA id, V2 is the gene id, V3 is the score, V4 is energy in kcal/mol, V5 is the subject % identify, and V6 is the query % identity. Select these columns from my data, rearrange columns, and rename columns to V1-6. This will allow the following functions to properly read the miranda file. 
```{r}
# Select specific columns
# miranda_data_filt <- miranda_data %>%
#   dplyr::select(mirna, Score, Energy_kcal_mol, Subject_Identity, Query_Identity, gene_id)
# 
# # Remove any duplicate rows
# miranda_data_filt <- unique(miranda_data_filt)
# 
# # Rearrange columns
# miranda_data_filt <- miranda_data_filt[,c("mirna", "gene_id", "Score", "Energy_kcal_mol", "Subject_Identity", "Query_Identity")]
# 
# # Rename columns V1-6
# miranda_data_filt <- miranda_data_filt %>%    
#   rename(V1 = mirna, 
#          V2 = gene_id,
#          V3 = Score,
#          V4 = Energy_kcal_mol, 
#          V5 = Subject_Identity,
#          V6 = Query_Identity)
```

Intersect miranda data with significant relationships
```{r}
results_TP0_v_TP5_heat <- mirandaIntersectInter(sig_interRel, outs_background, mrna_data, mirna_data, miranda_data_filt)

head(results_TP0_v_TP5_heat$mirna)
head(results_TP0_v_TP5_heat$mrna)
```

There are lots of NAs in the data. I need to get rid of those before plotting. I usually combine and save the df after I plot but I will 

Combine results data into a single df 
```{r}
combined_df <- cbind(results_TP0_v_TP5_heat$mirna, results_TP0_v_TP5_heat$mrna, results_TP0_v_TP5_heat$corrs)

colnames(combined_df) <- c("mirna_FC", "mrna_FC", "V1", "V2", "value", "V3", "V4", "V5", "V6", "pvalue")

# Remove NAs
combined_df <- na.omit(combined_df)
```

I needed to label some columns V1, V2, etc in order for the plotting functions to understand the input. The functions are looking for columns named V1, V2, etc 

Make plots showing miRNA and mRNA FC relationships 
```{r}
# Create results file for plotting 
results_TP0_v_TP5_heat_df <- finInterResult(results_TP0_v_TP5_heat)

# Remove NAs

par(mar=c(4,4,2,1))
drawInterPlots(mrna_data, mirna_data, results_TP0_v_TP5_heat_df)
```

Make heatmap for significant relationships
```{r}
mirRnaHeatmapDiff(combined_df, fontsize = 10)

# Save plot as PDF
#pdf("../output/Molecular/interactions/TP0_v_TP5_amv_comparisons_sigcorr.pdf")
#mirRnaHeatmapDiff(results_TP0_v_TP5_amb$corrs, fontsize = 10)
#dev.off() # Close the PDF device

# Save plot as png
png("../output/Molecular/interactions/TP0_v_TP5_heat_comparisons_sigcorr.png")
mirRnaHeatmapDiff(combined_df, fontsize = 10)
dev.off() # Close the png device
```

Rename columns to save and save results
```{r}
colnames(combined_df) <- c("mirna_FC", "mrna_FC", "mirna", "mrna", "value", "score", "energy", "subj_identity", "query_identity", "pvalue")

length(unique(combined_df$mrna))
unique(combined_df$mrna)

write.csv(combined_df, file = "../output/Molecular/interactions/TP0_v_5_heat_miRNA_mRNA_correlations.csv")
```

When comparing this df to the `TP0_v_5_amb_miRNA_mRNA_correlations.csv`, the patterns are the same in terms of direction of regulation and miRNA to mRNA combo. 

Less differences between TP0 and high treatment corals? 

Next comparison: TP0 amb v TP7 heat

Read in mRNA file of interest
```{r}
TP0_v_7_heat_df <- read.csv("../output/Molecular/mRNA/TP0_v_7_heat.csv") %>%
  dplyr::select(X, log2FoldChange, padj) %>%
  dplyr::rename(gene_id = X, pvalue = padj)
```

Get mRNAs with particular fold change and pvalues 
```{r}
mrna_files <- list(TP0_v_7_heat_df)
mrna_data <- one2OneRnaMiRNA(mrna_files, gene_colname = "gene_id", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

Read in miRNA file of interest
```{r}
TP0_v_7_heat_miRNA_df <- read.csv("../output/Molecular/smRNA/TP0_v_7_heat_miRNA.csv") %>%
  dplyr::select(X, log2FoldChange, padj) %>%
  dplyr::rename(gene_id = X, pvalue = padj)
```

Get miRNAs with particular fold change and pvalues 
```{r}
mirna_files <- list(TP0_v_7_heat_miRNA_df)
mirna_data <- one2OneRnaMiRNA(mirna_files, gene_colname = "gene_id", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

Produce a mRNA miRNA interrelation df 
```{r}
inter0 <- twoTimePoint(mrna_data, mirna_data)
```

This function is calculating the absolute value of the difference between the mRNA and miRNA fold changes for each potential miRNA mRNA combo. 

Make background distribution
```{r}
outs_background <- twoTimePointSamp(mrna_data, mirna_data, Shrounds = 100, Srounds = 1000)
```

Read in miranda data (doesn't need to be read in again if it was already read in above)
```{r}
#miranda_data <- read.csv("~/Desktop/PutnamLab/Astrangia/2020-2021/miranda/miranda_gene_ids_all.csv", header = T)
```

The vignette includes the following information for the example mouse miranda data:  
```{r}
#head(miRanda)
```

V1 is the miRNA id, V2 is the gene id, V3 is the score, V4 is energy in kcal/mol, V5 is the subject % identify, and V6 is the query % identity. Select these columns from my data, rearrange columns, and rename columns to V1-6. This will allow the following functions to properly read the miranda file. 
```{r}
# Select specific columns
# miranda_data_filt <- miranda_data %>%
#   dplyr::select(mirna, Score, Energy_kcal_mol, Subject_Identity, Query_Identity, gene_id)
# 
# # Remove any duplicate rows
# miranda_data_filt <- unique(miranda_data_filt)
# 
# # Rearrange columns
# miranda_data_filt <- miranda_data_filt[,c("mirna", "gene_id", "Score", "Energy_kcal_mol", "Subject_Identity", "Query_Identity")]
# 
# # Rename columns V1-6
# miranda_data_filt <- miranda_data_filt %>%    
#   rename(V1 = mirna, 
#          V2 = gene_id,
#          V3 = Score,
#          V4 = Energy_kcal_mol, 
#          V5 = Subject_Identity,
#          V6 = Query_Identity)
```

Intersect miranda data with significant relationships
```{r}
results_TP0_v_TP7_heat <- mirandaIntersectInter(sig_interRel, outs_background, mrna_data, mirna_data, miranda_data_filt)

head(results_TP0_v_TP7_heat$mirna)
head(results_TP0_v_TP7_heat$mrna)
```

There are lots of NAs in the data. I need to get rid of those before plotting. I usually combine and save the df after I plot but I will 

Combine results data into a single df 
```{r}
combined_df <- cbind(results_TP0_v_TP7_heat$mirna, results_TP0_v_TP7_heat$mrna, results_TP0_v_TP7_heat$corrs)

colnames(combined_df) <- c("mirna_FC", "mrna_FC", "V1", "V2", "value", "V3", "V4", "V5", "V6", "pvalue")

# Remove NAs
combined_df <- na.omit(combined_df)
```

I needed to label some columns V1, V2, etc in order for the plotting functions to understand the input. The functions are looking for columns named V1, V2, etc 

Make plots showing miRNA and mRNA FC relationships 
```{r}
# Create results file for plotting 
results_TP0_v_TP7_heat_df <- finInterResult(results_TP0_v_TP7_heat)

# Remove NAs

par(mar=c(4,4,2,1))
drawInterPlots(mrna_data, mirna_data, results_TP0_v_TP7_heat_df)
```

Make heatmap for significant relationships
```{r}
mirRnaHeatmapDiff(combined_df, fontsize = 10)

# Save plot as PDF
#pdf("../output/Molecular/interactions/TP0_v_TP5_amv_comparisons_sigcorr.pdf")
#mirRnaHeatmapDiff(results_TP0_v_TP5_amb$corrs, fontsize = 10)
#dev.off() # Close the PDF device

# Save plot as png
png("../output/Molecular/interactions/TP0_v_TP7_heat_comparisons_sigcorr.png")
mirRnaHeatmapDiff(combined_df, fontsize = 10)
dev.off() # Close the png device
```

Rename columns to save and save results
```{r}
colnames(combined_df) <- c("mirna_FC", "mrna_FC", "mirna", "mrna", "value", "score", "energy", "subj_identity", "query_identity", "pvalue")

length(unique(combined_df$mrna))
unique(combined_df$mrna)

write.csv(combined_df, file = "../output/Molecular/interactions/TP0_v_7_heat_miRNA_mRNA_correlations.csv")
```

When comparing this df to the `TP0_v_5_amb_miRNA_mRNA_correlations.csv`, the patterns are the same in terms of direction of regulation and miRNA to mRNA combo. 





### Functional analyses 

Now that I have the information about miRNA mRNA relationships, I can explore the functional implications of these relationships. 

Read in all data
```{r}
test1 <- read.csv("../output/Molecular/interactions/TP0_v_5_amb_miRNA_mRNA_correlations.csv")
length(unique(test1$mirna))
length(unique(test1$mrna))

test2 <- read.csv("../output/Molecular/interactions/TP0_v_7_amb_miRNA_mRNA_correlations.csv")
length(unique(test2$mirna))
length(unique(test2$mrna))

test3 <- read.csv("../output/Molecular/interactions/TP0_v_5_heat_miRNA_mRNA_correlations.csv")
length(unique(test3$mirna))
length(unique(test3$mrna))

test4 <- read.csv("../output/Molecular/interactions/TP0_v_7_heat_miRNA_mRNA_correlations.csv")
length(unique(test4$mirna))
length(unique(test4$mrna))
```

Extract column of interest (mRNAs)
```{r}
col1 <- test1$mrna
col2 <- test2$mrna
col3 <- test3$mrna
col4 <- test4$mrna
```

Bind together and save as csv 
```{r}
# Use `list` and then bind the rows
result_list <- list(col1, col2, col3, col4)

# To bind rows, you can convert them to a data frame first
result_df <- do.call(rbind, lapply(result_list, function(x) data.frame(value = x, stringsAsFactors = FALSE)))

# Remove duplicates
result_df <- unique(result_df)

# Save as csv 
write.csv(result_df, "../output/Molecular/interactions/all_sigcorr.csv")
write_delim(result_df, "../output/Molecular/interactions/all_sigcorr.txt")
```

I need to subset the genes of interest and blast them!

Identify unique miRNAs. First extract miRNAs from dfs
```{r}
col1 <- test1$mirna
col2 <- test2$mirna
col3 <- test3$mirna
col4 <- test4$mirna
```

Bind together and save as csv 
```{r}
# Use `list` and then bind the rows
result_list <- list(col1, col2, col3, col4)

# To bind rows, you can convert them to a data frame first
result_df <- do.call(rbind, lapply(result_list, function(x) data.frame(value = x, stringsAsFactors = FALSE)))

# Remove duplicates
result_df <- unique(result_df)

# Save as csv 
write.csv(result_df, "../output/Molecular/interactions/all_sigcorr.csv")
write_delim(result_df, "../output/Molecular/interactions/all_sigcorr.txt")
```


