---
title: "mirTarRnaSeq"
author: "Jill Ashey"
date: "2024-04-22"
output: html_document
---

This script will test out the R package [mirTarRnaSeq](https://bioconductor.org/packages/3.13/bioc/html/mirTarRnaSeq.html), which an be used for interactive mRNA miRNA sequencing statistical analysis. This package utilizes expression or differential expression mRNA and miRNA sequencing results and performs interactive correlation and various GLMs (Regular GLM, Multivariate GLM, and Interaction GLMs ) analysis between mRNA and miRNA expriments. These experiments can be time point experiments, and or condition expriments.

Load packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

BiocManager::install("mirTarRnaSeq")

library(tidyverse)
library(mirTarRnaSeq)
```

My goal is to identify miRNA mRNA correlations across timepoints and treatments. I'll be using part 2 of the [vignette](https://bioconductor.org/packages/3.13/bioc/vignettes/mirTarRnaSeq/inst/doc/mirTarRnaSeq.pdf) to test the package with my own data. This identifies miRNA-mRNA correlations across timepoints or treatments (>3 time points according the the vignette and paper). 

Load mRNA data. Not sure if it needs to be counts data? 
```{r}
mRNA_file <- read.csv("output/Molecular/mRNA/filtered_gene_counts.csv")

# Make gene counts the rownames 
#rownames(mRNA_file) <- mRNA_file[,1] #set first column that contains gene names as rownames
#mRNA_file <- mRNA_file[,-1] # remove column w/ gene names 

# Remove AST-1105, as it sequenced poorly 
mRNA_file <- mRNA_file %>%
  dplyr::select(-"AST.1105") %>%
  dplyr::rename(gene_id = X)
```

The `one2OneRnaMiRNA` wants a list of dataframe as opposed to just a single dataframe. I will separate out the columns into separate dfs by Condition (ie Temperature and Treatment). 

- TP0_Acclimation: "AST-1147", "AST-1567", "AST-1722", "AST-2398"
- TP5_Amb: "AST-1412", "AST-1617", "AST-2000", "AST-2360"
- TP5_Heat: "AST-2404", "AST-2412", "AST-2512", "AST-2563"
- TP7_Amb: "AST-1560", "AST-2302", "AST-2523"
- TP7_Heat: "AST-1065", "AST-2007", "AST-2729", "AST-2755"

Subset the dataframe into multiple dataframes 
```{r}
TP0_Acclimation <- mRNA_file[, c("gene_id", "AST.1147", "AST.1567", "AST.1722", "AST.2398")]
TP5_Amb <- mRNA_file[, c("gene_id", "AST.1412", "AST.1617", "AST.2000", "AST.2360")]
TP5_Heat <- mRNA_file[, c("gene_id", "AST.2404", "AST.2412", "AST.2512", "AST.2563")]
TP7_Amb <- mRNA_file[, c("gene_id", "AST.1560", "AST.2302", "AST.2523")]
TP7_Heat <- mRNA_file[, c("gene_id", "AST.1065", "AST.2007", "AST.2729", "AST.2755")]

list <- list(TP0_Acclimation, TP5_Amb, TP5_Heat, TP7_Amb, TP7_Heat)

lapply(list, str)

list[[1]]


test1 <- mRNA_file[, c("gene_id", "AST.1147")]
test2 <- mRNA_file[, c("gene_id", "AST.1412")]

test_list <- list(test1, test2)
dim(test_list)
```

Get mRNAs with particular fold change 
```{r}
mRNA <- one2OneRnaMiRNA(test_list[,2], gene_colname = "gene_id", pthreshold = 0.05)$fold
```

It seems like I need to give it matrix with DE fold change in it. Not sure how to do this since there are different fold changes for different comparisons. Load DE mRNA data
```{r}
TP0_amb_v_TP5_amb_DEG <- read.csv("output/Molecular/mRNA/TP0_amb_v_TP5_amb_DEG.csv")
TP0_amb_v_TP7_amb_DEG <- read.csv("output/Molecular/mRNA/TP0_amb_v_TP7_amb_DEG.csv")

test_list <- list(TP0_amb_v_TP5_amb_DEG, TP0_amb_v_TP7_amb_DEG)

mRNA <- one2OneRnaMiRNA(test_list, gene_colname = "X", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

This worked, but reduced to 282 genes. Maybe this is the overlap between the two comparisons? 

```{r}
blah <- inner_join(TP0_amb_v_TP5_amb_DEG, TP0_amb_v_TP7_amb_DEG, by = "X")
```

Confirmed, it is the number of genes that is shared between the two comparisons. 

Load DE miRNA data
```{r}
TP0_amb_v_TP5_amb_DEM <- read.csv("output/Molecular/smRNA/TP0_amb_v_TP5_amb_DEM.csv")
TP0_amb_v_TP7_amb_DEM <- read.csv("output/Molecular/smRNA/TP0_amb_v_TP7_amb_DEM.csv")

miRNA_list <- list(TP0_amb_v_TP5_amb_DEM, TP0_amb_v_TP7_amb_DEM)
```

Get all with particular fold change
```{r}
miRNA <- one2OneRnaMiRNA(miRNA_list, gene_colname = "X", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

Get mRNA-miRNA correlation
```{r}
test <- corMirnaRna(mRNA, miRNA, method = "spearman")
```

This only returned values of 1 and -1...Why? 

Make background distribution correlation
```{r}
outs <- sampCorRnaMirna(mRNA, miRNA, method="pearson",Shrounds = 100, Srounds = 1000)
```

Plot density. The grey is the background distribution and the red is the actual data. 
```{r}
mirRnaDensityCor(test, outs)
```

Something is weird...it doesn't look correct to have all of the correlations either be 1 or -1. 

Get correlations below certain threshold
```{r}
sig_corrs <- threshSig(test, outs,pvalue = 0.05)
```

As I suspected, when we try to get correlations below a particular threshold, we do not get any significant correlations because the values were either 1 or -1. 

Let's try part 3 of the [vignette](https://bioconductor.org/packages/3.13/bioc/vignettes/mirTarRnaSeq/inst/doc/mirTarRnaSeq.pdf). This identified significant mRNA-miRNA correlations for two time points. I'm going to go with TP0 vs TP5 ambient. 

The DEG (TP0_amb_v_TP5_amb_DEG) and DEM (TP0_amb_v_TP5_amb_DEM) data was already loaded above. Get fold changes above a certain threshold. 
```{r}
#TP0_amb_v_TP5_amb_DEG <- as.list(TP0_amb_v_TP5_amb_DEG)

mRNA <- one2OneRnaMiRNA(TP0_amb_v_TP5_amb_DEG, gene_colname = "X", fc_colname = "log2FoldChange", pthreshold = 0.05)$foldchanges
```

Giving me the undefined column error again. Try using code from this [analysis](https://github.com/yusri-dh/Keratinization_LUSC/blob/master/4_mirna_analysis.R), which used the same package starting on line 91. 
```{r}
my_mRNA_fc <- TP0_amb_v_TP5_amb_DEG %>%
  dplyr::select(log2FoldChange) %>%
  dplyr::rename(FC1 = log2FoldChange)

my_miRNA_fc <- TP0_amb_v_TP5_amb_DEM %>%
  dplyr::select(log2FoldChange) %>%
  dplyr::rename(FC1 = log2FoldChange)

test_inter0 <- twoTimePoint(my_mRNA_fc, my_miRNA_fc)

test_outs <- twoTimePointSamp(my_mRNA_fc, my_miRNA_fc)

# Identify correlations below a p-value threshold 
test_sig <- threshSigInter(test_inter0, test_outs)
```

But what is V1 and V2? Does this correspond to the row number for gene and miRNA?

Load miranda data 
```{r}
my_miranda <- read.delim("output/Molecular/interactions/miranda_de_strict_parsed.txt", header = F) %>% 
  dplyr::rename(mirna = V1,
        Target = V2, 
        Score = V3, 
        Energy_kcal_mol = V4, 
        Query_start_end = V5, 
        Subject_start_end = V6,
        Length = V7, 
        Subject_Identity = V8, 
        Query_Identity = V9) %>%
  dplyr::select(mirna, Target, Score, Energy_kcal_mol, Subject_Identity, Query_Identity)

my_miranda$mirna <- gsub(">", "", my_miranda$mirna)
```

The columns need to have the same names in order to intersect. The `test_sig` has V1 and V2 for the miRNA and mRNAs, respectively. V1 and V2 are just numbers, presumably corresponding to the row number of that gene id. Subset gene id names and add column that lists 1-XXX
```{r}
n <- nrow(TP0_amb_v_TP5_amb_DEG)
mRNA_id <- TP0_amb_v_TP5_amb_DEG %>%
  dplyr::select(X) %>%
  mutate(V2 = as.character(1:n)) %>%
  dplyr::rename(gene_id = X)

n <- nrow(TP0_amb_v_TP5_amb_DEM)
miRNA_id <- TP0_amb_v_TP5_amb_DEM %>%
  dplyr::select(X) %>%
  mutate(V1 = as.character(1:n)) %>%
  dplyr::rename(mirna = X)
```

Now join the id dataframes with `test_sig`. 
```{r}
test_sig_yay <- test_sig %>%
  inner_join(mRNA_id, by = "V2") %>%
  inner_join(miRNA_id, by = "V1") %>%
  select(mirna, gene_id, value)
```

In the `my_miranda` df, I need to change it so that Target corresponds to the gene id instead of the 3'UTR for a specific gene. To get the corresponding gene name for the 3'UTR, read in DEG 3'UTR text file.
```{r}
deg_3utr <- read.delim("output/Molecular/interactions/uniq_modified_closest_degs_3UTRid.txt", header = F) %>%
  dplyr::select(V1, V12) %>%
  dplyr::rename(UTR = V1, 
         gene_id = V12)

# Remove everything after the space in the UTR column
deg_3utr$UTR <- sub("\\s.*", "", deg_3utr$UTR)
```

Merge the miranda results with the deg 3UTR info.
```{r}
miranda_gene <- my_miranda %>%
  inner_join(deg_3utr, by = c("Target" = "UTR"))
```

It looks like the `mirandaIntersectInter` is trying to join things by V1 and V2. I will change those back in the dfs. 
```{r}
test_sig_yay <- test_sig_yay %>% 
  dplyr::rename(V1 = mirna, V2 = gene_id)

miranda_gene <- miranda_gene %>%
  dplyr::rename(V1 = mirna, V2 = gene_id)
```

Miranda intersection with correlation results 
```{r}
results <- mirandaIntersectInter(test_sig_yay, test_outs, my_mRNA_fc, my_miRNA_fc, miranda_gene)
```

Error says no common mRNA/miRNAs found...I may need to wait until the current miranda run finishes. 










Can I do it with their test data that is included in the package? Trying to two time point comparison using `mRNA_fc2` and `miRNA_fc2` for Part 3. These already have fold changes estimated
```{r}
dim(mRNA_fc2)
dim(miRNA_fc2)

# estimate differences based on FC
inter0_test <- twoTimePoint(mRNA_fc2, miRNA_fc2)
head(inter0_test)
head(inter0)
```

What is the value telling us here? The [manual](https://bioconductor.org/packages/3.13/bioc/manuals/mirTarRnaSeq/man/mirTarRnaSeq.pdf) says it provides a miRNA mRNA interrelation dataframe. Interrelation values are estimated by the absolute difference between fold change for miRNA and mRNA. All miRNAs are being compared against all mRNAs. 

Make background distribution
```{r}
outs_test <- twoTimePointSamp(mRNA_fc2, miRNA_fc2)
```

This function samples the mRNA and miRNA FCs by shuffling over the FC data (default is 100). Then it samples from the shuffled data (default is 1000) so that there is a predicted distribution based on our data. 

Import miranda data. This package has miranda data built in for a few species, but a user-created miranda file can also be used as input. 
```{r}
miRanda_test <- getInputSpecies("Mouse", threshold = 140)
```

Threshold is referring to the miranda score (the higher the score, the better the interaction). I could also set the energy score to cut off at a certain value (ie -10 kcal/mol).

Identify relationships below threshold (default is 0.05). 
```{r}
sig_InterR_test <- threshSigInter(inter0_test, outs_test)
```

This df represents the significant interactions. Annoying that they don't provide the p-value. The function is using the predicted distribution to determine an appropriate threshold for our data. This seems to correspond to a higher value, ie a higher absolute difference between miRNA and mRNA FC. 

Intersect miranda data with results. 
```{r}
results_test <- mirandaIntersectInter(sig_InterR_test, outs_test, mRNA_fc2, miRNA_fc2, miRanda_test)
head(results_test$corrs)
```

Yay! `results_test$corrs` has the information about the correlation. It includes the miRNA, mRNA, value, score, energy, subject identity, query identity, and p-value for the putative interaction. 

```{r}
#Create a results file for heatmap
final_results <- finInterResult(results_test)
#Draw plots of miRNA mRNA fold changes for your results file
par(mar=c(4,4,2,1))
drawInterPlots(mRNA_fc2, miRNA_fc2,final_results)
```

These plots are showing the distribution of FC for mRNA and miRNA, as well as the correlation between the mRNA and miRNAs that are interacting. 

Make heatmap of miRNA mRNA FC differences 
```{r}
CorRes <- results_test$corrs
mirRnaHeatmap(CorRes)
mirRnaHeatmapDiff(CorRes)
```

mRNA is on the x axis, miRNA is on the y-axis. It is plotting the value as the color (ie a higher absolute difference between miRNA and mRNA FC). What is this value telling us? Larger difference means larger values. We know that the colored blocks are correlated mRNA and miRNAs and that they can putatively bind. It's not telling us about the strength of the correlations. 






