---
title: "Amassing data"
author: "Jill Ashey"
date: "2025-01-29"
output: html_document
---

I am getting overwhelmed with the amount of data that I have in different places so I am making this script that amasses the mRNA miRNA data thus far. I will be making a df that contains:

- Expression information for all mRNAs for all comparisons
- Expression information for all miRNAs for all comparisons
- Miranda interaction data
- Annotation data 
- PCC data 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(viridis)
library(cowplot)
library(igraph)
library(tidygraph)
library(ggraph)
```

Read in mRNA expression information
```{r}
tp0_v_tp5_amb <- read.csv("../output/Molecular/mRNA/TP0_v_5_amb.csv")
colnames(tp0_v_tp5_amb)[colnames(tp0_v_tp5_amb) == "X"] <- "mRNA"
tp0_v_tp5_amb$comparison <- "TP0_v_TP5_amb"

tp0_v_tp7_amb <- read.csv("../output/Molecular/mRNA/TP0_v_7_amb.csv")
colnames(tp0_v_tp7_amb)[colnames(tp0_v_tp7_amb) == "X"] <- "mRNA"
tp0_v_tp7_amb$comparison <- "TP0_v_TP7_amb"

tp0_v_tp5_heat <- read.csv("../output/Molecular/mRNA/TP0_v_5_heat.csv")
colnames(tp0_v_tp5_heat)[colnames(tp0_v_tp5_heat) == "X"] <- "mRNA"
tp0_v_tp5_heat$comparison <- "TP0_v_TP5_heat"

tp0_v_tp7_heat <- read.csv("../output/Molecular/mRNA/TP0_v_7_heat.csv")
colnames(tp0_v_tp7_heat)[colnames(tp0_v_tp7_heat) == "X"] <- "mRNA"
tp0_v_tp7_heat$comparison <- "TP0_v_TP7_heat"
```

Bind all mRNA expression dfs and mark if the mRNA is differentially expressed 
```{r}
all_mrna <- rbind(tp0_v_tp5_amb, tp0_v_tp7_amb, tp0_v_tp5_heat, tp0_v_tp7_heat)

all_mrna <- all_mrna %>%
  mutate(DE = ifelse(padj < 0.05, "Y", "N"))
```

Read in miRNA expression information
```{r}
tp0_v_tp5_amb_mi <- read.csv("../output/Molecular/smRNA/shortstack/TP0_v_5_amb_miRNA.csv")
colnames(tp0_v_tp5_amb_mi)[colnames(tp0_v_tp5_amb_mi) == "X"] <- "miRNA"
tp0_v_tp5_amb_mi$comparison <- "TP0_v_TP5_amb"

tp0_v_tp7_amb_mi <- read.csv("../output/Molecular/smRNA/shortstack/TP0_v_7_amb_miRNA.csv")
colnames(tp0_v_tp7_amb_mi)[colnames(tp0_v_tp7_amb_mi) == "X"] <- "miRNA"
tp0_v_tp7_amb_mi$comparison <- "TP0_v_TP7_amb"

tp0_v_tp5_heat_mi <- read.csv("../output/Molecular/smRNA/shortstack/TP0_v_5_heat_miRNA.csv")
colnames(tp0_v_tp5_heat_mi)[colnames(tp0_v_tp5_heat_mi) == "X"] <- "miRNA"
tp0_v_tp5_heat_mi$comparison <- "TP0_v_TP5_heat"

tp0_v_tp7_heat_mi <- read.csv("../output/Molecular/smRNA/shortstack/TP0_v_7_heat_miRNA.csv")
colnames(tp0_v_tp7_heat_mi)[colnames(tp0_v_tp7_heat_mi) == "X"] <- "miRNA"
tp0_v_tp7_heat_mi$comparison <- "TP0_v_TP7_heat"
```

Bind all miRNA expression dfs, mark if the mRNA is differentially expressed, and format miRNA names
```{r}
all_mirna <- rbind(tp0_v_tp5_amb_mi, tp0_v_tp7_amb_mi, tp0_v_tp5_heat_mi, tp0_v_tp7_heat_mi)

all_mirna <- all_mirna %>%
  mutate(DE = ifelse(padj < 0.05, "Y", "N")) %>%
  mutate(miRNA = str_replace(miRNA, "^([^_]*_[^_]*)_.*$", "\\1"))
```

Read in df that contains pcc, miranda and annot data
```{r}
combined_data_pcc <- read.csv("../output/Molecular/interactions/miRNA_mRNA_interactions_annot.csv")

length(unique(combined_data_pcc$miRNA))
length(unique(combined_data_pcc$mRNA))
```

Plot using igraph
```{r}
# Create the graph
g <- graph_from_data_frame(combined_data_pcc, directed = FALSE)

# Add edge attributes
E(g)$weight <- abs(E(g)$PCC.cor)  # Use absolute PCC for edge weight
E(g)$color <- ifelse(E(g)$PCC.cor > 0, "blue", "red")  # Blue for positive, red for negative correlations

# Add node attributes
V(g)$type <- ifelse(V(g)$name %in% combined_data_pcc$miRNA, "miRNA", "mRNA")

# Convert to tbl_graph for ggraph
g_tbl <- as_tbl_graph(g)

# Create the plot
p <- ggraph(g_tbl, layout = "auto") +
  geom_edge_link(aes(edge_width = weight, color = color), alpha = 0.6) +
  geom_node_point(aes(color = type), size = 10) +
  #geom_node_text(aes(label = name), repel = TRUE) +
  #theme(aspect.ratio = 1) +
  #coord_fixed(expand = TRUE) +
  scale_edge_width(range = c(1, 5)) +
  scale_color_manual(values = c("miRNA" = "purple", "mRNA" = "orange", "Positive correlation" = "blue", "Negative correlation" = "red")) +
  theme_graph();p
ggsave("../output/Molecular/interactions/miRNA_mRNA_network.png", p, width = 35, height = 30, dpi = 300)
```

Assess data 
```{r}
sum(combined_data_pcc$p_value < 0.05)

blah <- combined_data_pcc %>%
  filter(p_value < 0.05)
length(unique(blah$miRNA))
length(unique(blah$mRNA))
```

Merge `combined_data_pcc` df with mRNA and miRNA dfs
```{r}
combined_data_pcc_mrna <- combined_data_pcc %>%
  left_join(all_mrna, by = "mRNA")

combined_data_pcc_mrna_mirna <- combined_data_pcc_mrna %>%
  left_join(all_mirna, by = "miRNA")
```

Rename and reorder columns 
```{r}
combined_data_pcc_mrna_mirna <- combined_data_pcc_mrna_mirna %>%
   dplyr::rename(
    mRNA_baseMean = baseMean.x,
    mRNA_log2FoldChange = log2FoldChange.x,
    mRNA_lfcSE = lfcSE.x,
    mRNA_stat = stat.x,
    mRNA_pvalue = pvalue.x,
    mRNA_padj = padj.x,
    #mRNA_comparison = comparison.x,
    mRNA_DE = DE.x,
    miRNA_baseMean = baseMean.y,
    miRNA_log2FoldChange = log2FoldChange.y,
    miRNA_lfcSE = lfcSE.y,
    miRNA_stat = stat.y,
    miRNA_pvalue = pvalue.y,
    miRNA_padj = padj.y,
    #miRNA_comparison = comparison.y,
    miRNA_DE = DE.y) %>%
  dplyr::select(miRNA, mRNA, PCC.cor, p_value, adjusted_p_value, mRNA_log2FoldChange, mRNA_padj, mRNA_DE, miRNA_log2FoldChange, miRNA_padj, miRNA_DE, NCBI_Description, Swiss.Prot_Description, score, energy, query_similar, subject_similar, total_bp_shared, GO_Swiss.Prot, GO_Trembl) %>%
  dplyr::filter(!is.na(miRNA_padj) & !is.na(mRNA_padj))

dim(combined_data_pcc_mrna_mirna)
combined_data_pcc_mrna_mirna <- unique(combined_data_pcc_mrna_mirna)
dim(combined_data_pcc_mrna_mirna)
```

Assess df 
```{r}
dim(combined_data_pcc_mrna_mirna)
length(unique(combined_data_pcc_mrna_mirna$miRNA))
length(unique(combined_data_pcc_mrna_mirna$mRNA))
```

Plot using igraph
```{r}
# Create the graph
g <- graph_from_data_frame(combined_data_pcc_mrna_mirna, directed = FALSE)

# Add edge attributes
E(g)$weight <- abs(E(g)$PCC.cor)  # Use absolute PCC for edge weight
E(g)$color <- ifelse(E(g)$PCC.cor > 0, "blue", "red")  # Blue for positive, red for negative correlations

# Add node attributes
V(g)$type <- ifelse(V(g)$name %in% combined_data_pcc_mrna_mirna$miRNA, "miRNA", "mRNA")

# Convert to tbl_graph for ggraph
g_tbl <- as_tbl_graph(g)

# Create the plot
p <- ggraph(g_tbl, layout = "fr") +
  geom_edge_link(aes(edge_width = weight, color = color), alpha = 0.6) +
  geom_node_point(aes(color = type), size = 10) +
  #geom_node_text(aes(label = name), repel = TRUE) +
  #theme(aspect.ratio = 1) +
  #coord_fixed(expand = TRUE) +
  scale_edge_width(range = c(1, 5)) +
  scale_color_manual(values = c("miRNA" = "purple", "mRNA" = "orange", "Positive correlation" = "blue", "Negative correlation" = "red")) +
  theme_graph();p
ggsave("../output/Molecular/interactions/miRNA_mRNA_network_expressed.png", p, width = 35, height = 30, dpi = 300)

# components <- decompose(g_tbl)
# 
# # Plot each component
# lapply(components, function(comp) {
#   ggraph(comp, layout = "fr") +
#     geom_edge_link(aes(width = weight, color = color), alpha = 0.6) +
#     geom_node_point(aes(color = type), size = 4) +
#     theme_graph()
# })
```

Barplot of targets 
```{r}
# Summarize the data with unique mRNA counts
summary_data <- combined_data_pcc_mrna_mirna %>%
  group_by(miRNA) %>%
  summarise(
    positive_count = n_distinct(mRNA[PCC.cor > 0]),
    negative_count = n_distinct(mRNA[PCC.cor < 0])
  ) %>%
  mutate(total_count = positive_count + negative_count) %>%
  arrange(desc(total_count))

# Reshape the data for plotting
plot_data <- summary_data %>%
  pivot_longer(cols = c(positive_count, negative_count),
               names_to = "correlation_type",
               values_to = "count")

# Create the stacked bar plot
mirna_num <- ggplot(plot_data, aes(x = reorder(miRNA, -total_count), y = count, fill = correlation_type)) +
  geom_bar(stat = "identity", width = 0.8) +
  scale_fill_manual(values = c("positive_count" = "#CC3361", "negative_count" = "#33CC9E"),
                    labels = c("Positive", "Negative")) +
  theme_classic(base_size = 25) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.title = element_text(size = 18, face = "bold"),
    axis.text = element_text(color = "black"),
    legend.title = element_text(size = 25, face = "bold"),
    legend.text = element_text(size = 18),
    legend.position = "top",
    legend.key.size = unit(1.5, "cm"),
    legend.spacing.y = unit(0.5, "cm"),
    plot.margin = margin(t = 20, r = 20, b = 20, l = 20, unit = "pt")
  ) +
  labs(x = "miRNAs", y = "Number of mRNA Targets", fill = "Correlation Type") +
  scale_y_continuous(expand = c(0, 0)); mirna_num

ggsave("../output/Molecular/interactions/miRNA_num_targets.png", mirna_num, width = 18, height = 10, dpi = 300)
ggsave("../output/Molecular/interactions/miRNA_num_targets.pdf", mirna_num, width = 18, height = 10, dpi = 300)
```

Barplot of the number of miRNAs that target each mRNA 
```{r}
# Summarize the data with unique mRNA counts
summary_data <- combined_data_pcc_mrna_mirna %>%
  group_by(mRNA) %>%
  summarise(
    positive_count = n_distinct(mRNA[PCC.cor > 0]),
    negative_count = n_distinct(mRNA[PCC.cor < 0])
  ) %>%
  mutate(total_count = positive_count + negative_count) %>%
  arrange(desc(total_count))

# Reshape the data for plotting
plot_data <- summary_data %>%
  pivot_longer(cols = c(positive_count, negative_count),
               names_to = "correlation_type",
               values_to = "count")

# Create the stacked bar plot
mrna_num <- ggplot(plot_data, aes(x = reorder(mRNA, -total_count), y = count, fill = correlation_type)) +
  geom_bar(stat = "identity", width = 0.8) +
  scale_fill_manual(values = c("positive_count" = "#CC3361", "negative_count" = "#33CC9E"),
                    labels = c("Positive", "Negative")) +
  scale_y_continuous(breaks = c(1, 2), limits = c(0, 2), expand = c(0, 0)) +  # Remove padding at bottom
  theme_classic(base_size = 25) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.title = element_text(size = 18, face = "bold"),
    axis.text = element_text(color = "black"),
    legend.title = element_text(size = 25, face = "bold"),
    legend.text = element_text(size = 18),
    legend.position = "top",
    legend.key.size = unit(1.5, "cm"),
    legend.spacing.y = unit(0.5, "cm"),
    plot.margin = margin(t = 20, r = 20, b = 20, l = 20, unit = "pt")
  ) +
  labs(x = "mRNAs", y = "Number of Targeting miRNAs", fill = "Correlation Type"); mrna_num

ggsave("../output/Molecular/interactions/mRNA_num_targeting.png", mrna_num, width = 18, height = 10, dpi = 300)
ggsave("../output/Molecular/interactions/mRNA_num_targeting.pdf", mrna_num, width = 18, height = 10, dpi = 300)
```

Select only differentially expressed 
```{r}
combined_data_pcc_mrna_mirna_de <- combined_data_pcc_mrna_mirna %>%
  filter(miRNA_DE == "Y") %>%
  filter(mRNA_DE == "Y") #%>%
  #filter(p_value < 0.05)

length(unique(combined_data_pcc_mrna_mirna_de$miRNA))
length(unique(combined_data_pcc_mrna_mirna_de$mRNA))
# Save as csv 
#write.csv(combined_data_pcc_mrna_mirna_de, "../output/Molecular/interactions/interactions_annot_pcc_de.csv")
```

Plot using igraph
```{r}
# Create the graph
g <- graph_from_data_frame(combined_data_pcc_mrna_mirna_de, directed = FALSE)

# Add edge attributes
E(g)$weight <- abs(E(g)$PCC.cor)  # Use absolute PCC for edge weight
E(g)$color <- ifelse(E(g)$PCC.cor > 0, "blue", "red")  # Blue for positive, red for negative correlations

# Add node attributes
V(g)$type <- ifelse(V(g)$name %in% combined_data_pcc_mrna_mirna_de$miRNA, "miRNA", "mRNA")

# Convert to tbl_graph for ggraph
g_tbl <- as_tbl_graph(g)

# Create the plot
p <- ggraph(g_tbl, layout = "fr") +
  geom_edge_link(aes(edge_width = weight, color = color), alpha = 0.6) +
  geom_node_point(aes(color = type), size = 5) +
  #geom_node_text(aes(label = name), repel = TRUE, size = 3) +
  scale_edge_width(range = c(0.5, 3)) +
  scale_color_manual(values = c("miRNA" = "purple", "mRNA" = "orange", "Positive correlation" = "blue", "Negative correlation" = "red")) +
  theme_graph();p
ggsave("../output/Molecular/interactions/miRNA_mRNA_network_de.png", p, width = 15, height = 10, dpi = 300)
```

Plot!
```{r}
# Specify direction of correlation
combined_data_pcc_mrna_mirna_de <- combined_data_pcc_mrna_mirna_de %>%
  mutate(direction = ifelse(PCC.cor > 0, "Positive", "Negative")) %>%
  mutate(Swiss.Prot_Description_short = str_remove(Swiss.Prot_Description, "OS=.*$"))

plot<-ggplot(combined_data_pcc_mrna_mirna_de, aes(x = Swiss.Prot_Description_short, y = p_value, fill = direction)) +
  #ylim(0, 1) +
  #geom_hline(yintercept = 0.05, linetype = "solid", color = "black", linewidth = 1)+
  geom_hline(yintercept = 0.05, color = "black", linetype = "solid", linewidth = 1) +  # Add line at 0.05 
  geom_point(shape = 21, size = 10) + 
  #scale_size(range = c(2, 20)) + 
  xlab('') + 
  ylab("PCC p-value") +
  theme_bw(base_size = 30) +
  #facet_grid(vars(ontology), scales = "free", space = "free_y") +
  coord_flip(); plot

# Save plot 
ggsave("../output/Molecular/interactions/swissprot_pcc_de_target.pdf", plot, width = 40, height = 25, dpi = 300)
ggsave("../output/Molecular/interactions/swissprot_pcc_de_target.png", plot, width = 40, height = 25, dpi = 300)
```

Make plots for paper. plot of pos and neg correlations from just PCC+ miranda data → DEs, include the plots of pvalue v PCC. so there will be 4 figures in one plot 
```{r}
### PCC + miranda 
# Summarize data
summary_data <- combined_data_pcc %>%
  group_by(miRNA) %>%
  summarise(
    positive_count = sum(PCC.cor > 0),
    negative_count = sum(PCC.cor < 0)
  ) %>%
  mutate(total_count = positive_count + negative_count) %>%
  arrange(desc(total_count))

# Reshape the data for plotting
plot_data <- summary_data %>%
  tidyr::pivot_longer(cols = c(positive_count, negative_count),
                      names_to = "correlation_type",
                      values_to = "count")

# Plot number of correlations for PCC + miranda
plot_a <- ggplot(plot_data, aes(x = reorder(miRNA, -total_count), y = count, fill = correlation_type)) +
  geom_bar(stat = "identity", color = "black") +
  scale_fill_viridis(discrete = TRUE, option = "D") +  # Color-blind friendly palette
  theme_minimal(base_size = 25) +  # Increase base font size
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 25, colour = "black"),
    axis.text.y = element_text(size = 25, colour = "black"),
    axis.title = element_text(size = 27, face = "bold", colour = "black"),
    legend.title = element_text(size = 18, face = "bold", colour = "black"),
    legend.text = element_text(size = 23, colour = "black"),
    legend.position = c(1, 1),
    legend.justification = c(1, 1),
    axis.line.x = element_line(color = "black", size = 0.5),
    axis.ticks.x = element_line(color = "black", size = 0.5),
    axis.ticks.length.x = unit(0.25, "cm"),
    plot.margin = margin(b = 20, unit = "pt"),  # Add bottom margin
    axis.title.x = element_text(margin = margin(t = 15, unit = "pt"))  # Move x-axis title down
  ) +
  labs(
    x = "miRNA",
    y = "Count",
    fill = "Correlation Type"
  ) +
  scale_x_discrete(expand = c(0, 0)) +  # Remove padding at the ends of x-axis
  scale_y_continuous(expand = c(0, 0)); plot_a  # Remove padding at the bottom of y-axis

# Plot pvalue v PCC 
plot_b <- ggplot(combined_data_pcc, aes(x = PCC.cor, y = p_value, color = p_value < 0.05)) +
  geom_jitter(alpha = 0.7, width = 0.05, height = 0, size =3.5) +  # Increased point size and alpha
  scale_color_manual(values = c("FALSE" = "grey50", "TRUE" = "red")) +  # Colorblind-friendly palette
  geom_hline(yintercept = 0.05, linetype = "dashed", color = "black", size = 1) +  # Thicker line
  theme_minimal(base_size = 25) +  # Increase base font size
  theme(
    axis.text = element_text(size = 25, colour = "black"),
    axis.title = element_text(size = 27, face = "bold", colour = "black"),
    legend.title = element_text(size = 25, face = "bold", colour = "black"),
    legend.text = element_text(size = 23, colour = "black"),
    legend.position = c(1, 1),
    legend.justification = c(1, 1),
    axis.line = element_line(color = "black", size = 0.5),
    axis.ticks = element_line(color = "black", size = 0.5),
    axis.ticks.length = unit(0.25, "cm"),
    plot.margin = margin(b = 20, l = 20, unit = "pt")  # Add bottom and left margin
  ) +
  labs(
    x = "Pearson Correlation Coefficient",
    y = "P-value",
    color = "P-value < 0.05"
  ); plot_b

### PCC + miranda AND DEs
# Summarize data
summary_data <- combined_data_pcc_mrna_mirna_de %>%
  group_by(miRNA) %>%
  summarise(
    positive_count = sum(PCC.cor > 0),
    negative_count = sum(PCC.cor < 0)
  ) %>%
  mutate(total_count = positive_count + negative_count) %>%
  arrange(desc(total_count))

# Reshape the data for plotting
plot_data <- summary_data %>%
  tidyr::pivot_longer(cols = c(positive_count, negative_count),
                      names_to = "correlation_type",
                      values_to = "count")

# Plot number of correlations for PCC + miranda AND sig corrs
plot_c <- ggplot(plot_data, aes(x = reorder(miRNA, -total_count), y = count, fill = correlation_type)) +
  geom_bar(stat = "identity", color = "black") +
  scale_fill_viridis(discrete = TRUE, option = "D") +  # Color-blind friendly palette
  theme_minimal(base_size = 25) +  # Increase base font size
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 25, colour = "black"),
    axis.text.y = element_text(size = 25, colour = "black"),
    axis.title = element_text(size = 27, face = "bold", colour = "black"),
    legend.title = element_text(size = 18, face = "bold", colour = "black"),
    legend.text = element_text(size = 23, colour = "black"),
    legend.position = c(1, 1),
    legend.justification = c(1, 1),
    axis.line.x = element_line(color = "black", size = 0.5),
    axis.ticks.x = element_line(color = "black", size = 0.5),
    axis.ticks.length.x = unit(0.25, "cm"),
    plot.margin = margin(b = 20, unit = "pt"),  # Add bottom margin
    axis.title.x = element_text(margin = margin(t = 15, unit = "pt"))  # Move x-axis title down
  ) +
  labs(
    x = "miRNA",
    y = "Count",
    fill = "Correlation Type"
  ) +
  scale_x_discrete(expand = c(0, 0)) +  # Remove padding at the ends of x-axis
  scale_y_continuous(expand = c(0, 0)); plot_c  # Remove padding at the bottom of y-axis

# Plot pvalue v PCC 
plot_d <- ggplot(combined_data_pcc_mrna_mirna_de, aes(x = PCC.cor, y = p_value, color = p_value < 0.05)) +
  geom_jitter(alpha = 0.7, width = 0.05, height = 0, size = 3.5) +  # Increased point size and alpha
  scale_color_manual(values = c("FALSE" = "grey50", "TRUE" = "red")) +  # Colorblind-friendly palette
  geom_hline(yintercept = 0.05, linetype = "dashed", color = "black", size = 1) +  # Thicker line
  theme_minimal(base_size = 25) +  # Increase base font size
  theme(
    axis.text = element_text(size = 25, colour = "black"),
    axis.title = element_text(size = 27, face = "bold", colour = "black"),
    legend.title = element_text(size = 25, face = "bold", colour = "black"),
    legend.text = element_text(size = 23, colour = "black"),
    legend.position = c(1, 1),
    legend.justification = c(1, 1),
    axis.line = element_line(color = "black", size = 0.5),
    axis.ticks = element_line(color = "black", size = 0.5),
    axis.ticks.length = unit(0.25, "cm"),
    plot.margin = margin(b = 20, l = 20, unit = "pt")  # Add bottom and left margin
  ) +
  labs(
    x = "Pearson Correlation Coefficient",
    y = "P-value",
    color = "P-value < 0.05"
  ); plot_d

# Put plots together
combined_plot <- plot_grid(plot_a, plot_b, plot_c, plot_d, 
          labels = c("A", "B", "C", "D"),
          label_size = 30,
          ncol = 2, 
          align = "hv")

# Save plot 
ggsave("../output/Molecular/interactions/combined_plot_pcc_target.pdf", combined_plot, width = 40, height = 30, dpi = 300)
ggsave("../output/Molecular/interactions/combined_plot_pcc_target.png", combined_plot, width = 40, height = 30, dpi = 300)
```









## Examine miRNAs of interest 

Cluster_4640
```{r}
clust_4640 <- combined_data_pcc_mrna_mirna %>%
  filter(miRNA == "Cluster_4640") %>%
  #filter(PCC.cor < 0) %>%
  filter(mRNA_log2FoldChange > 0)
unique(clust_4640$NCBI_Description)
```

Cluster_2003
```{r}
clust_2003 <- combined_data_pcc_mrna_mirna %>%
  filter(miRNA == "Cluster_2003") %>%
  filter(PCC.cor < 0) %>%
  filter(mRNA_log2FoldChange > 0) %>%
  filter(miRNA_log2FoldChange < 0)
unique(clust_2003$NCBI_Description)
```

Cluster_1453
```{r}
clust_1453 <- combined_data_pcc_mrna_mirna %>%
  filter(miRNA == "Cluster_1453") %>%
  filter(PCC.cor < 0) %>%
  filter(mRNA_log2FoldChange > 0) %>%
  filter(miRNA_log2FoldChange < 0)
unique(clust_1453$NCBI_Description)
```

Cluster_3037
```{r}
clust_3037 <- combined_data_pcc_mrna_mirna %>%
  filter(miRNA == "Cluster_3037") %>%
  filter(PCC.cor < 0) %>%
  filter(mRNA_log2FoldChange > 0) %>%
  filter(miRNA_log2FoldChange < 0)
unique(clust_3037$NCBI_Description)
```

## Test the hypothesis that a miRNA can upregulate a gene by inhibiting its upstream suppressor

To do this for a significant positive mRNA-miRNA pair, assess if there are any mRNAs that are negatively correlated with both the miRNA and mRNA expression. 

Filter the df so that only the significant positive DE mRNA-miRNA pairs remain 
```{r}
combined_data_pcc_mrna_mirna_sig_pos <- combined_data_pcc_mrna_mirna %>%
    filter(PCC.cor > 0)# %>%
  # filter(p_value < 0.05) %>%
  # filter(mRNA_DE == "Y") %>%
  # filter(miRNA_DE == "Y")
unique(combined_data_pcc_mrna_mirna_sig_pos$miRNA)
unique(combined_data_pcc_mrna_mirna_sig_pos$mRNA)
```

Extract relevant mRNAs
```{r}
mRNAs_list1 <- unique(combined_data_pcc_mrna_mirna_sig_pos$mRNA) # miRNAs with significant positive 
mRNAs_list2 <- unique(combined_data_pcc_mrna_mirna$mRNA) # all miRNAs interacting with miRNAs and has a correlation value 
```

Read in mRNA count data and normalize 
```{r}
mrna_counts <- read.csv("../output/Molecular/mRNA/filtered_gene_counts.csv")

# Set row names
rownames(mrna_counts) <- mrna_counts[,1] #set first column that contains gene names as rownames
mrna_counts <- mrna_counts[,-1] # remove column w/ gene names 

# Remove AST-1105
mrna_counts <- mrna_counts %>%
  select(-AST.1105)
```

Subset the counts data 
```{r}
# Subset the counts matrix
mrna_counts_subset <- mrna_counts[c(mRNAs_list1, mRNAs_list2), ]
```

Normalize mRNA counts
```{r}
# Normalize 
## Function to normalize counts (simple RPM normalization)
normalize_counts <- function(counts) {
  rpm <- t(t(counts) / colSums(counts)) * 1e6
  return(rpm)
}

mrna_counts_subset_norm <- normalize_counts(mrna_counts_subset)
```

Run PCC
```{r}
# Create an empty dataframe to store results
results <- data.frame(mRNA1 = character(), 
                      mRNA2 = character(), 
                      estimate = numeric(), 
                      p_value = numeric(), 
                      stringsAsFactors = FALSE)

# Run cor.test for each pair
for (mRNA1 in mRNAs_list1) {
  for (mRNA2 in mRNAs_list2) {
    if (mRNA1 != mRNA2) {
      test_result <- cor.test(mrna_counts_subset_norm[mRNA1,], mrna_counts_subset_norm[mRNA2,], method = "pearson")
      results <- rbind(results, data.frame(mRNA1 = mRNA1,
                                           mRNA2 = mRNA2,
                                           estimate = test_result$estimate,
                                           p_value = test_result$p.value))
    }
  }
}

# Adjust p-values for FDR
results <- results %>%
  mutate(adjusted_p_value = p.adjust(p_value, method = "fdr"))
```

Filter `results_neg` so that only negative correlations remain 
```{r}
results_neg <- results %>%
  filter(estimate < 0)
```

Filter `combined_data_pcc_mrna_mirna` so that only negative correlations remain 
```{r}
combined_data_pcc_mrna_mirna_neg <- combined_data_pcc_mrna_mirna %>%
  filter(PCC.cor < 0)
```

Merge `results_neg` with `combined_data_pcc_mrna_mirna_neg`
```{r}
merge <- results_neg %>%
  inner_join(combined_data_pcc_mrna_mirna_neg, by = c("mRNA2" = "mRNA"))
```

Identify miRNAs of interest from sig pos df and subset `merge` to only include these miRNAs 
```{r}
unique_clusters <- unique(combined_data_pcc_mrna_mirna_sig_pos$miRNA)
filtered_merge <- merge[merge$miRNA %in% unique_clusters, ]
```

mRNA1 are the mRNAs from the significant positive correlations (`combined_data_pcc_mrna_mirna_sig_pos$mRNA`) and mRNA2 are the mRNAs that are negatively correlated with the mRNA1s. 

Remove any correlations >-0.1. 
```{r}
filtered_merge.1 <- filtered_merge %>%
  filter(estimate < -0.1) %>%
  filter(PCC.cor < -0.1) %>%
  filter(mRNA_DE == "Y") %>%
  filter(miRNA_DE == "Y")
length(unique(filtered_merge.1$mRNA1)) # should be 11
length(unique(filtered_merge.1$mRNA2))
length(unique(filtered_merge.1$miRNA)) # should be 8
```

### Examine specific correlations of interest 

One of the pathways that I have detected is the miR-100 mediated uptake of inorganic phosphate in Feb when nutrients are low. Based on my data, I think that E2 ligases, acid phosphates and miR100 are forming a pathway to facilitate this. I want to see if the E2 ligases are also correlated with the acid phosphotases 

Extract relevant mRNAs
```{r}
mRNAs_list1 <- "evm.TU.chromosome_13.1876" 
mRNAs_list2 <- "evm.TU.chromosome_12.2473" 
```

Subset the counts data 
```{r}
# Subset the counts matrix
mrna_counts_subset <- mrna_counts[c(mRNAs_list1, mRNAs_list2), ]
```

```{r}
mrna_counts_reshaped <- mrna_counts_subset %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  rename(
    Gene1 = `evm.TU.chromosome_13.1876`,
    Gene2 = `evm.TU.chromosome_12.2473`
  )

# Create the plot
ggplot(mrna_counts_reshaped, aes(x = Gene2, y = Gene1)) +
  geom_point() +
  geom_text(aes(label = Sample), hjust = 0, vjust = 0) +
  labs(title = "Gene counts comparison") +
  theme_minimal()

cor.test(mrna_counts_reshaped$Gene2, mrna_counts_reshaped$Gene1)
```

Confused...these two mRNAs have a positive correlation but I expected them to have a negative correlation...