---
title: "topGO"
author: "Jill Ashey"
date: "2025-02-04"
output: html_document
---

This script will use topGO to analyze functional enrichment of miRNA targets. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(topGO)
library(tidyverse)
```

Before analysis, make gene2go file from annotation information
```{r}
# Read in annotation info
annot <- read.delim("/Users/jillashey/Desktop/PutnamLab/Astrangia_Genome/Apoculata_v2.0_GeneAnnotation_combined_prelim.txt", header = T)
annot$Protein_ID <- gsub("model", "TU", annot$Protein_ID)
annot$GO <- paste(annot$GO_Swiss.Prot, annot$GO_Trembl, sep = ";")

# Select gene name and GO columns only 
go_ast <- annot %>%
  dplyr::select(Protein_ID, GO) %>%
  mutate(GO = case_when(
    grepl("^NA$|^NA;|;$|^;$", GO, ignore.case = TRUE) ~ NA_character_,
    TRUE ~ GO
  ))

# Rename cols
colnames(go_ast) <- c("mRNA", "GO")

# Save as tsv
write_tsv(go_ast, file = "../data/Molecular/Ast_gene2go.tab")
```

## Functional enrichment of all targets, regardless of differential expression status 

Read in miranda data - entire mRNA binding
```{r}
miranda_data <- read.delim("~/Desktop/PutnamLab/Astrangia/Molecular/miranda_strict_all_mrna_apoc_shortstack_parsed.txt", header = F)
colnames(miranda_data) <- c("miRNA", "mRNA", "score", "energy", "query_start_end", "subject_start_end", "total_bp_shared", "query_similar", "subject_similar")

# Format miranda df 
miranda_data$miRNA <- sub("^>", "", miranda_data$miRNA)  # Remove leading ">"
miranda_data$miRNA <- sub("\\..*", "", miranda_data$miRNA)  # Remove everything from the first period onwards
miranda_data$mRNA <- sub(";.*", "", miranda_data$mRNA)  # Remove everything from "::" onwards
miranda_data$mRNA <- sub("ID=", "", miranda_data$mRNA)  # Remove everything from "::" onwards
miranda_data$mRNA <- sub("model", "TU", miranda_data$mRNA)  # Remove everything from "::" onwards

dim(miranda_data)
length(unique(miranda_data$miRNA))
length(unique(miranda_data$mRNA))
```

Read in expressed genes file 
```{r}
filt_counts <- read.csv("../output/Molecular/mRNA/filtered_gene_counts.csv")
```

Join miranda and filtered gene counts df 
```{r}
joined_df <- miranda_data %>%
  left_join(filt_counts, by = c("mRNA" = "X")) %>%
  filter(if_all(starts_with("AST"), ~ !is.na(.)))

length(unique(joined_df$miRNA))
length(unique(joined_df$mRNA))
```

Read in gene2go information (generated above)
```{r}
ast_gene2go <- read.delim("../data/Molecular/Ast_gene2go.tab", sep = "\t")
```

Make list of genes for input to topGO
```{r}
# Genes of interest - ie those targeted by miRNAs
ast_target_genes <- as.character(unique(joined_df$mRNA))

# All genes 
ast_all_genes <- as.character(ast_gene2go$mRNA)

# Apply 1 or 0 if gene is gene of interest 
ast_GeneList <- factor(as.integer(ast_all_genes %in% ast_target_genes))
names(ast_GeneList) <- ast_all_genes
```

The following code will perform GO enrichment using the weighted Fisher's exact test to assess whether specific GO terms are overrepresented in the genes targeted by miRNAs. 

Read in gene-to-go-mappings
```{r}
ast_gene2go_topgo<-readMappings("../data/Molecular/Ast_gene2go.tab", IDsep=";", sep="\t")
```

Set function to select genes of interest (ie those that have pvalue < 0.05)
```{r}
topDiffGenes <- function(allScore) {
return(allScore < 0.05)}
```

### Biological Processes

Create `topGOdata` object, which is required for topGO analysis
```{r}
GO_ast_BP <-new("topGOdata", ontology="BP", gene2GO=ast_gene2go_topgo, allGenes=ast_GeneList, annot = annFUN.gene2GO, geneSel=topDiffGenes)
```

Run GO enrichment test 
```{r}
GO_ast_BP_FE <- runTest(GO_ast_BP, algorithm="weight01", statistic="fisher")
```

Generate results table 
```{r}
GO_ast_BP_En <- GenTable(GO_ast_BP, Fisher = GO_ast_BP_FE, orderBy = "Fisher",  topNodes = 100, numChar = 51)
```
Only taking the top 100 GO terms 

Filter by significant results
```{r}
GO_ast_BP_En$Fisher<-as.numeric(GO_ast_BP_En$Fisher)
GO_ast_BP_En_sig<-GO_ast_BP_En[GO_ast_BP_En$Fisher<0.05,]
```

Merge `GO_ast_BP_En_sig` with GO and gene info. 
```{r}
# Separate GO terms 
ast_gene2go <- ast_gene2go %>%
  separate_rows(GO, sep = ";")

# Ensure GO terms in both datasets are formatted similarly (trim whitespaces)
ast_gene2go$GO <- trimws(ast_gene2go$GO)
GO_ast_BP_En_sig$GO.ID <- trimws(GO_ast_BP_En_sig$GO.ID)

# Join the datasets based on GO term
GO_ast_BP_En_sig_gene <- ast_gene2go %>%
  left_join(GO_ast_BP_En_sig, by = c("GO" = "GO.ID")) %>%
  na.omit()

# Join the datasets based on genes upregulated in Feb (ie those used for the functional enrichment)
GO_ast_BP_En_sig_gene <- GO_ast_BP_En_sig_gene %>%
  filter(mRNA %in% ast_target_genes)

# Add ontology column 
GO_ast_BP_En_sig_gene$ontology <- "Biological Processes"
```

### Molecular Function

Create `topGOdata` object, which is required for topGO analysis
```{r}
GO_ast_MF <-new("topGOdata", ontology="MF", gene2GO=ast_gene2go_topgo, allGenes=ast_GeneList, annot = annFUN.gene2GO, geneSel=topDiffGenes)
```

Run GO enrichment test 
```{r}
GO_ast_MF_FE <- runTest(GO_ast_MF, algorithm="weight01", statistic="fisher")
```

Generate results table 
```{r}
GO_ast_MF_En <- GenTable(GO_ast_MF, Fisher = GO_ast_MF_FE, orderBy = "Fisher",  topNodes = 100, numChar = 51)
```
Only taking the top 100 GO terms 

Filter by significant results
```{r}
GO_ast_MF_En$Fisher<-as.numeric(GO_ast_MF_En$Fisher)
GO_ast_MF_En_sig<-GO_ast_MF_En[GO_ast_MF_En$Fisher<0.05,]
```

Merge `GO_ast_MF_En_sig` with GO and gene info. 
```{r}
# Separate GO terms 
ast_gene2go <- ast_gene2go %>%
  separate_rows(GO, sep = ";")

# Ensure GO terms in both datasets are formatted similarly (trim whitespaces)
ast_gene2go$GO <- trimws(ast_gene2go$GO)
GO_ast_MF_En_sig$GO.ID <- trimws(GO_ast_MF_En_sig$GO.ID)

# Join the datasets based on GO term
GO_ast_MF_En_sig_gene <- ast_gene2go %>%
  left_join(GO_ast_MF_En_sig, by = c("GO" = "GO.ID")) %>%
  na.omit()

# Join the datasets based on genes upregulated in Feb (ie those used for the functional enrichment)
GO_ast_MF_En_sig_gene <- GO_ast_MF_En_sig_gene %>%
  filter(mRNA %in% ast_target_genes)

# Add ontology column 
GO_ast_MF_En_sig_gene$ontology <- "Molecular Functions"
```

### Join ontologies 

Bind so there is a df that has significantly enriched GO terms for all ontologies 
```{r}
GO_En_sig_gene <- rbind(GO_ast_BP_En_sig_gene, GO_ast_MF_En_sig_gene)

# Calculate proportion of significant v annotated genes 
GO_En_sig_gene <- GO_En_sig_gene %>%
  mutate(sig.prop = Significant/Annotated) %>%
  na.omit()
length(unique(GO_En_sig_gene$Term))
length(unique(GO_En_sig_gene$mRNA))

# Save as csv 
write.csv(GO_En_sig_gene, "../output/Molecular/interactions/enrichment/GO_en_sig_miRNA_target_genes.csv")
```

Plot by number of genes that have enriched terms
```{r}
plot_data <- GO_En_sig_gene %>%
  #filter(ontology != "Cellular Components") %>%
  filter(ontology != "Molecular Functions") %>%
  group_by(Term, ontology) %>%
  summarise(gene_count = n_distinct(mRNA),
            Fisher = mean(Fisher), .groups = 'drop') %>%
  group_by(ontology) %>%
  slice_max(order_by = gene_count, n = 15)

ggplot(plot_data, aes(x = reorder(Term, gene_count), y = gene_count, fill = Fisher)) +
  geom_bar(stat = "identity") +
  #geom_text(aes(label = gene_count), vjust = -0.5, color = "black", size = 3) +
  coord_flip() +
  scale_fill_gradient(low = "blue", high = "red") +
  #facet_grid(vars(ontology), scales = "free", space = "free_y") + 
  labs(x = "",
       y = "Number of Genes",
       fill = "Fisher Value") +
  theme_bw() +
  guides(fill = guide_colorbar(barwidth = 1.5, barheight = 30)) +
  theme(
    legend.title = element_text(size = 38, face = "bold"),
    legend.text = element_text(size = 36),
    legend.key.height = unit(2, "cm"),
    legend.key.width  = unit(1, "cm"),
    legend.margin = margin(0, 20, 0, 0)
  ) +
  theme(
    axis.title = element_text(size = 70, face = "bold"),   # slightly larger axis titles
    axis.text = element_text(size = 68, colour = "black"), # slightly larger axis text
    legend.title = element_text(size = 68, face = "bold"), # larger legend title
    legend.text = element_text(size = 66),                 # larger legend text
    strip.text = element_text(size = 68, face = "bold"),
    strip.background = element_rect(fill = "lightgray", color = "black", size = 1.5),
    axis.line = element_line(size = 1, colour = "black"),
    axis.ticks = element_line(size = 1),
    panel.border = element_rect(color = "black", size = 1.2),
    panel.grid.major = element_line(size = 0.5, color = "gray"),
    panel.spacing = unit(1, "lines"),
    strip.placement = "outside",
    legend.position = "right",           # move legend outside plot on right
    #legend.justification = c(0, 1),      # align legend top left
    legend.margin = margin(0, 20, 0, 0)  # add right margin so legend is separate
  )

ggsave(filename = "../output/Molecular/interactions/enrichment/GO_en_sig_miRNA_target_genes_top10.pdf", last_plot(), width = 49, height = 40)
ggsave(filename = "../output/Molecular/interactions/enrichment/GO_en_sig_miRNA_target_genes_top10.png", last_plot(), width = 49, height = 40)
```

Other ways to plot 
```{r}
ggplot(plot_data, aes(x = ontology, y = reorder(Term, gene_count), fill = Fisher)) +
  geom_tile(color = "white") +
  scale_fill_gradient(low = "blue", high = "red") +
  labs(x = "Ontology", y = "Term", fill = "-log10(Fisher)") +
  theme_minimal()

ggplot(plot_data, aes(x = ontology, 
                      y = reorder(Term, gene_count),
                      size = gene_count, 
                      color = -log10(Fisher))) +
  geom_point() +
  scale_color_gradient(low = "blue", high = "red") +
  labs(x = "Ontology", y = "Term", 
       size = "Gene Count", 
       color = "-log10(Fisher)") +
  theme_bw()

ggplot(plot_data, aes(x = gene_count, y = reorder(Term, gene_count), color = -log10(Fisher))) +
  geom_segment(aes(x = 0, xend = gene_count, yend = Term), linewidth = 1.2) +
  geom_point(size = 5) +
  scale_color_gradient(low = "blue", high = "red") +
  labs(x = "Number of Genes", y = "", color = "-log10(Fisher)") +
  theme_bw() +
  theme(
    axis.title = element_text(size = 40, face = "bold"),   # slightly larger axis titles
    axis.text = element_text(size = 38, colour = "black"), # slightly larger axis text
    legend.title = element_text(size = 38, face = "bold"), # larger legend title
    legend.text = element_text(size = 36),                 # larger legend text
    strip.text = element_text(size = 38, face = "bold"),
    strip.background = element_rect(fill = "lightgray", color = "black", size = 1.5),
    axis.line = element_line(size = 1, colour = "black"),
    axis.ticks = element_line(size = 1),
    panel.border = element_rect(color = "black", size = 1.2),
    panel.grid.major = element_line(size = 0.5, color = "gray"),
    panel.spacing = unit(1, "lines"),
    strip.placement = "outside",
    legend.position = "right",           # move legend outside plot on right
    #legend.justification = c(0, 1),      # align legend top left
    legend.margin = margin(0, 20, 0, 0)  # add right margin so legend is separate
  )
ggsave(filename = "../output/Molecular/interactions/enrichment/GO_en_sig_miRNA_target_genes_top10_lollipop.pdf", last_plot(), width = 30, height = 40)
ggsave(filename = "../output/Molecular/interactions/enrichment/GO_en_sig_miRNA_target_genes_top10_lollipop.png", last_plot(), width = 30, height = 40)
```

