---
title: "topGO"
author: "Jill Ashey"
date: "2025-02-04"
output: html_document
---

This script will use topGO to analyze functional enrichment of miRNA targets. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(topGO)
library(tidyverse)
```

Before analysis, make gene2go file from annotation information
```{r}
# Read in annotation info
annot <- read.delim("/Users/jillashey/Desktop/PutnamLab/Astrangia_Genome/Apoculata_v2.0_GeneAnnotation_combined_prelim.txt", header = T)
annot$Protein_ID <- gsub("model", "TU", annot$Protein_ID)
annot$GO <- paste(annot$GO_Swiss.Prot, annot$GO_Trembl, sep = ";")

# Select gene name and GO columns only 
go_ast <- annot %>%
  dplyr::select(Protein_ID, GO) %>%
  mutate(GO = case_when(
    grepl("^NA$|^NA;|;$|^;$", GO, ignore.case = TRUE) ~ NA_character_,
    TRUE ~ GO
  ))

# Rename cols
colnames(go_ast) <- c("mRNA", "GO")

# Save as tsv
write_tsv(go_ast, file = "../data/Molecular/Ast_gene2go.tab")
```

## Functional enrichment of all targets, regardless of differential expression status 

Read in miranda data
```{r}
miranda_data <- read.delim("../output/Molecular/interactions/miranda_strict_all_1kb_apoc_shortstack_parsed.txt", header = F)
colnames(miranda_data) <- c("miRNA", "mRNA", "score", "energy", "query_start_end", "subject_start_end", "total_bp_shared", "query_similar", "subject_similar")

# Format miranda df 
miranda_data$miRNA <- sub("^>", "", miranda_data$miRNA)  # Remove leading ">"
miranda_data$miRNA <- sub("\\..*", "", miranda_data$miRNA)  # Remove everything from the first period onwards
miranda_data$mRNA <- sub(";.*", "", miranda_data$mRNA)  # Remove everything from "::" onwards
miranda_data$mRNA <- sub("ID=", "", miranda_data$mRNA)  # Remove everything from "::" onwards

dim(miranda_data)
```

Read in gene2go information (generated above)
```{r}
ast_gene2go <- read.delim("../data/Molecular/Ast_gene2go.tab", sep = "\t")
```

Make list of genes for input to topGO
```{r}
# Genes of interest - ie those targeted by miRNAs
ast_target_genes <- as.character(unique(miranda_data$mRNA))

# All genes 
ast_all_genes <- as.character(ast_gene2go$mRNA)

# Apply 1 or 0 if gene is gene of interest 
ast_GeneList <- factor(as.integer(ast_all_genes %in% ast_target_genes))
names(ast_GeneList) <- ast_all_genes
```

The following code will perform GO enrichment using the weighted Fisher's exact test to assess whether specific GO terms are overrepresented in the genes targeted by miRNAs. 

Read in gene-to-go-mappings
```{r}
ast_gene2go_topgo<-readMappings("../data/Molecular/Ast_gene2go.tab", IDsep=";", sep="\t")
```

Set function to select genes of interest (ie those that have pvalue < 0.05)
```{r}
topDiffGenes <- function(allScore) {
return(allScore < 0.05)}
```

### Biological Processes

Create `topGOdata` object, which is required for topGO analysis
```{r}
GO_ast_BP <-new("topGOdata", ontology="BP", gene2GO=ast_gene2go_topgo, allGenes=ast_GeneList, annot = annFUN.gene2GO, geneSel=topDiffGenes)
```

Run GO enrichment test 
```{r}
GO_ast_BP_FE <- runTest(GO_ast_BP, algorithm="weight01", statistic="fisher")
```

Generate results table 
```{r}
GO_ast_BP_En <- GenTable(GO_ast_BP, Fisher = GO_ast_BP_FE, orderBy = "Fisher",  topNodes = 100, numChar = 51)
```
Only taking the top 100 GO terms 

Filter by significant results
```{r}
GO_ast_BP_En$Fisher<-as.numeric(GO_ast_BP_En$Fisher)
GO_ast_BP_En_sig<-GO_ast_BP_En[GO_ast_BP_En$Fisher<0.05,]
```

Merge `GO_ast_BP_En_sig` with GO and gene info. 
```{r}
# Separate GO terms 
ast_gene2go <- ast_gene2go %>%
  separate_rows(GO, sep = ";")

# Ensure GO terms in both datasets are formatted similarly (trim whitespaces)
ast_gene2go$GO <- trimws(ast_gene2go$GO)
GO_ast_BP_En_sig$GO.ID <- trimws(GO_ast_BP_En_sig$GO.ID)

# Join the datasets based on GO term
GO_ast_BP_En_sig_gene <- ast_gene2go %>%
  left_join(GO_ast_BP_En_sig, by = c("GO" = "GO.ID")) %>%
  na.omit()

# Add ontology column 
GO_ast_BP_En_sig_gene$ontology <- "Biological Processes"
```









## Functional enrichment of differentially expressed targets by differentially expressed miRNAs (regardless of correlation significance)

Read in data - this data is the DE miRNAs and DE target mRNAs but not filtered by significant correlation values
```{r}
pcc_de <- read.csv("../output/Molecular/interactions/interactions_annot_pcc_de.csv")
```

Read in gene2go information (generated above)
```{r}
ast_gene2go <- read.delim("../data/Molecular/Ast_gene2go.tab", sep = "\t")
```

Make list of genes for input to topGO
```{r}
# Genes of interest - ie those targeted by miRNAs
ast_target_genes <- as.character(unique(pcc_de$mRNA))

# All genes 
ast_all_genes <- as.character(ast_gene2go$mRNA)

# Apply 1 or 0 if gene is gene of interest 
ast_GeneList <- factor(as.integer(ast_all_genes %in% ast_target_genes))
names(ast_GeneList) <- ast_all_genes
```

The following code will perform GO enrichment using the weighted Fisher's exact test to assess whether specific GO terms are overrepresented in the DE genes targeted by DE miRNAs, regardless of correlation significance. 

Read in gene-to-go-mappings
```{r}
ast_gene2go_topgo<-readMappings("../data/Molecular/Ast_gene2go.tab", IDsep=";", sep="\t")
```

Set function to select genes of interest (ie those that have pvalue < 0.05)
```{r}
topDiffGenes <- function(allScore) {
return(allScore < 0.05)}
```

### Biological Processes

Create `topGOdata` object, which is required for topGO analysis
```{r}
GO_ast_BP <-new("topGOdata", ontology="BP", gene2GO=ast_gene2go_topgo, allGenes=ast_GeneList, annot = annFUN.gene2GO, geneSel=topDiffGenes)
```

Run GO enrichment test 
```{r}
GO_ast_BP_FE <- runTest(GO_ast_BP, algorithm="weight01", statistic="fisher")
```

Generate results table 
```{r}
GO_ast_BP_En <- GenTable(GO_ast_BP, Fisher = GO_ast_BP_FE, orderBy = "Fisher",  topNodes = 100, numChar = 51)
```
Only taking the top 100 GO terms 

Filter by significant results
```{r}
GO_ast_BP_En$Fisher<-as.numeric(GO_ast_BP_En$Fisher)
GO_ast_BP_En_sig<-GO_ast_BP_En[GO_ast_BP_En$Fisher<0.05,]
```

Merge `GO_ast_BP_En_sig` with GO and gene info. 
```{r}
# Separate GO terms 
ast_gene2go <- ast_gene2go %>%
  separate_rows(GO, sep = ";")

# Ensure GO terms in both datasets are formatted similarly (trim whitespaces)
ast_gene2go$GO <- trimws(ast_gene2go$GO)
GO_ast_BP_En_sig$GO.ID <- trimws(GO_ast_BP_En_sig$GO.ID)

# Join the datasets based on GO term
GO_ast_BP_En_sig_gene <- ast_gene2go %>%
  left_join(GO_ast_BP_En_sig, by = c("GO" = "GO.ID")) %>%
  na.omit()

# Add ontology column 
GO_ast_BP_En_sig_gene$ontology <- "Biological Processes"

# Keep only unique rows 
GO_ast_BP_En_sig_gene <- unique(GO_ast_BP_En_sig_gene)
```

### Cellular Components 

Create `topGOdata` object, which is required for topGO analysis
```{r}
GO_ast_CC <-new("topGOdata", ontology="CC", gene2GO=ast_gene2go_topgo, allGenes=ast_GeneList, annot = annFUN.gene2GO, geneSel=topDiffGenes)
```

Run GO enrichment test 
```{r}
GO_ast_CC_FE <- runTest(GO_ast_CC, algorithm="weight01", statistic="fisher")
```

Generate results table 
```{r}
GO_ast_CC_En <- GenTable(GO_ast_CC, Fisher = GO_ast_CC_FE, orderBy = "Fisher",  topNodes = 100, numChar = 51)
```
Only taking the top 100 GO terms 

Filter by significant results
```{r}
GO_ast_CC_En$Fisher<-as.numeric(GO_ast_CC_En$Fisher)
GO_ast_CC_En_sig<-GO_ast_CC_En[GO_ast_CC_En$Fisher<0.05,]
```

Merge `GO_ast_CC_En_sig` with GO and gene info. 
```{r}
# Separate GO terms 
ast_gene2go <- ast_gene2go %>%
  separate_rows(GO, sep = ";")

# Ensure GO terms in both datasets are formatted similarly (trim whitespaces)
ast_gene2go$GO <- trimws(ast_gene2go$GO)
GO_ast_CC_En_sig$GO.ID <- trimws(GO_ast_CC_En_sig$GO.ID)

# Join the datasets based on GO term
GO_ast_CC_En_sig_gene <- ast_gene2go %>%
  left_join(GO_ast_CC_En_sig, by = c("GO" = "GO.ID")) %>%
  na.omit()

# Add ontology column 
GO_ast_CC_En_sig_gene$ontology <- "Cellular Components"

# Keep only unique rows 
GO_ast_CC_En_sig_gene <- unique(GO_ast_CC_En_sig_gene)
```

### Molecular Functions 

Create `topGOdata` object, which is required for topGO analysis
```{r}
GO_ast_MF <-new("topGOdata", ontology="MF", gene2GO=ast_gene2go_topgo, allGenes=ast_GeneList, annot = annFUN.gene2GO, geneSel=topDiffGenes)
```

Run GO enrichment test 
```{r}
GO_ast_MF_FE <- runTest(GO_ast_MF, algorithm="weight01", statistic="fisher")
```

Generate results table 
```{r}
GO_ast_MF_En <- GenTable(GO_ast_MF, Fisher = GO_ast_MF_FE, orderBy = "Fisher",  topNodes = 100, numChar = 51)
```
Only taking the top 100 GO terms 

Filter by significant results
```{r}
GO_ast_MF_En$Fisher<-as.numeric(GO_ast_MF_En$Fisher)
GO_ast_MF_En_sig<-GO_ast_MF_En[GO_ast_MF_En$Fisher<0.05,]
```

Merge `GO_ast_MF_En_sig` with GO and gene info. 
```{r}
# Separate GO terms 
ast_gene2go <- ast_gene2go %>%
  separate_rows(GO, sep = ";")

# Ensure GO terms in both datasets are formatted similarly (trim whitespaces)
ast_gene2go$GO <- trimws(ast_gene2go$GO)
GO_ast_MF_En_sig$GO.ID <- trimws(GO_ast_MF_En_sig$GO.ID)

# Join the datasets based on GO term
GO_ast_MF_En_sig_gene <- ast_gene2go %>%
  left_join(GO_ast_MF_En_sig, by = c("GO" = "GO.ID")) %>%
  na.omit()

# Add ontology column 
GO_ast_MF_En_sig_gene$ontology <- "Molecular Functions"

# Keep only unique rows 
GO_ast_MF_En_sig_gene <- unique(GO_ast_MF_En_sig_gene)
```

### Join intologies 

Bind so there is a df that has significantly enriched GO terms for all ontologies 
```{r}
GO_ast_de <- rbind(GO_ast_BP_En_sig_gene, GO_ast_CC_En_sig_gene, GO_ast_MF_En_sig_gene)
```

Merge with `pcc_de`
```{r}
test <- GO_ast_de %>%
  inner_join(pcc_de, by = "mRNA") %>%
  mutate(direction = ifelse(PCC.cor > 0, "Positive", "Negative")) %>%
  filter(ontology != "Cellular Components") #%>%
  #filter(p_value < 0.1)

# Save as csv 
write.csv(test, "../output/Molecular/interactions/topGO_pcc_de_target_enrichment.csv")
```

Plot!
```{r}
plot<-ggplot(test, aes(x = Term, y = p_value, size = Fisher, fill = direction)) +
  #ylim(0, 1) +
  #geom_hline(yintercept = 0.05, linetype = "solid", color = "black", linewidth = 1)+
  geom_hline(yintercept = 0.05, color = "black", linetype = "solid", linewidth = 0.5) +  # Add line at 0.05 
  geom_point(shape = 21) + 
  scale_size(range = c(2, 20)) + 
  xlab('') + 
  ylab("PCC p-value") +
  theme_bw(base_size = 24) +
  facet_grid(vars(ontology), scales = "free", space = "free_y") +
  coord_flip(); plot

# Save plot 
ggsave("../output/Molecular/interactions/topGO_pcc_de_target_enrichment.pdf", plot, width = 20, height = 25, dpi = 300)
ggsave("../output/Molecular/interactions/topGO_pcc_de_target_enrichment.png", plot, width = 20, height = 25, dpi = 300)
```

Plot! But leave out PCC information
```{r}
test <- test %>%
  arrange(desc(Fisher)) %>%
  mutate(Term = factor(Term, levels = unique(Term)))

plot<-ggplot(test, aes(x = Term, y = Fisher)) +
  #ylim(0, 1) +
  #geom_hline(yintercept = 0.05, linetype = "solid", color = "black", linewidth = 1)+
  #geom_hline(yintercept = 0.05, color = "black", linetype = "solid", linewidth = 0.5) +  # Add line at 0.05 
  geom_point(size = 10, color = "black") + 
  scale_size(range = c(2, 20)) + 
  xlab('') + 
  ylab("Fisher p-value") +
  theme_bw(base_size = 35) +
  facet_grid(vars(ontology), scales = "free", space = "free_y") +
  coord_flip(); plot

# Save plot 
ggsave("../output/Molecular/interactions/topGO_de_target_enrichment.pdf", plot, width = 20, height = 40, dpi = 300)
ggsave("../output/Molecular/interactions/topGO_de_target_enrichment.png", plot, width = 20, height = 40, dpi = 300)
```

Assess data
```{r}
length(unique(test$mRNA)) # Count number of DEGs
length(unique(test$miRNA)) # Count number of DEMs
sum(test$ontology == "Biological Processes", na.rm = TRUE) # Count number of BP GO terms that are significantly enriched

# Function to get top unique terms
get_top_unique <- function(data, ontology_type) {
  data %>%
    filter(ontology == ontology_type) %>%
    arrange(Fisher) %>%
    distinct(Term, .keep_all = TRUE) 
}
top_BP <- get_top_unique(test, "Biological Processes")
top_MF <- get_top_unique(test, "Molecular Functions")
```

Examine the top 10 most significant GO terms for BP and MF
```{r}
# Function to get top 5 unique terms
get_top_10_unique <- function(data, ontology_type) {
  data %>%
    filter(ontology == ontology_type) %>%
    arrange(Fisher) %>%
    distinct(Term, .keep_all = TRUE) %>%
    slice_head(n = 10)
}

# Get top 5 unique Biological Processes
top_10_BP <- get_top_10_unique(test, "Biological Processes")

# Get top 5 unique Molecular Functions
top_10_MF <- get_top_10_unique(test, "Molecular Functions")

# Combine results
top_10_combined <- bind_rows(top_10_BP, top_10_MF)
unique(top_5_combined$Term)

# Plot 
top_10_combined <- top_10_combined %>%
  arrange(desc(Fisher)) %>%
  mutate(Term = factor(Term, levels = unique(Term)))

plot<-ggplot(top_10_combined, aes(x = Term, y = Fisher)) +
  #ylim(0, 1) +
  #geom_hline(yintercept = 0.05, linetype = "solid", color = "black", linewidth = 1)+
  #geom_hline(yintercept = 0.05, color = "black", linetype = "solid", linewidth = 0.5) +  # Add line at 0.05 
  geom_point(size = 10, color = "black") + 
  scale_size(range = c(2, 20)) + 
  xlab('') + 
  ylab("Fisher p-value") +
  theme_bw(base_size = 35) +
  facet_grid(vars(ontology), scales = "free", space = "free_y") +
  coord_flip(); plot

# Save plot 
ggsave("../output/Molecular/interactions/top10GO_de_target_enrichment.pdf", plot, width = 20, height = 25, dpi = 300)
ggsave("../output/Molecular/interactions/top10GO_de_target_enrichment.png", plot, width = 20, height = 25, dpi = 300)
```





