---
title: "Weighted Jaccard Index for mRNA-miRNA networks"
author: "Jill Ashey"
date: "2025-07-16"
output: html_document
---

Steps 
- Subset samples for one treatment 
- Run PCC for those samples 
- Build networks using the PCC and miranda data 
- Examine intersection of two groups and calculate jaccard index - this will tell us how similar/different the groups are 
- Look at the PCC signs - this will tell us if the correlation direction is changing 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(DESeq2)
library(pbapply)
library(igraph)
library(ggraph)
library(tidygraph)
library(DT)
library(pheatmap)
```

Read in mRNA count data
```{r}
mrna_counts <- read.csv("../output/Molecular/mRNA/filtered_gene_counts.csv")

# Set row names
rownames(mrna_counts) <- mrna_counts[,1] #set first column that contains gene names as rownames
mrna_counts <- mrna_counts[,-1] # remove column w/ gene names 
```

Read in miRNA count data
```{r}
mirna_counts <- read.table("../output/Molecular/smRNA/shortstack/Counts.csv", header = T) 
```

Select only miRNAs with a Y in MIRNA column. Combine Name and Coords columns for miRNA name. Remove MIRNA column 
```{r}
mirna_counts <- mirna_counts %>%
  dplyr::filter(MIRNA == "Y") %>%
  mutate(miRNA_name = paste(Name, Coords, sep = "_")) %>%
  dplyr::select(-c(Name, Coords, MIRNA))

# Set row names 
rownames(mirna_counts) <- mirna_counts[,20] #set column that contains names as rownames
mirna_counts <- mirna_counts[,-20] # remove column w/ gene names 
```

Remove extra info from column names
```{r}
colnames(mirna_counts) <- gsub("_R1_001.fastq.gz_1", "", colnames(mirna_counts))
colnames(mirna_counts) <- gsub("AST.", "AST-", colnames(mirna_counts))
colnames(mrna_counts) <- gsub("AST.", "AST-", colnames(mrna_counts))
```

Remove sample that did not sequence well
```{r}
mrna_counts <- mrna_counts %>%
    dplyr::select(-"AST-1105")
```

Read in metadata 
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv")
meta <- dplyr::arrange(meta, ID) # rearrange metadata so IDs are sorted in descending order 
#meta$ID <- gsub("AST-", "AST.", meta$ID)

# Set variables as factors 
meta$Timepoint <- factor(meta$Timepoint, levels = c("TP0", "TP5", "TP7"))
meta$Treatment <- factor(meta$Treatment, levels = c("Acclimation", "Ambient", "Heat"))

# Remove sample that did not sequence well 
meta <- meta %>%
    dplyr::filter(!ID=="AST-1105")
```

## February network 

Subset February samples from counts dfs
```{r}
# Get TP0 sample IDs
tp0_ids <- meta$ID[meta$Timepoint == "TP0"]

# Subset columns in counts
mirna_tp0 <- mirna_counts[, tp0_ids]
mrna_tp0  <- mrna_counts[, tp0_ids]
rownames(mirna_tp0) <- rownames(mirna_counts)
rownames(mrna_tp0)  <- rownames(mrna_counts)
```

Perform rlog on counts 
```{r}
# Filter meta for TP0
tp0_meta <- meta[meta$Timepoint == "TP0", ]
tp0_ids <- tp0_meta$ID

# Subset counts to TP0 samples
mirna_tp0 <- mirna_counts[, tp0_ids]
mrna_tp0  <- mrna_counts[, tp0_ids]

# Add dummy colData if you don’t have any covariates
dds_mirna <- DESeqDataSetFromMatrix(countData = mirna_tp0,
                                    colData = tp0_meta,
                                    design = ~1)

dds_mrna  <- DESeqDataSetFromMatrix(countData = mrna_tp0,
                                    colData = tp0_meta,
                                    design = ~1)

# Run rlog (blind = TRUE for no design)
rlog_mirna <- assay(rlog(dds_mirna, blind = TRUE))
rlog_mrna  <- assay(rlog(dds_mrna, blind = TRUE))

# Check shapes
dim(rlog_mirna)
dim(rlog_mrna)

```

Run PCC on February data 
```{r}
miRNA_ids <- rownames(rlog_mirna)
mRNA_ids  <- rownames(rlog_mrna)

# Make grid of all combos
grid <- expand.grid(miRNA = miRNA_ids, mRNA = mRNA_ids, stringsAsFactors = FALSE)

# Compute PCC and p-value
results <- pbapply::pblapply(1:nrow(grid), function(i) {
  mir <- rlog_mirna[grid$miRNA[i], ]
  mr  <- rlog_mrna[grid$mRNA[i], ]
  ct  <- cor.test(mir, mr, method = "pearson")
  data.frame(miRNA = grid$miRNA[i],
             mRNA = grid$mRNA[i],
             PCC = ct$estimate,
             Pvalue = ct$p.value)
})

results_df <- do.call(rbind, results)
str(results_df)

# format miRNA names 
results_df <- results_df %>%
    mutate(miRNA = str_replace(miRNA, "^([^_]*_[^_]*)_.*$", "\\1"))
```

### Full binding 

Read in full binding miranda data 
```{r}
miranda_full <- read.delim("~/Desktop/PutnamLab/Astrangia/Molecular/miranda_strict_all_mrna_apoc_shortstack_parsed.txt", header = F)
colnames(miranda_full) <- c("miRNA", "mRNA", "score", "energy", "query_start_end", "subject_start_end", "total_bp_shared", "query_similar", "subject_similar")

# Format miranda df 
miranda_full$miRNA <- sub("^>", "", miranda_full$miRNA)  # Remove leading ">"
miranda_full$miRNA <- sub("\\..*", "", miranda_full$miRNA)  # Remove everything from the first period onwards
miranda_full$mRNA <- sub(";.*", "", miranda_full$mRNA)  # Remove everything from "::" onwards
miranda_full$mRNA <- sub("ID=", "", miranda_full$mRNA)  # Remove everything from "::" onwards
miranda_full$mRNA <- sub("model", "TU", miranda_full$mRNA)  # Remove everything from "::" onwards
```

Merge miranda results with pcc data
```{r}
full_pcc <- results_df %>%
  inner_join(miranda_full, by = c("miRNA", "mRNA"))

length(unique(full_pcc$miRNA))
length(unique(full_pcc$mRNA))

# Save as csv 
write.csv(full_pcc, "../output/Molecular/interactions/jaccard/feb_pcc_miranda_full_binding.csv")
```

Subset by significant PCCs
```{r}
full_pcc_05 <- full_pcc %>%
  filter(Pvalue < 0.05)
```

Create network plot of the mRNA-miRNA interactions. 
```{r}
# Prepare edges
edges <- full_pcc_05[, c("miRNA", "mRNA", "PCC", "Pvalue", "score", "energy")]
colnames(edges) <- c("from", "to", "PCC", "Pvalue", "score", "energy")

miRNAs <- unique(edges$from)
mRNAs  <- unique(edges$to)

nodes <- data.frame(
  name = c(miRNAs, mRNAs),
  type = c(rep("miRNA", length(miRNAs)), rep("mRNA", length(mRNAs)))
)

# Create graph
graph <- tbl_graph(nodes = nodes, edges = edges, directed = TRUE);graph

# Plot 
ggraph(graph, layout = "fr") +  # fr = Fruchterman-Reingold layout
  geom_edge_link(aes(color = PCC), arrow = arrow(length = unit(4, 'mm')),
                 end_cap = circle(3, 'mm'), width = 0.8) +
  scale_edge_color_gradient2(low = "red", mid = "grey", high = "blue", midpoint = 0) +
  geom_node_point(aes(shape = type, color = type), size = 4) +
  geom_node_text(aes(label = name), repel = TRUE, size = 3) +
  theme_graph() +
  labs(edge_color = "PCC") +
  ggtitle("miRNA–mRNA PCC network")
```

Another version
```{r}
# Example edges — use your filtered data
edges <- full_pcc_05[, c("miRNA", "mRNA", "PCC", "Pvalue")]
colnames(edges) <- c("from", "to", "PCC", "Pvalue")

# Create igraph object
g <- graph_from_data_frame(edges, directed = FALSE)

# Add edge weight and color
E(g)$weight <- abs(E(g)$PCC)  # Absolute PCC for line thickness
E(g)$color <- ifelse(E(g)$PCC > 0, "red", "blue")  # Red = positive, Blue = negative

# Add node type attribute
V(g)$type <- ifelse(V(g)$name %in% edges$from, "miRNA", "mRNA")

# Convert to tidygraph for ggraph
g_tbl <- as_tbl_graph(g)

# Plot
p <- ggraph(g_tbl, layout = "fr") +  # Fruchterman-Reingold layout
  geom_edge_link(aes(edge_width = weight, color = color), alpha = 0.6) +
  geom_node_point(aes(color = type), size = 5) +
  # geom_node_text(aes(label = name), repel = TRUE, size = 3) +  # Uncomment to show labels
  scale_edge_width(range = c(0.5, 3)) +
  scale_edge_color_identity() +  # Use exact edge colors we set
  scale_color_manual(values = c("miRNA" = "purple", "mRNA" = "orange")) +
  theme_graph() +
  labs(
    title = "miRNA–mRNA Interaction Network",
    subtitle = "Edge width = |PCC|, Red = positive, Blue = negative"
  )

# Show plot
p
```

### 3'UTR

Read in 3'UTR binding miranda data 
```{r}
miranda_3utr <- read.delim("../output/Molecular/interactions/miranda_strict_all_1kb_apoc_shortstack_parsed.txt", header = F)
colnames(miranda_3utr) <- c("miRNA", "mRNA", "score", "energy", "query_start_end", "subject_start_end", "total_bp_shared", "query_similar", "subject_similar")

# Format miranda df 
miranda_3utr$miRNA <- sub("^>", "", miranda_3utr$miRNA)  # Remove leading ">"
miranda_3utr$miRNA <- sub("\\..*", "", miranda_3utr$miRNA)  # Remove everything from the first period onwards
miranda_3utr$mRNA <- sub(";.*", "", miranda_3utr$mRNA)  # Remove everything from "::" onwards
miranda_3utr$mRNA <- sub("ID=", "", miranda_3utr$mRNA)  # Remove everything from "::" onwards
miranda_3utr$mRNA <- sub("model", "TU", miranda_3utr$mRNA)  # Remove everything from "::" onwards
```

Merge miranda results with pcc data
```{r}
utr_pcc <- results_df %>%
  inner_join(miranda_3utr, by = c("miRNA", "mRNA"))

length(unique(utr_pcc$miRNA))
length(unique(utr_pcc$mRNA))

# Save as csv 
write.csv(utr_pcc, "../output/Molecular/interactions/jaccard/feb_pcc_miranda_3UTR_binding.csv")
```

Subset by significant PCCs
```{r}
utr_pcc_05 <- utr_pcc %>%
  filter(Pvalue < 0.05)
```

Create network plot of the mRNA-miRNA interactions. 
```{r}
# Prepare edges
edges <- utr_pcc_05[, c("miRNA", "mRNA", "PCC", "Pvalue", "score", "energy")]
colnames(edges) <- c("from", "to", "PCC", "Pvalue", "score", "energy")

miRNAs <- unique(edges$from)
mRNAs  <- unique(edges$to)

nodes <- data.frame(
  name = c(miRNAs, mRNAs),
  type = c(rep("miRNA", length(miRNAs)), rep("mRNA", length(mRNAs)))
)

# Create graph
graph <- tbl_graph(nodes = nodes, edges = edges, directed = TRUE);graph

# Plot 
ggraph(graph, layout = "fr") +  # fr = Fruchterman-Reingold layout
  geom_edge_link(aes(color = PCC), arrow = arrow(length = unit(4, 'mm')),
                 end_cap = circle(3, 'mm'), width = 0.8) +
  scale_edge_color_gradient2(low = "red", mid = "grey", high = "blue", midpoint = 0) +
  geom_node_point(aes(shape = type, color = type), size = 4) +
  geom_node_text(aes(label = name), repel = TRUE, size = 3) +
  theme_graph() +
  labs(edge_color = "PCC") +
  ggtitle("miRNA–mRNA PCC network")
```

Another version
```{r}
# Example edges — use your filtered data
edges <- utr_pcc_05[, c("miRNA", "mRNA", "PCC", "Pvalue")]
colnames(edges) <- c("from", "to", "PCC", "Pvalue")

# Create igraph object
g <- graph_from_data_frame(edges, directed = FALSE)

# Add edge weight and color
E(g)$weight <- abs(E(g)$PCC)  # Absolute PCC for line thickness
E(g)$color <- ifelse(E(g)$PCC > 0, "red", "blue")  # Red = positive, Blue = negative

# Add node type attribute
V(g)$type <- ifelse(V(g)$name %in% edges$from, "miRNA", "mRNA")

# Convert to tidygraph for ggraph
g_tbl <- as_tbl_graph(g)

# Plot
p <- ggraph(g_tbl, layout = "fr") +  # Fruchterman-Reingold layout
  geom_edge_link(aes(edge_width = weight, color = color), alpha = 0.6) +
  geom_node_point(aes(color = type), size = 5) +
  # geom_node_text(aes(label = name), repel = TRUE, size = 3) +  # Uncomment to show labels
  scale_edge_width(range = c(0.5, 3)) +
  scale_edge_color_identity() +  # Use exact edge colors we set
  scale_color_manual(values = c("miRNA" = "purple", "mRNA" = "orange")) +
  theme_graph() +
  labs(
    title = "miRNA–mRNA Interaction Network",
    subtitle = "Edge width = |PCC|, Red = positive, Blue = negative"
  )

# Show plot
p
```

## June ambient network 

Subset June ambient samples from counts dfs
```{r}
# Get TP5 ambient sample IDs
tp5_amb_ids <- meta$ID[meta$Timepoint == "TP5" & meta$Treatment == "Ambient"]

# Subset columns in counts
mirna_tp5_amb <- mirna_counts[, tp5_amb_ids]
mrna_tp5_amb  <- mrna_counts[, tp5_amb_ids]
rownames(mirna_tp5_amb) <- rownames(mirna_counts)
rownames(mrna_tp5_amb)  <- rownames(mrna_counts)
```

Perform rlog on counts 
```{r}
# Filter meta for TP5 amb
tp5_amb_meta <- meta[meta$Timepoint == "TP5" & meta$Treatment == "Ambient", ]
tp5_amb_ids <- tp5_amb_meta$ID

# Subset counts to TP5 amb samples
mirna_tp5_amb <- mirna_counts[, tp5_amb_ids]
mrna_tp5_amb  <- mrna_counts[, tp5_amb_ids]

# Add dummy colData if you don’t have any covariates
dds_mirna <- DESeqDataSetFromMatrix(countData = mirna_tp5_amb,
                                    colData = tp5_amb_meta,
                                    design = ~1)

dds_mrna  <- DESeqDataSetFromMatrix(countData = mrna_tp5_amb,
                                    colData = tp5_amb_meta,
                                    design = ~1)

# Run rlog (blind = TRUE for no design)
rlog_mirna <- assay(rlog(dds_mirna, blind = TRUE))
rlog_mrna  <- assay(rlog(dds_mrna, blind = TRUE))

# Check shapes
dim(rlog_mirna)
dim(rlog_mrna)

```

Run PCC on June ambient data 
```{r}
miRNA_ids <- rownames(rlog_mirna)
mRNA_ids  <- rownames(rlog_mrna)

# Make grid of all combos
grid <- expand.grid(miRNA = miRNA_ids, mRNA = mRNA_ids, stringsAsFactors = FALSE)

# Compute PCC and p-value
results_tp5_amb <- pbapply::pblapply(1:nrow(grid), function(i) {
  mir <- rlog_mirna[grid$miRNA[i], ]
  mr  <- rlog_mrna[grid$mRNA[i], ]
  ct  <- cor.test(mir, mr, method = "pearson")
  data.frame(miRNA = grid$miRNA[i],
             mRNA = grid$mRNA[i],
             PCC = ct$estimate,
             Pvalue = ct$p.value)
})

results_tp5_amb_df <- do.call(rbind, results_tp5_amb)
str(results_tp5_amb_df)

# format miRNA names 
results_tp5_amb_df <- results_tp5_amb_df %>%
    mutate(miRNA = str_replace(miRNA, "^([^_]*_[^_]*)_.*$", "\\1"))
```

### Full binding 

Merge full miranda results with pcc data
```{r}
full_pcc_tp5_amb <- results_tp5_amb_df %>%
  inner_join(miranda_full, by = c("miRNA", "mRNA"))

length(unique(full_pcc_tp5_amb$miRNA))
length(unique(full_pcc_tp5_amb$mRNA))


# Save as csv 
write.csv(full_pcc_tp5_amb, "../output/Molecular/interactions/jaccard/june_amb_pcc_miranda_full_binding.csv")
```

Subset by significant PCCs
```{r}
full_pcc_tp5_amb_05 <- full_pcc_tp5_amb %>%
  filter(Pvalue < 0.05)
```

Create network plot of the mRNA-miRNA interactions. 
```{r}
# Prepare edges
edges <- full_pcc_tp5_amb_05[, c("miRNA", "mRNA", "PCC", "Pvalue", "score", "energy")]
colnames(edges) <- c("from", "to", "PCC", "Pvalue", "score", "energy")

miRNAs <- unique(edges$from)
mRNAs  <- unique(edges$to)

nodes <- data.frame(
  name = c(miRNAs, mRNAs),
  type = c(rep("miRNA", length(miRNAs)), rep("mRNA", length(mRNAs)))
)

# Create graph
graph <- tbl_graph(nodes = nodes, edges = edges, directed = TRUE);graph

# Plot 
ggraph(graph, layout = "fr") +  # fr = Fruchterman-Reingold layout
  geom_edge_link(aes(color = PCC), arrow = arrow(length = unit(4, 'mm')),
                 end_cap = circle(3, 'mm'), width = 0.8) +
  scale_edge_color_gradient2(low = "red", mid = "grey", high = "blue", midpoint = 0) +
  geom_node_point(aes(shape = type, color = type), size = 4) +
  geom_node_text(aes(label = name), repel = TRUE, size = 3) +
  theme_graph() +
  labs(edge_color = "PCC") +
  ggtitle("miRNA–mRNA PCC network")
```

Another version
```{r}
# Example edges — use your filtered data
edges <- full_pcc_tp5_amb_05[, c("miRNA", "mRNA", "PCC", "Pvalue")]
colnames(edges) <- c("from", "to", "PCC", "Pvalue")

# Create igraph object
g <- graph_from_data_frame(edges, directed = FALSE)

# Add edge weight and color
E(g)$weight <- abs(E(g)$PCC)  # Absolute PCC for line thickness
E(g)$color <- ifelse(E(g)$PCC > 0, "red", "blue")  # Red = positive, Blue = negative

# Add node type attribute
V(g)$type <- ifelse(V(g)$name %in% edges$from, "miRNA", "mRNA")

# Convert to tidygraph for ggraph
g_tbl <- as_tbl_graph(g)

# Plot
p <- ggraph(g_tbl, layout = "fr") +  # Fruchterman-Reingold layout
  geom_edge_link(aes(edge_width = weight, color = color), alpha = 0.6) +
  geom_node_point(aes(color = type), size = 5) +
  # geom_node_text(aes(label = name), repel = TRUE, size = 3) +  # Uncomment to show labels
  scale_edge_width(range = c(0.5, 3)) +
  scale_edge_color_identity() +  # Use exact edge colors we set
  scale_color_manual(values = c("miRNA" = "purple", "mRNA" = "orange")) +
  theme_graph() +
  labs(
    title = "miRNA–mRNA Interaction Network",
    subtitle = "Edge width = |PCC|, Red = positive, Blue = negative"
  )

# Show plot
p
```

### 3'UTR

Merge full miranda results with pcc data
```{r}
utr_pcc_tp5_amb <- results_tp5_amb_df %>%
  inner_join(miranda_3utr, by = c("miRNA", "mRNA"))

length(unique(utr_pcc_tp5_amb$miRNA))
length(unique(utr_pcc_tp5_amb$mRNA))


# Save as csv 
write.csv(utr_pcc_tp5_amb, "../output/Molecular/interactions/jaccard/june_amb_pcc_miranda_3UTR_binding.csv")
```








## Comparison of February and June ambient

### Full binding 

Add pair ID to each df and then select the pairs and their associated PCC
```{r}
# Add pair ID to each df
full_pcc <- full_pcc %>%
  mutate(pair = paste(miRNA, mRNA, sep = "_"))

full_pcc_tp5_amb <- full_pcc_tp5_amb %>%
  mutate(pair = paste(miRNA, mRNA, sep = "_"))

# Bring Feb PCC
df_feb <- full_pcc %>%
  select(pair, PCC_feb = PCC)

# Bring TP5 PCC
df_tp5 <- full_pcc_tp5_amb %>%
  select(pair, PCC_tp5 = PCC)
```

Join dfs and calculate Jaccard index 
```{r}
# Combine dfs
jaccard_df <- full_join(df_feb, df_tp5, by = "pair")

# Calculate jaccard index
jaccard_df <- jaccard_df %>%
  mutate(
    abs_feb  = abs(PCC_feb),
    abs_tp5  = abs(PCC_tp5),
    weighted_jaccard = pmin(abs_feb, abs_tp5) / pmax(abs_feb, abs_tp5),
    sign_change = sign(PCC_feb) != sign(PCC_tp5)
  )

# Assess what the correlation is doing based on sign change and jaccard
jaccard_df <- jaccard_df %>%
  mutate(
    stability = case_when(
      is.na(PCC_feb) | is.na(PCC_tp5) ~ "One missing",
      sign_change ~ "Sign flip",
      weighted_jaccard >= 0.8 ~ "Stable same sign",
      weighted_jaccard < 0.8 ~ "Diverging same sign"
    )
  )

jaccard_df <- unique(jaccard_df)

# Show summary table
table(jaccard_df$stability)
```

Subset by sign flip 
```{r}
jaccard_df_flip <- jaccard_df %>%
  filter(stability == "Sign flip") %>%
  separate(pair, into = c("miRNA", "mRNA"), sep = "_evm\\.", remove = FALSE)
jaccard_df_flip$mRNA <- paste0("evm.", jaccard_df_flip$mRNA)
```

Read in mRNA DE results from Feb and June ambient 
```{r}
de_tp0_tp5_amb <- read.csv("../output/Molecular/mRNA/TP0_amb_v_TP5_amb_DEG.csv")[,1:7]
```

Join DE results with jaccard 
```{r}
test <- de_tp0_tp5_amb %>%
  inner_join(jaccard_df_flip, by = c("X" = "mRNA"))
test<- unique(test)

# DEGs where the miRNA-mRNA correlation flips sign or changes greatly in magnitude.
highlights <- test %>%
  filter(sign_change == TRUE) %>%
  filter(weighted_jaccard < 0.2) %>%
  mutate(
    seasonal_direction = case_when(
      log2FoldChange > 0 ~ "Up in February",
      log2FoldChange < 0 ~ "Up in June",
      TRUE               ~ "No Change"
    )
  )

table(highlights$seasonal_direction, highlights$stability)
ggplot(highlights, aes(x = seasonal_direction, fill = stability)) +
  geom_bar(position = "dodge") +
  labs(title = "Seasonal DEGs with miRNA-mRNA Regulatory Rewiring",
       x = "Gene Upregulation Season", y = "Number of Pairs", fill = "Regulatory Stability") +
  theme_classic()

heatmap_mat <- highlights %>%
  select(pair, PCC_feb, PCC_tp5) %>%
  tibble::column_to_rownames("pair") %>%
  as.matrix()

pheatmap(heatmap_mat, cluster_rows = TRUE, cluster_cols = FALSE,
         main = "miRNA-mRNA Correlations: Feb vs June (Top Rewired Pairs)",
         labels_col = c("February", "June"))
```

Filter for genes downregulated in Feb and have negative PCCs 
```{r}
blah <- highlights %>%
  filter(log2FoldChange < 0) %>%
  filter(PCC_feb < 0)
```

Read in annot and merge 
```{r}
annot <- read.delim("/Users/jillashey/Desktop/PutnamLab/Astrangia_Genome/Apoculata_v2.0_GeneAnnotation_combined_prelim.txt", header = T)
annot$Protein_ID <- gsub("model", "TU", annot$Protein_ID)

blah2 <- blah %>%
  inner_join(annot, by = c("X" = "Protein_ID"))
```

Potential visualizations 
```{r}
heatmap_data <- blah %>%
  dplyr::select(pair, PCC_feb, PCC_tp5) %>%
  tibble::column_to_rownames("pair")

pheatmap(as.matrix(heatmap_data), cluster_rows = TRUE, cluster_cols = FALSE,
         main = "miRNA-mRNA Correlations: February vs June",
         labels_col = c("February", "June"))

blah$cor_diff <- abs(blah$PCC_feb - blah$PCC_tp5)
ggplot(blah, aes(x = log2FoldChange, y = cor_diff, color = sign_change)) +
  geom_point(size=3, alpha=0.7) +
  labs(title="Expression Change vs. Regulatory Change by Pair",
       x="log2 Fold Change (Feb vs June)", y="Abs Change in PCC",
       color = "Sign Flip?") +
  theme_minimal()

pair1 <- blah$pair[1]
df <- blah %>%
  filter(pair == pair1) %>%
  tidyr::pivot_longer(cols = c(PCC_feb, PCC_tp5),
                      names_to = "Timepoint", values_to = "Correlation")
ggplot(df, aes(x = Timepoint, y = Correlation, group = 1)) +
  geom_line(size=1.5) +
  geom_point(size=3) +
  labs(title=paste("Correlation Dynamics for", pair1), 
       x="Timepoint", y="Pearson Correlation") +
  theme_classic()

ggplot(highlights, aes(x = pair, y = 1, fill = PCC_feb)) +
  geom_tile() +
  geom_tile(aes(y = 2, fill = PCC_tp5)) +
  scale_fill_gradient2(midpoint=0, low="blue", mid="white", high="red") +
  facet_grid(~seasonal_direction, scales = "free_x", space = "free_x") +
  labs(
    title = "miRNA-mRNA Correlation Dynamics for DEGs",
    x = "miRNA–Gene Pair", y = "Timepoint",
    fill = "Pearson Corr."
  ) +
  theme(axis.text.x = element_text(angle=90, hjust=1))
```

### 3UTR binding 

Add pair ID to each df and then select the pairs and their associated PCC
```{r}
# Add pair ID to each df
utr_pcc <- utr_pcc %>%
  mutate(pair = paste(miRNA, mRNA, sep = "_"))

utr_pcc_tp5_amb <- utr_pcc_tp5_amb %>%
  mutate(pair = paste(miRNA, mRNA, sep = "_"))

# Bring Feb PCC
df_feb_3utr <- utr_pcc %>%
  select(pair, PCC_feb = PCC)

# Bring TP5 PCC
df_tp5_amb_3utr <- utr_pcc_tp5_amb %>%
  select(pair, PCC_tp5 = PCC)
```

Join dfs and calculate Jaccard index 
```{r}
# Combine dfs
jaccard_df_3utr <- full_join(df_feb_3utr, df_tp5_amb_3utr, by = "pair")

# Calculate jaccard index
jaccard_df_3utr <- jaccard_df_3utr %>%
  mutate(
    abs_feb  = abs(PCC_feb),
    abs_tp5  = abs(PCC_tp5),
    weighted_jaccard = pmin(abs_feb, abs_tp5) / pmax(abs_feb, abs_tp5),
    sign_change = sign(PCC_feb) != sign(PCC_tp5)
  )

# Assess what the correlation is doing based on sign change and jaccard
jaccard_df_3utr <- jaccard_df_3utr %>%
  mutate(
    stability = case_when(
      is.na(PCC_feb) | is.na(PCC_tp5) ~ "One missing",
      sign_change ~ "Sign flip",
      weighted_jaccard >= 0.8 ~ "Stable same sign",
      weighted_jaccard < 0.8 ~ "Diverging same sign"
    )
  )

jaccard_df_3utr <- unique(jaccard_df_3utr)

# Show summary table
table(jaccard_df_3utr$stability)
```

Subset by sign flip 
```{r}
jaccard_df_3utr_flip <- jaccard_df_3utr %>%
  filter(stability == "Sign flip") %>%
  separate(pair, into = c("miRNA", "mRNA"), sep = "_evm\\.", remove = FALSE)
jaccard_df_3utr_flip$mRNA <- paste0("evm.", jaccard_df_3utr_flip$mRNA)
```

Join DE results with jaccard 
```{r}
test_3utr <- de_tp0_tp5_amb %>%
  inner_join(jaccard_df_3utr_flip, by = c("X" = "mRNA"))
test_3utr<- unique(test_3utr)

# DEGs where the miRNA-mRNA correlation flips sign or changes greatly in magnitude.
highlights_3utr <- test_3utr %>%
  filter(sign_change == TRUE) %>%
  filter(weighted_jaccard < 0.2) %>%
  mutate(
    seasonal_direction = case_when(
      log2FoldChange > 0 ~ "Up in February",
      log2FoldChange < 0 ~ "Up in June",
      TRUE               ~ "No Change"
    )
  )
```

Join annot results 
```{r}
hello <- highlights_3utr %>%
   inner_join(annot, by = c("X" = "Protein_ID"))
```








20250722

Read in data
```{r}
june <- read.csv("../output/Molecular/interactions/jaccard/june_amb_pcc_miranda_3UTR_binding.csv")
feb <- read.csv("../output/Molecular/interactions/jaccard/feb_pcc_miranda_3UTR_binding.csv")

# Select only mRNA, miRNA and weight (ie PCC)
june <- june %>%
  select(miRNA, mRNA, PCC) %>%
  unique()
feb <- feb %>%
  select(miRNA, mRNA, PCC) %>%
  unique()
```

```{r}
merged_df <- merge(feb, june, by = c("miRNA", "mRNA"), suffixes = c("_feb", "_june"))
head(merged_df)
merged_df$sign_feb  <- sign(merged_df$PCC_feb)
merged_df$sign_june <- sign(merged_df$PCC_june)
merged_df$direction_changed <- merged_df$sign_feb != merged_df$sign_june
merged_df$strength_change <- abs(merged_df$PCC_june - merged_df$PCC_feb)
threshold <- 0.5
merged_df$large_strength_change <- merged_df$strength_change > threshold
which_changed <- merged_df[merged_df$direction_changed, ]

# Filter top pairs for table/figure
candidate_pairs <- merged_df[
  merged_df$direction_changed | merged_df$large_strength_change, 
]
# Optionally further filter or rank by strength_change
top_pairs <- candidate_pairs[order(-candidate_pairs$strength_change), ][1:20, ]
```

Read in mRNA DE results from Feb and June ambient 
```{r}
de_tp0_tp5_amb <- read.csv("../output/Molecular/mRNA/TP0_amb_v_TP5_amb_DEG.csv")[,1:7]
```

```{r}
blah <- de_tp0_tp5_amb %>%
  inner_join(candidate_pairs, by = c("X" = "mRNA")) %>%
  filter(direction_changed == TRUE & large_strength_change == TRUE)
blah<- unique(blah)
```












Create igraph network objects
```{r}
net_feb <- graph_from_data_frame(feb, directed=FALSE)
net_june <- graph_from_data_frame(june, directed=FALSE)

# Very important: edge order in graph matches edge list order to assign correct weights
E(net_feb)$weight <- feb$PCC
E(net_june)$weight <- june$PCC
```

Calculate network metrics 
```{r}
network_metrics <- function(net) {
  data.frame(
    nodes = vcount(net),
    edges = ecount(net),
    density = edge_density(net),
    average_degree = mean(degree(net)),
    weighted_average_strength = mean(strength(net, weights = E(net)$weight)),
    weighted_clustering = transitivity(net, type = "weighted", weights = abs(E(net)$weight))
  )
}

metrics_feb <- network_metrics(net_feb)
metrics_june <- network_metrics(net_june)

metrics_feb
metrics_june

```




```{r}
hmm <- strength(net_feb, vids = V(net_feb), weights = E(net_feb)$PCC)
strength(net_jun, vids = V(net_jun), weights = E(net_jun)$PCC)


summary(feb$PCC)
summary(june$PCC)
hist(feb$PCC, main="Feb PCC", xlab="PCC")
hist(june$PCC, main="June PCC", xlab="PCC")
```



I want to summarize by miRNA to get avg PCC, number of mRNAs interacted with 
```{r}
library(dplyr)

# For February
feb_summary <- feb %>%
  group_by(miRNA) %>%
  summarise(n_mRNAs = n(),
            avg_PCC = mean(PCC, na.rm = TRUE),
            median_PCC = median(PCC, na.rm = TRUE))

# For June
june_summary <- june %>%
  group_by(miRNA) %>%
  summarise(n_mRNAs = n(),
            avg_PCC = mean(PCC, na.rm = TRUE),
            median_PCC = median(PCC, na.rm = TRUE))

summary_merged <- full_join(
  feb_summary %>% rename(n_mRNAs_feb = n_mRNAs, avg_PCC_feb = avg_PCC),
  june_summary %>% rename(n_mRNAs_june = n_mRNAs, avg_PCC_june = avg_PCC),
  by = "miRNA"
)

# For plotting:
ggplot(summary_merged, aes(x=avg_PCC_feb, y=avg_PCC_june)) +
  geom_point() +
  xlab("Avg PCC Feb") + ylab("Avg PCC June") +
  ggtitle("Average PCC per miRNA: Feb vs. June")

```

```{r}
library(igraph)

# We must ensure the 'PCC' is properly set as edge weights
E(net_feb)$weight <- feb$PCC
E(net_jun)$weight <- june$PCC

# Weighted node strength (sum of PCCs for each node)
feb_miRNA_strength <- strength(net_feb, vids=V(net_feb), weights=E(net_feb)$weight, mode="out")
june_miRNA_strength <- strength(net_jun, vids=V(net_jun), weights=E(net_jun)$weight, mode="out")

# Weighted clustering coefficient (uses abs(weight) to avoid negatives disrupting the sum)
feb_cluster <- transitivity(net_feb, type="weighted", weights=abs(E(net_feb)$weight))
june_cluster <- transitivity(net_jun, type="weighted", weights=abs(E(net_jun)$weight))

```

```{r}
net_feb <- graph_from_data_frame(feb, directed=FALSE)
E(net_feb)$weight <- feb$PCC

# Example weighted degree (strength)
strength_feb <- strength(net_feb, weights=E(net_feb)$weight)

# Clustering coefficient
transitivity(net_feb, type="weighted", weights=abs(E(net_feb)$weight))

```



