---
title: "WGCNA miRNA"
author: "Jill Ashey"
date: "2025-05-05"
output: html_document
---

WGCNA analysis of miRNA data for AST 2021 project. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# The following setting is important, do not omit.
options(stringsAsFactors = FALSE) #Set Strings to character
```

Load required libraries.  
```{r}
if ("tidyverse" %in% rownames(installed.packages()) == 'FALSE') install.packages('tidyverse') 
if ("genefilter" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install("genefilter") 
if ("DESeq2" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('DESeq2') 
if ("RColorBrewer" %in% rownames(installed.packages()) == 'FALSE') install.packages('RColorBrewer') 
if ("impute" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('impute') 
if ("preprocessCore" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('preprocessCore') 
if ("WGCNA" %in% rownames(installed.packages()) == 'FALSE') install.packages('WGCNA') 
if ("flashClust" %in% rownames(installed.packages()) == 'FALSE') install.packages('flashClust') 
if ("gridExtra" %in% rownames(installed.packages()) == 'FALSE') install.packages('gridExtra') 
if ("ComplexHeatmap" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('ComplexHeatmap') 
if ("goseq" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('goseq')
if ("dplyr" %in% rownames(installed.packages()) == 'FALSE') install.packages('dplyr') 
if ("clusterProfiler" %in% rownames(installed.packages()) == 'FALSE') BiocManager::install('clusterProfiler') 
if ("pheatmap" %in% rownames(installed.packages()) == 'FALSE') install.packages('pheatmap') 
if ("magrittr" %in% rownames(installed.packages()) == 'FALSE') install.packages('magrittr') 
if ("vegan" %in% rownames(installed.packages()) == 'FALSE') install.packages('vegan') 
if ("factoextra" %in% rownames(installed.packages()) == 'FALSE') install.packages('factoextra') 

library("tidyverse")
library("genefilter")
library("DESeq2")
library("RColorBrewer")
library("impute")
library("preprocessCore")
library("WGCNA")
library("flashClust")
library("gridExtra")
library("ComplexHeatmap")
library("goseq")
library("dplyr")
library("clusterProfiler")
library("pheatmap")
library("magrittr")
library("vegan")
library("factoextra")
library("dplyr")
```

Read in data and select only counts data 
```{r}
counts <- read.table("../output/Molecular/smRNA/shortstack/Counts.csv", header = T) 
```

Select only miRNAs with a Y in MIRNA column 
```{r}
counts <- counts %>%
  dplyr::filter(MIRNA == "Y")
```

Should be 51 miRNAs.

Combine Name and Coords columns for miRNA name. Remove MIRNA column 
```{r}
counts <- counts %>%
  mutate(miRNA_name = paste(Name, Coords, sep = "_")) %>%
  dplyr::select(-c(Name, Coords, MIRNA))
```

Make miRNA_name the row names and remove it as a column 
```{r}
rownames(counts) <- counts$miRNA_name
counts <- counts %>%
  dplyr::select(-miRNA_name)
```

Remove extra info from column names
```{r}
colnames(counts) <- gsub("_R1_001.fastq.gz_1", "", colnames(counts))
colnames(counts) <- gsub("AST.", "AST-", colnames(counts))
```

Read in metadata
```{r}
meta <- read.csv("../data/Molecular/RNA_metadata.csv") %>%
 dplyr::arrange(ID)
head(meta)

# Remove AST-1105
meta_clean <- meta %>%
  filter(!ID == "AST-1105") %>%
  mutate(Treatment = ifelse(Treatment == "Acclimation", "Ambient", Treatment))
```

Reorder columns in filtered counts matrix so that it matches the metadata
```{r}
counts <- counts[, meta_clean$ID]

# Sanity check! 
meta_clean$ID %in% colnames(counts) #are all of the sample names (rows) in the metadata df in the gene count matrix? Should be TRUE. 
all(rownames(meta_clean$ID) == colnames(counts)) #are they the same in the same order? Should be TRUE
```

Merge Treatment and Timepoint into a single column and set levels of factor. 
```{r}
meta_clean$Condition <- paste(meta_clean$Timepoint, meta_clean$Treatment, sep = "_")
meta_clean$Condition <- factor(meta_clean$Condition, levels = c("TP0_Ambient", "TP5_Ambient", "TP5_Heat", "TP7_Ambient", "TP7_Heat"))
```

Construct DESeq data set
```{r}
gdds <- DESeqDataSetFromMatrix(countData = counts, colData = meta_clean, design = ~Condition)
```

First we are going to log-transform the data using a variance stabilizing transforamtion (VST). This is only for visualization purposes. Essentially, this is roughly similar to putting the data on the log2 scale. It will deal with the sampling variability of low counts by calculating within-group variability (if blind=FALSE). Importantly, it does not use the design to remove variation in the data, and so can be used to examine if there may be any variability do to technical factors such as extraction batch effects.

To do this we first need to calculate the size factors of our samples. This is a rough estimate of how many reads each sample contains compared to the others. In order to use VST (the faster log2 transforming process) to log-transform our data, the size factors need to be less than 4. Otherwise, there could be artefacts in our results.
```{r}
SF.gdds <- estimateSizeFactors(gdds) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than for to use vst
print(sizeFactors(SF.gdds)) #View size factors
```

Size factors are all less than 4, so we can use VST!
```{r}
gvst <- varianceStabilizingTransformation(gdds, blind=FALSE) #apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
head(assay(gvst), 3) #view transformed gene count data
```

Plot a heatmap of sample-to-sample distances
```{r}
gsampleDists <- dist(t(assay(gvst))) #calculate distance matix
gsampleDistMatrix <- as.matrix(gsampleDists) #distance matrix
rownames(gsampleDistMatrix) <- colnames(gvst) #assign row names
colnames(gsampleDistMatrix) <- NULL #assign col names
pheatmap(gsampleDistMatrix, #plot matrix
         clustering_distance_rows=gsampleDists, #cluster rows
         clustering_distance_cols=gsampleDists) #cluster columns
```

Compile WGCNA dataset 

Transpose the filtered gene count matrix so that the gene IDs are rows and the sample IDs are columns.
```{r}
datExpr <- as.data.frame(t(assay(gvst))) #transpose to output to a new data frame with the column names as row names. And make all data numeric
```

Look for outliers by examining sample clusters
```{r}
sampleTree = hclust(dist(datExpr), method = "average");
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
# The user should change the dimensions if the window is too large or too small.
plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5, cex.axis = 1.5, cex.main = 2)
```

Network construction and consensus module detection

First choose a soft-threshold power. The soft thresholding power (Î²) is the number to which the co-expression similarity is raised to calculate adjacency. The function pickSoftThreshold performs a network topology analysis. The user chooses a set of candidate powers, however the default parameters are suitable values.
```{r, message=FALSE, warning=FALSE}
##Soft threshold
dim(datExpr) #  19 51
# Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12, to=20, by=2))
# Call the network topology analysis function

#the below takes a long time to run, so is commented out and the pre-run results are loaded in below. I need to save it to my desktop because it is too large to push to github. Instead, it will go on OSF
sft <- pickSoftThreshold(datExpr, powerVector = powers, verbose = 5) #...wait for this to finish
save(sft, file = "~/Desktop/PutnamLab/Astrangia/Molecular/sft_miRNA.RData")
load("~/Desktop/PutnamLab/Astrangia/Molecular/sft_miRNA.RData")

# pickSoftThreshold 
#  performs the analysis of network topology and aids the
# user in choosing a proper soft-thresholding power.
# The user chooses a set of candidate powers (the function provides suitable default values)
# function returns a set of network indices that should be inspected

sizeGrWindow(9, 5) # set window size 
# png to output 
png("../output/Molecular/smRNA/shortstack/WGCNA/sft_miRNA.png", 1000, 1000, pointsize=20)
par(mfrow = c(1,2));
cex1 = 0.9;
# # Scale-free topology fit index as a function of the soft-thresholding power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
    main = paste("Scale independence"));
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
    labels=powers,cex=cex1,col="red");
# # this line corresponds to using an R^2 cut-off
abline(h=0.8,col="red")
# # Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
    xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
    main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
dev.off() # output 
#I used a scale-free topology fit index **R^2 of 0.85**. This lowest recommended R^2 by Langfelder and Horvath is 0.8. I chose 0.85 because we want to use the smallest soft thresholding power that maximizes with model fit. It appears that our **soft thresholding power is 14** because it is the loweest power before the R^2=0.8 threshold that maximizes with model fit.
```

## Step wise module construction 

#### Step 1: Create adjacency matrix 

```{r}
softPower = 8 # set the soft threshold based on the plots above 

# signed 
#to get this to work with a ton of genes >30K, I had to increase my memory limit using: memory.limit(size = 35000) 
adjacency_sign = adjacency(datExpr, power = softPower, type="signed")  #Calculate adjacency
```

###  Step 2: Turn adjacency into topological overlap: Calculation of the topological overlap matrix, (TOM) and the corresponding dissimilarity, from a given adjacency matrix.

```{r}
#the below takes a long time to run, so is commented out and the pre-run results are loaded in below
TOM_sign = TOMsimilarity(adjacency_sign, TOMType="signed") #Translate adjacency into topological overlap matrix
save(TOM_sign, file = "~/Desktop/PutnamLab/Astrangia/Molecular/TOM_sign_miRNA.Rdata")
load("~/Desktop/PutnamLab/Astrangia/Molecular/TOM_sign_miRNA.Rdata")

dissTOM_sign = 1-TOM_sign
```

###  Step 3: Call the hierarchical clustering function - plot the tree

```{r}
# Call the hierarchical clustering function
#to get this to work, I had to increase my memory limit using: memory.limit(size = 45000) 
geneTree_sign   = hclust(as.dist(dissTOM_sign), method = "average");

# Plot the resulting clustering tree (dendrogram) Each leaf corresponds to a gene, branches grouping together densely are interconnected, highly co-expressed genes.  
sizeGrWindow(12,9)

plot(geneTree_sign, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity - SIGNED",
     labels = FALSE, hang = 0.04)
```

###  Step 4: Set module size and 'cutreeDynamic' to create clusters 

```{r}
#Module identification is essentially cutting the branches off the tree in the dendrogram above. We like large modules, so we set the **minimum module size** relatively high, so we will set the minimum size at 30. 
minModuleSize = 2; # set this for the subseqent call...

dynamicMods_sign = cutreeDynamic(dendro = geneTree_sign, distM = dissTOM_sign,
                            deepSplit = 1, pamRespectsDendro = FALSE,
                            minClusterSize = minModuleSize);
table(dynamicMods_sign) # number of genes per module. Module 0 is reserved for unassigned genes. The are other modules will be listed largest to smallest. 
```

###  Step 5: convert numeric network to colors and plot the dendrogram

```{r} 
# Convert numeric lables into colors
dynamicColors_sign = labels2colors(dynamicMods_sign) # add colors to module labels (previously numbers)
table(dynamicColors_sign) # lets look at this table...
# Plot the dendrogram and colors underneath

plotDendroAndColors(geneTree_sign, dynamicColors_sign, "Dynamic Tree Cut - SIGNED",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene dendrogram and module colors 'SIGNED'")

sizeGrWindow(8,6)
png("../output/Molecular/smRNA/shortstack/WGCNA/GeneDendrogram_miRNA.png", 1000, 1000, pointsize=20)
plotDendroAndColors(geneTree_sign, dynamicColors_sign, "Dynamic Tree Cut - SIGNED",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene dendrogram and module colors 'SIGNED'")
dev.off()
```

###  Step 6: Calculate Eigengenes - view thier connectivity based on 'MEDiss = 1-cor(MEs)'

```{r}
# Calculate eigengenes
MEList = moduleEigengenes(datExpr, colors = dynamicColors_sign, softPower = 8)

MEs = MEList$eigengenes
MEs

library(dplyr)
MEs <- MEs %>% 
    select_if(~ !any(is.na(.)))

# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs, use = "pairwise.complete.obs");

# Cluster module eigengenes
METree = hclust(as.dist(MEDiss), method = "average");
# Plot the result
sizeGrWindow(7, 6)
png("../output/Molecular/smRNA/shortstack/WGCNA/ClusterEigengenes_mRNA.png", 1000, 1000, pointsize=20)
plot(METree, main = "Clustering of module eigengenes - SIGNED (dissimilarity calc = MEDiss = 1-cor(MEs))",
     xlab = "", sub = "")
MEDissThres = 0.15 
abline(h=MEDissThres, col = "red")
dev.off()
```

###  Step 7: Specify the cut line for the dendrogram (module) - Calc MODULE EIGENGENES (mergeMEs)

```{r}
MEDissThres = 0.15 # **Merge modules with >85% eigengene similarity.** Most studies use somewhere between 80-90% similarity. I will use 85% similarity as my merging threshold.
# Plot the cut line into the dendrogram
#abline(h=MEDissThres, col = "red")
# Call an automatic merging function
# merge = mergeCloseModules(dds.d0_vst, dynamicColors, cutHeight = MEDissThres, verbose = 3)
merge = mergeCloseModules(datExpr, dynamicColors_sign, cutHeight = MEDissThres, verbose = 3)
# The merged module colors
mergedColors = merge$colors;
# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs;
library(dplyr)
mergedMEs <- mergedMEs %>% 
    select_if(~ !any(is.na(.)))
# Cluster module eigengenes
MEDiss2 = 1-cor(mergedMEs,use = 'pairwise.complete.obs');
MEDiss2
METree2 = hclust(as.dist(MEDiss2), method = "average");
# Plot the result
plot(METree2, main = "Clustering of module eigengenes - SIGNED (dissimilarity calc = MEDiss = 1-cor(MEs))",
     xlab = "", sub = "")

sizeGrWindow(7, 6)
png("../output/Molecular/smRNA/shortstack/WGCNA/ClusterEigengenes_merged_miRNA.png", 1000, 1000, pointsize=20)
plot(METree2, main = "Clustering of module eigengenes - SIGNED (dissimilarity calc = MEDiss = 1-cor(MEs))",
     xlab = "", sub = "")
dev.off()
```

###  Step 8: Plot dendrogram with the cut line 'MEDissThres' 

```{r}
plotDendroAndColors(geneTree_sign, cbind(dynamicColors_sign, mergedColors),
                    c("Dynamic Tree Cut", "Merged dynamic"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)

sizeGrWindow(12, 9)

png("../output/Molecular/smRNA/shortstack/WGCNA/ClusterDendrogram_signed_mRNA.png", 1000, 1000, pointsize=20)
plotDendroAndColors(geneTree_sign, cbind(dynamicColors_sign, mergedColors),
                    c("Dynamic Tree Cut", "Merged dynamic"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
dev.off()
```

###  Step 9: Commit to mergedcolors as 'MEs' and 'moduleColors'
```{r}
# Rename to moduleColors
moduleColors = mergedColors
# Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels = match(moduleColors, colorOrder)-1;
MEs = mergedMEs;
# Save module colors and labels for use in subsequent parts
save(MEs, moduleLabels, moduleColors, geneTree_sign, file = "~/Desktop/PutnamLab/Astrangia/Molecular/networkConstruction-stepByStep_miRNA.RData")
# write csv - save the module eigengenes
write.csv(MEs, file = "../output/Molecular/smRNA/shortstack/WGCNA/WGCNA_ModuleEigengenes_miRNA.csv")
table(mergedColors)
```

Join `mergedColors` and `filt_counts_clean`
```{r}
counts$Gene <- rownames(counts)
counts$Module <- mergedColors
write.csv(counts, file = "../output/Molecular/smRNA/shortstack/WGCNA/miRNA_modules.csv")

# Double check the numbers match above 
module_counts <- table(counts$Module)
module_counts
```

